// @gui-agent/operator-aio/dist/AIOHybridOperator.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Operator } from "@gui-agent/shared/base";
import { ConsoleLogger, LogLevel } from "@agent-infra/logger";
import { Base64ImageParser } from "@agent-infra/media-utils";
import { AIOComputer, keyNameMap } from "./AIOComputer.mjs";
import { AIOBrowser } from "./AIOBrowser.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const defaultLogger = new ConsoleLogger(void 0, LogLevel.DEBUG);
class AIOHybridOperator extends Operator {
    async initialize() {
        var _this_aioBrowser;
        await this.aioComputer.screenshot(0);
        this.aioBrowser = await AIOBrowser.create({
            baseURl: this.options.baseURL,
            logger: this.logger
        });
        await (null == (_this_aioBrowser = this.aioBrowser) ? void 0 : _this_aioBrowser.launch({
            timeout: 1000,
            defaultViewport: {
                width: this.screenshotWidth,
                height: this.screenshotHeight
            }
        }));
        this.logger.info('AIOBrowser initialized successfully');
    }
    supportedActions() {
        return [
            'navigate',
            'navigate_back',
            'wait',
            'mouse_move',
            'click',
            'double_click',
            'right_click',
            'middle_click',
            'drag',
            'type',
            'hotkey',
            'press',
            'scroll',
            'call_user',
            'finished'
        ];
    }
    screenContext() {
        return {
            screenWidth: this.screenshotWidth,
            screenHeight: this.screenshotHeight,
            scaleX: 1,
            scaleY: 1
        };
    }
    async screenshot() {
        this.logger.info('Taking Screenshot...');
        try {
            var _result_data;
            const result = await this.aioComputer.screenshot();
            if (!result.success) throw new Error(`Screenshot failed: ${result.message || 'Unknown error'}`);
            if (null == (_result_data = result.data) ? void 0 : _result_data.base64) {
                var _result_data1;
                const base64Tool = new Base64ImageParser(null == (_result_data1 = result.data) ? void 0 : _result_data1.base64);
                const dimensions = base64Tool.getDimensions();
                if (dimensions) {
                    this.screenshotWidth = null == dimensions ? void 0 : dimensions.width;
                    this.screenshotHeight = null == dimensions ? void 0 : dimensions.height;
                }
                this.logger.info('Screenshot size:', JSON.stringify(dimensions));
                return {
                    status: 'success',
                    base64: result.data.base64,
                    url: await this.getCurrentUrl()
                };
            }
            throw new Error('No base64 image data received from screenshot API');
        } catch (error) {
            this.logger.error('Screenshot failed:', error);
            throw error;
        }
    }
    async execute(params) {
        const { actions } = params;
        for (const action of actions){
            this.logger.info('Execute action', action);
            await this.singleActionExecutor(action);
        }
        return {
            status: 'success'
        };
    }
    async singleActionExecutor(action) {
        const { type: actionType, inputs: actionInputs } = action;
        this.logger.info('Executing action', actionType, actionInputs, ', screen context', await this.getScreenContext());
        let startXPercent = null, startYPercent = null;
        try {
            switch(actionType){
                case 'navigate':
                    {
                        var _this_aioBrowser;
                        const { url, content } = actionInputs;
                        if (!url && !content) throw new Error('url is required when navigate');
                        this.logger.info('Navigating to', url || content);
                        await (null == (_this_aioBrowser = this.aioBrowser) ? void 0 : _this_aioBrowser.handleNavigate({
                            url: url || content
                        }));
                        break;
                    }
                case 'navigate_back':
                    var _this_aioBrowser1;
                    this.logger.info('Navigating back');
                    await (null == (_this_aioBrowser1 = this.aioBrowser) ? void 0 : _this_aioBrowser1.handleNavigateBack());
                    break;
                case 'move':
                case 'move_to':
                case 'mouse_move':
                case 'hover':
                    {
                        var _point_normalized, _point_normalized1;
                        const { point } = actionInputs;
                        if (!point) throw new Error('point is required when mouse move');
                        const { realX, realY } = await this.calculateRealCoords(point);
                        await this.aioComputer.moveTo(realX, realY);
                        startXPercent = null == (_point_normalized = point.normalized) ? void 0 : _point_normalized.x;
                        startYPercent = null == (_point_normalized1 = point.normalized) ? void 0 : _point_normalized1.y;
                        break;
                    }
                case 'click':
                case 'left_click':
                case 'left_single':
                    {
                        var _point_normalized2, _point_normalized3;
                        await this.handleClick(actionInputs, 'left');
                        const { point } = actionInputs;
                        startXPercent = null == point ? void 0 : null == (_point_normalized2 = point.normalized) ? void 0 : _point_normalized2.x;
                        startYPercent = null == point ? void 0 : null == (_point_normalized3 = point.normalized) ? void 0 : _point_normalized3.y;
                        break;
                    }
                case 'left_double':
                case 'double_click':
                    {
                        var _point_normalized4, _point_normalized5;
                        await this.handleClick(actionInputs, 'left', 2);
                        const { point } = actionInputs;
                        startXPercent = null == point ? void 0 : null == (_point_normalized4 = point.normalized) ? void 0 : _point_normalized4.x;
                        startYPercent = null == point ? void 0 : null == (_point_normalized5 = point.normalized) ? void 0 : _point_normalized5.y;
                        break;
                    }
                case 'right_click':
                case 'right_single':
                    {
                        var _point_normalized6, _point_normalized7;
                        await this.handleClick(actionInputs, 'right');
                        const { point } = actionInputs;
                        startXPercent = null == point ? void 0 : null == (_point_normalized6 = point.normalized) ? void 0 : _point_normalized6.x;
                        startYPercent = null == point ? void 0 : null == (_point_normalized7 = point.normalized) ? void 0 : _point_normalized7.y;
                        break;
                    }
                case 'middle_click':
                    {
                        var _point_normalized8, _point_normalized9;
                        await this.handleClick(actionInputs, 'middle');
                        const { point } = actionInputs;
                        startXPercent = null == point ? void 0 : null == (_point_normalized8 = point.normalized) ? void 0 : _point_normalized8.x;
                        startYPercent = null == point ? void 0 : null == (_point_normalized9 = point.normalized) ? void 0 : _point_normalized9.y;
                        break;
                    }
                case 'left_click_drag':
                case 'drag':
                case 'select':
                    await this.handleDrag(actionInputs);
                    break;
                case 'type':
                    {
                        var _actionInputs_content;
                        const content = null == (_actionInputs_content = actionInputs.content) ? void 0 : _actionInputs_content.trim();
                        if (!content) throw new Error('content is required when type');
                        const stripContent = content.replace(/\\n$/, '').replace(/\n$/, '');
                        await this.aioComputer.type(stripContent);
                        break;
                    }
                case 'hotkey':
                case 'press':
                    {
                        const keyStr = (null == actionInputs ? void 0 : actionInputs.key) || (null == actionInputs ? void 0 : actionInputs.hotkey);
                        if ('string' != typeof keyStr) throw new Error('key string is required when press or hotkey');
                        const lowerKeyStr = keyStr.toLowerCase();
                        const keys = lowerKeyStr.split(/\s+/).filter((k)=>k.length > 0);
                        const mappedKeys = keys.map((key)=>keyNameMap[key] || key);
                        if (0 === mappedKeys.length) throw new Error('key string is required when press or hotkey');
                        this.logger.info('Press/hotkey action mappedKeys:', mappedKeys.join('+'));
                        if (mappedKeys.length > 1) await this.aioComputer.hotkey(mappedKeys);
                        else await this.aioComputer.press(mappedKeys[0]);
                        break;
                    }
                case 'scroll':
                    await this.handleScroll(actionInputs);
                    break;
                case 'wait':
                    this.logger.info('Waiting for 3 seconds');
                    let sleepTime = 3000;
                    if (null == actionInputs ? void 0 : actionInputs.time) sleepTime = 1000 * actionInputs.time;
                    await new Promise((resolve)=>setTimeout(resolve, sleepTime));
                    break;
                default:
                    this.logger.warn(`Unsupported action type: ${actionType}`);
                    throw new Error(`Unsupported action type: ${actionType}`);
            }
            this.logger.info(`position percent return: (${startXPercent}, ${startYPercent})`);
            return {
                status: 'success',
                startXPercent,
                startYPercent,
                actionInputs
            };
        } catch (error) {
            this.logger.error('Execute action failed:', error);
            throw new Error(`Execute action failed: ${actionType}, message: ${error.message}`);
        }
    }
    async getCurrentUrl() {
        try {
            var _this_aioBrowser;
            const retUrl = await (null == (_this_aioBrowser = this.aioBrowser) ? void 0 : _this_aioBrowser.getActiveUrl());
            if (retUrl) return retUrl;
        } catch (error) {
            this.logger.warn('Failed to get page url:', error);
        }
    }
    async calculateRealCoords(coords) {
        if (!coords.normalized) {
            if (!coords.raw) throw new Error('Invalide coordinates');
            return {
                realX: coords.raw.x,
                realY: coords.raw.y
            };
        }
        const screenContext = await this.getScreenContext();
        return {
            realX: coords.normalized.x * screenContext.screenWidth * screenContext.scaleX,
            realY: coords.normalized.y * screenContext.screenHeight * screenContext.scaleY
        };
    }
    async handleClick(actionInputs, button, clickNum = 1) {
        const { point } = actionInputs;
        if (!point) throw new Error('point is required when click');
        const { realX, realY } = await this.calculateRealCoords(point);
        if (1 === clickNum) {
            if ('left' === button) await this.aioComputer.click(realX, realY);
            else if ('right' === button) await this.aioComputer.rightClick(realX, realY);
            else if ('middle' === button) await this.aioComputer.click(realX, realY, 'middle');
        } else await this.aioComputer.doubleClick(realX, realY);
    }
    async handleDrag(actionInputs) {
        const { start: startPoint, end: endPoint } = actionInputs;
        if (!startPoint) throw new Error('start point is required when drag/select');
        if (!endPoint) throw new Error('end point is required when drag/select');
        const { realX: startX, realY: startY } = await this.calculateRealCoords(startPoint);
        const { realX: endX, realY: endY } = await this.calculateRealCoords(endPoint);
        await this.aioComputer.moveTo(startX, startY);
        await this.aioComputer.mouseDown();
        await this.aioComputer.dragTo(endX, endY);
        await this.aioComputer.mouseUp();
    }
    async handleScroll(actionInputs) {
        const { direction, point } = actionInputs;
        if (point) {
            const { realX, realY } = await this.calculateRealCoords(point);
            await this.aioComputer.moveTo(realX, realY);
        }
        if ('string' != typeof direction) throw new Error('direction is required when scroll');
        const normalizedDirection = direction.toLowerCase();
        let dx = 0;
        let dy = 0;
        switch(normalizedDirection){
            case 'up':
                dy = 10;
                break;
            case 'down':
                dy = -10;
                break;
            case 'left':
                dx = 10;
                break;
            case 'right':
                dx = -10;
                break;
        }
        if (0 !== dx || 0 !== dy) await this.aioComputer.scroll(dx, dy);
    }
    constructor(options, logger = defaultLogger){
        super(), _define_property(this, "options", void 0), _define_property(this, "logger", void 0), _define_property(this, "aioBrowser", null), _define_property(this, "aioComputer", void 0), _define_property(this, "screenshotWidth", 1280), _define_property(this, "screenshotHeight", 1024);
        this.options = options;
        this.logger = logger.spawn('[AIOHybridOperator]');
        this.aioComputer = new AIOComputer(options);
    }
}
export { AIOHybridOperator };
