// @omni-tars/core/dist/plugins/snapshot.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { AgentPlugin } from "../AgentPlugin.mjs";
import { existsSync, mkdirSync, writeFileSync } from "fs";
import { join } from "path";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class SnapshotPlugin extends AgentPlugin {
    async onLLMRequest(id, payload) {
        this.saveSnapshot(id, 'request.json', payload);
    }
    async onLLMResponse(id, payload) {
        this.saveSnapshot(id, 'response.json', payload);
    }
    async onEachAgentLoopStart() {
        this.loop++;
    }
    async onAgentLoopEnd() {
        this.loop++;
        return Promise.resolve();
    }
    saveSnapshot(id, filename, payload) {
        try {
            const dir = join(this.baseDir, `${id}/loop-${this.loop}`);
            this.ensureDirectoryExists(dir);
            const filePath = join(dir, filename);
            const content = JSON.stringify(payload, null, 2);
            writeFileSync(filePath, content, {
                encoding: 'utf-8'
            });
        } catch (error) {
            console.error('saveSnapshot err: ', error);
        }
    }
    ensureDirectoryExists(dir) {
        if (!existsSync(dir)) mkdirSync(dir, {
            recursive: true
        });
    }
    constructor(option){
        super(), _define_property(this, "baseDir", ''), _define_property(this, "loop", 0), _define_property(this, "name", 'snapshot-plugin');
        this.baseDir = option.baseDir;
    }
}
export { SnapshotPlugin };
