// @tarko/model-provider/dist/llm-client.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { TokenJS } from "@tarko/llm-client";
const NATIVE_PROVIDERS = new Set([
    'openrouter',
    'openai-compatible',
    'azure-openai'
]);
function createLLMClient(agentModel, requestInterceptor) {
    const { provider, id, baseProvider, baseURL, apiKey, headers, params } = agentModel;
    const client = new TokenJS({
        apiKey,
        baseURL,
        defaultHeaders: headers
    });
    if (baseProvider && !NATIVE_PROVIDERS.has(baseProvider)) {
        const extendableClient = client;
        if ('function' == typeof extendableClient.extendModelList) extendableClient.extendModelList(baseProvider, id, {
            streaming: true,
            json: true,
            toolCalls: true,
            images: true
        });
    }
    return {
        chat: {
            completions: {
                async create (requestParams) {
                    const requestPayload = {
                        ...requestParams,
                        provider,
                        model: id,
                        ...params
                    };
                    const finalRequest = requestInterceptor ? requestInterceptor(provider, requestPayload, baseURL) : requestPayload;
                    return client.chat.completions.create({
                        ...finalRequest,
                        provider: baseProvider || 'openai'
                    });
                }
            }
        }
    };
}
export { createLLMClient };
