// @gui-agent/shared/dist/base/operator.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class BaseOperator {
}
class Operator extends BaseOperator {
    async doInitialize() {
        if (this._initialized) return;
        if (this._initializing && this._initPromise) return this._initPromise;
        this._initializing = true;
        this._initPromise = (async ()=>{
            try {
                await this.initialize();
                this._initialized = true;
            } finally{
                this._initializing = false;
            }
        })();
        return this._initPromise;
    }
    async ensureInitialized() {
        if (this._initialized || this._initializing) {
            if (this._initializing && this._initPromise) await this._initPromise;
        } else await this.doInitialize();
    }
    getSupportedActions() {
        return this.supportedActions();
    }
    async getScreenContext() {
        await this.ensureInitialized();
        return this.screenContext();
    }
    async doScreenshot() {
        try {
            await this.ensureInitialized();
            return await this.screenshot();
        } catch (error) {
            console.error('Error in doScreenshot:', error);
            return {
                base64: '',
                status: 'failed',
                errorMessage: error.message
            };
        }
    }
    async doExecute(params) {
        try {
            await this.ensureInitialized();
            return await this.execute(params);
        } catch (error) {
            console.error('Error in doExecute:', error);
            return {
                status: 'failed',
                errorMessage: error.message
            };
        }
    }
    constructor(){
        super(), _define_property(this, "_initialized", false), _define_property(this, "_initializing", false), _define_property(this, "_initPromise", null);
    }
}
export { BaseOperator, Operator };
