// @omni-tars/code-agent/dist/CodeToolCallEngine.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { SYSTEM_PROMPT_GROUP, createT5InitState, processT5StreamingChunk } from "@omni-tars/core";
import { ToolCallEngine, getLogger } from "@tarko/agent";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class CodeToolCallEngine extends ToolCallEngine {
    preparePrompt(instructions, tools) {
        return SYSTEM_PROMPT_GROUP;
    }
    prepareRequest(context) {
        return {
            model: context.model,
            messages: context.messages,
            temperature: context.temperature || 1.0,
            top_p: context.top_p,
            stream: true
        };
    }
    initStreamProcessingState() {
        return createT5InitState();
    }
    processStreamingChunk(chunk, state) {
        return processT5StreamingChunk(chunk, state);
    }
    finalizeStreamProcessing(state) {
        this.logger.info('finalizeStreamProcessing state \n', state);
        return {
            content: state.accumulatedChatContentBuffer || '',
            rawContent: state.contentBuffer,
            reasoningContent: state.reasoningBuffer ?? '',
            toolCalls: state.toolCalls,
            finishReason: (state.toolCalls || []).length > 0 ? 'tool_calls' : 'stop'
        };
    }
    buildHistoricalAssistantMessage(currentLoopAssistantEvent) {
        return {
            role: 'assistant',
            content: currentLoopAssistantEvent.rawContent || currentLoopAssistantEvent.content
        };
    }
    buildHistoricalToolCallResultMessages(toolCallResults) {
        return toolCallResults.map((result)=>{
            const textContent = result.content.filter((part)=>'text' === part.type).map((part)=>part.text).join('');
            return {
                role: 'user',
                content: `Tool "${result.toolName}" result:\n${textContent}`
            };
        });
    }
    constructor(...args){
        super(...args), _define_property(this, "logger", getLogger('CodeToolCallEngine'));
    }
}
export { CodeToolCallEngine };
