// @omni-tars/code-agent/dist/CodeAgentPlugin.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { AgentPlugin, CODE_ENVIRONMENT } from "@omni-tars/core";
import { LogLevel } from "@tarko/agent";
import { ExcuteBashProvider } from "./tools/ExcuteBash.mjs";
import { JupyterCIProvider } from "./tools/JupyterCI.mjs";
import { StrReplaceEditorProvider } from "./tools/StrReplaceEditor.mjs";
import { AioClient } from "@agent-infra/sandbox";
import assert from "assert";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class CodeAgentPlugin extends AgentPlugin {
    async checkSandbox(ignoreSandboxCheck) {
        if (ignoreSandboxCheck) return;
        await this.client.healthCheck();
    }
    async onLLMRequest(id, payload) {}
    async onLLMResponse(id, payload) {}
    async onEachAgentLoopStart() {}
    async onAgentLoopEnd() {}
    constructor(option){
        super(), _define_property(this, "name", 'code-agent-plugin'), _define_property(this, "environmentSection", CODE_ENVIRONMENT), _define_property(this, "client", void 0);
        assert(option.sandboxUrl, 'no AIO_SANDBOX_URL url providered.');
        this.client = new AioClient({
            baseUrl: option.sandboxUrl,
            retries: 0,
            logLevel: process.env.AIO_DEBUG ? LogLevel.DEBUG : LogLevel.INFO
        });
        this.tools = [
            new ExcuteBashProvider(this.client).getTool(),
            new JupyterCIProvider(this.client).getTool(),
            new StrReplaceEditorProvider(this.client).getTool()
        ];
        this.checkSandbox(option.ignoreSandboxCheck);
    }
}
export { CodeAgentPlugin };
