// @omni-tars/code-agent/dist/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ComposableAgent } from "@omni-tars/core";
import { CodeAgentPlugin } from "./CodeAgentPlugin.mjs";
import { CodeToolCallEngine } from "./CodeToolCallEngine.mjs";
import { CodeToolCallEngineProvider } from "./CodeToolCallEngineProvider.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const codePluginBuilder = (option)=>new CodeAgentPlugin(option);
class CodeAgent extends ComposableAgent {
    constructor(options){
        const { ignoreSandboxCheck, ...restOptions } = options;
        super({
            ...restOptions,
            plugins: [
                codePluginBuilder({
                    sandboxUrl: options.sandboxUrl,
                    ignoreSandboxCheck
                })
            ],
            toolCallEngine: CodeToolCallEngine
        });
    }
}
_define_property(CodeAgent, "label", void 0);
export { CodeToolCallEngineProvider, codePluginBuilder, CodeAgent as default };
