// @agent-infra/sandbox/dist/modules/browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserModule = void 0;
class BrowserModule {
    constructor(deps) {
        this.deps = deps;
    }
    /**
     * Get CDP version information
     */
    async cdpVersion(option) {
        return this.deps.request("/cdp/json/version", {
            method: "GET",
            ...option,
        });
    }
    /**
     * Get browser information
     */
    async info(option) {
        return this.deps.request("/v1/browser/info", {
            method: "GET",
            ...option,
        });
    }
    /**
     * Take a screenshot of the browser
     */
    async screenshot(option) {
        return this.deps.request("/v1/browser/screenshot", {
            method: "GET",
            raw: true,
            ...option,
        });
    }
    /**
     * Get browser information
     */
    async actions(params, option) {
        return this.deps.request("/v1/browser/actions", {
            method: "POST",
            body: JSON.stringify(params),
            ...option,
        });
    }
}
exports.BrowserModule = BrowserModule;
