// @agent-infra/sandbox/dist/modules/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileModule = void 0;
class FileModule {
    constructor(deps) {
        this.deps = deps;
    }
    /**
     * File editor operations
     */
    async editor(params, option) {
        return this.deps.request("/v1/file/str_replace_editor", {
            method: "POST",
            body: JSON.stringify(params),
            ...option,
        });
    }
    /**
     * Download file
     */
    async download(params, option) {
        return this.deps.request("/v1/file/download", {
            method: "POST",
            body: JSON.stringify(params),
            ...option,
        });
    }
    /**
     * List files in directory
     */
    async list({ path, sort_by = "name", file_types = ["string"], recursive = false, show_hidden = false, sort_desc = false, include_permissions = false, include_size = true, }, option) {
        return this.deps.request("/v1/file/list", {
            method: "POST",
            body: JSON.stringify({
                path,
                sort_by,
                sort_desc,
                file_types,
                recursive,
                show_hidden,
                include_permissions,
                include_size,
            }),
            ...option,
        });
    }
    /**
     * Execute Jupyter code
     */
    async jupyterExecute(params, option) {
        const requestParams = {
            ...params,
            kernel_name: params.kernel_name || "python3",
            timeout: params.timeout || 30,
        };
        return this.deps.request("/v1/jupyter/execute", {
            method: "POST",
            body: JSON.stringify(requestParams),
            ...option,
        });
    }
}
exports.FileModule = FileModule;
