// @ui-tars/shared/dist/types/agent.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
var agent_ErrorStatusEnum = /*#__PURE__*/ function(ErrorStatusEnum) {
    ErrorStatusEnum[ErrorStatusEnum["SCREENSHOT_RETRY_ERROR"] = -100000] = "SCREENSHOT_RETRY_ERROR";
    ErrorStatusEnum[ErrorStatusEnum["INVOKE_RETRY_ERROR"] = -100001] = "INVOKE_RETRY_ERROR";
    ErrorStatusEnum[ErrorStatusEnum["EXECUTE_RETRY_ERROR"] = -100002] = "EXECUTE_RETRY_ERROR";
    ErrorStatusEnum[ErrorStatusEnum["MODEL_SERVICE_ERROR"] = -100003] = "MODEL_SERVICE_ERROR";
    ErrorStatusEnum[ErrorStatusEnum["REACH_MAXLOOP_ERROR"] = -100004] = "REACH_MAXLOOP_ERROR";
    ErrorStatusEnum[ErrorStatusEnum["ENVIRONMENT_ERROR"] = -100005] = "ENVIRONMENT_ERROR";
    ErrorStatusEnum[ErrorStatusEnum["UNKNOWN_ERROR"] = -100099] = "UNKNOWN_ERROR";
    return ErrorStatusEnum;
}({});
class GUIAgentError extends Error {
    constructor(status, message, stack){
        super(message), _define_property(this, "status", void 0), _define_property(this, "message", void 0), _define_property(this, "stack", void 0);
        this.status = status;
        this.message = message;
        this.stack = stack;
    }
}
var agent_StatusEnum = /*#__PURE__*/ function(StatusEnum) {
    StatusEnum["INIT"] = "init";
    StatusEnum["RUNNING"] = "running";
    StatusEnum["PAUSE"] = "pause";
    StatusEnum["END"] = "end";
    StatusEnum["CALL_USER"] = "call_user";
    StatusEnum["MAX_LOOP"] = "max_loop";
    StatusEnum["USER_STOPPED"] = "user_stopped";
    StatusEnum["ERROR"] = "error";
    return StatusEnum;
}({});
export { agent_ErrorStatusEnum as ErrorStatusEnum, GUIAgentError, agent_StatusEnum as StatusEnum };
