// @omni-tars/core/dist/utils/streamingParserT5/extractToolCallT5.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
function extractToolCallT5(content, state) {
    if (!state.toolCallBuffer) state.toolCallBuffer = '';
    state.toolCallBuffer += content;
    let hasToolCallUpdate = false;
    const streamingToolCallUpdates = [];
    while(state.toolCallBuffer.length > 0)if (state.insideToolCall) {
        const closeMatch = state.toolCallBuffer.match(/<\/seed:tool_call>/);
        if (state.insideFunction) if (state.insideParameter) {
            const parameterCloseMatch = state.toolCallBuffer.match(/<\/parameter>/);
            const functionCloseMatchInside = state.toolCallBuffer.match(/<\/function>/);
            if (functionCloseMatchInside && parameterCloseMatch && functionCloseMatchInside.index < parameterCloseMatch.index) {
                state.insideParameter = false;
                continue;
            }
            if (parameterCloseMatch) {
                const paramValue = state.toolCallBuffer.substring(0, parameterCloseMatch.index);
                if (state.currentParameterName) state.currentParameters[state.currentParameterName] = (state.currentParameters[state.currentParameterName] || '') + paramValue;
                state.toolCallBuffer = state.toolCallBuffer.substring(parameterCloseMatch.index + parameterCloseMatch[0].length);
                state.insideParameter = false;
                const toolCall = state.toolCalls.find((tc)=>tc.id === state.currentToolCallId);
                if (toolCall) {
                    const escapedValue = paramValue.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t');
                    toolCall.function.arguments += escapedValue + '"';
                }
                hasToolCallUpdate = true;
                streamingToolCallUpdates.push({
                    toolCallId: state.currentToolCallId || '',
                    toolName: state.currentToolName || '',
                    argumentsDelta: paramValue.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t') + '"',
                    isComplete: false
                });
                continue;
            }
            if (hasPartialTagAtEnd(state.toolCallBuffer, [
                '</parameter>'
            ])) {
                const lastBracketIndex = state.toolCallBuffer.lastIndexOf('<');
                const paramValue = state.toolCallBuffer.substring(0, lastBracketIndex);
                if (paramValue && state.currentParameterName) {
                    state.currentParameters[state.currentParameterName] = (state.currentParameters[state.currentParameterName] || '') + paramValue;
                    const toolCall = state.toolCalls.find((tc)=>tc.id === state.currentToolCallId);
                    if (toolCall) {
                        const escapedValue = paramValue.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t');
                        toolCall.function.arguments += escapedValue;
                    }
                    hasToolCallUpdate = true;
                    streamingToolCallUpdates.push({
                        toolCallId: state.currentToolCallId || '',
                        toolName: state.currentToolName || '',
                        argumentsDelta: paramValue.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t'),
                        isComplete: false
                    });
                }
                state.toolCallBuffer = state.toolCallBuffer.substring(lastBracketIndex);
            } else {
                if (state.currentParameterName && state.toolCallBuffer) {
                    state.currentParameters[state.currentParameterName] = (state.currentParameters[state.currentParameterName] || '') + state.toolCallBuffer;
                    const toolCall = state.toolCalls.find((tc)=>tc.id === state.currentToolCallId);
                    if (toolCall) {
                        const escapedValue = state.toolCallBuffer.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t');
                        toolCall.function.arguments += escapedValue;
                    }
                    hasToolCallUpdate = true;
                    streamingToolCallUpdates.push({
                        toolCallId: state.currentToolCallId || '',
                        toolName: state.currentToolName || '',
                        argumentsDelta: state.toolCallBuffer.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t'),
                        isComplete: false
                    });
                }
                state.toolCallBuffer = '';
            }
            break;
        } else {
            const parameterMatch = state.toolCallBuffer.match(/<parameter=([^>]+)>/);
            const functionCloseMatch = state.toolCallBuffer.match(/<\/function>/);
            if (parameterMatch && (!functionCloseMatch || parameterMatch.index < functionCloseMatch.index)) {
                const paramName = parameterMatch[1];
                state.currentParameterName = paramName;
                state.insideParameter = true;
                state.parameterBuffer = '';
                state.toolCallBuffer = state.toolCallBuffer.substring(parameterMatch.index + parameterMatch[0].length);
                const isFirstParam = !state.currentParameters || 0 === Object.keys(state.currentParameters).length;
                const argsDelta = isFirstParam ? `{"${paramName}": "` : `, "${paramName}": "`;
                const toolCall = state.toolCalls.find((tc)=>tc.id === state.currentToolCallId);
                if (toolCall) toolCall.function.arguments += argsDelta;
                hasToolCallUpdate = true;
                streamingToolCallUpdates.push({
                    toolCallId: state.currentToolCallId || '',
                    toolName: state.currentToolName || '',
                    argumentsDelta: argsDelta,
                    isComplete: false
                });
                continue;
            }
            if (functionCloseMatch) {
                state.toolCallBuffer = state.toolCallBuffer.substring(functionCloseMatch.index + functionCloseMatch[0].length);
                const toolCall = state.toolCalls.find((tc)=>tc.id === state.currentToolCallId);
                if (toolCall && !toolCall.function.arguments.endsWith('}') && toolCall.function.arguments.length > 0) {
                    toolCall.function.arguments += '}';
                    hasToolCallUpdate = true;
                    streamingToolCallUpdates.push({
                        toolCallId: state.currentToolCallId || '',
                        toolName: state.currentToolName || '',
                        argumentsDelta: '}',
                        isComplete: true
                    });
                }
                state.insideFunction = false;
                state.currentToolName = '';
                state.currentParameterName = '';
                state.currentParameters = {};
                state.currentToolCallId = '';
                continue;
            }
            if (closeMatch) {
                state.toolCallBuffer = state.toolCallBuffer.substring(closeMatch.index + closeMatch[0].length);
                state.insideToolCall = false;
                state.insideFunction = false;
                continue;
            } else {
                if (hasPartialTagAtEnd(state.toolCallBuffer, [
                    '<parameter=',
                    '</function>',
                    '</seed:tool_call>'
                ])) {
                    const lastBracketIndex = state.toolCallBuffer.lastIndexOf('<');
                    state.toolCallBuffer = state.toolCallBuffer.substring(lastBracketIndex);
                } else state.toolCallBuffer = '';
                break;
            }
        }
        {
            const functionMatch = state.toolCallBuffer.match(/<function=([^>]*)>/);
            if (functionMatch && functionMatch[1]) {
                const toolName = functionMatch[1];
                state.currentToolName = toolName;
                state.currentParameters = {};
                state.insideFunction = true;
                state.functionCount = (state.functionCount || 0) + 1;
                state.toolCallBuffer = state.toolCallBuffer.substring(functionMatch.index + functionMatch[0].length);
                state.currentToolCallId = generateToolCallId();
                state.toolCalls.push({
                    id: state.currentToolCallId,
                    type: 'function',
                    function: {
                        name: toolName,
                        arguments: ''
                    }
                });
                hasToolCallUpdate = true;
                streamingToolCallUpdates.push({
                    toolCallId: state.currentToolCallId,
                    toolName: toolName,
                    argumentsDelta: '',
                    isComplete: false
                });
                continue;
            }
            if (closeMatch) {
                state.toolCallBuffer = state.toolCallBuffer.substring(closeMatch.index + closeMatch[0].length);
                state.insideToolCall = false;
                continue;
            }
            if (hasPartialTagAtEnd(state.toolCallBuffer, [
                '<function=',
                '</seed:tool_call>'
            ])) {
                const lastBracketIndex = state.toolCallBuffer.lastIndexOf('<');
                state.toolCallBuffer = state.toolCallBuffer.substring(lastBracketIndex);
            } else state.toolCallBuffer = '';
            break;
        }
    } else {
        const openMatch = state.toolCallBuffer.match(/<seed:tool_call>/);
        if (openMatch) {
            state.toolCallBuffer = state.toolCallBuffer.substring(openMatch.index + openMatch[0].length);
            state.insideToolCall = true;
            state.functionCount = 0;
            continue;
        }
        if (hasPartialTagAtEnd(state.toolCallBuffer, [
            '<seed:tool_call>'
        ])) {
            const lastBracketIndex = state.toolCallBuffer.lastIndexOf('<');
            state.toolCallBuffer = state.toolCallBuffer.substring(lastBracketIndex);
        } else state.toolCallBuffer = '';
        break;
    }
    return {
        hasToolCallUpdate,
        streamingToolCallUpdates
    };
}
function hasPartialTagAtEnd(buffer, expectedTags) {
    const lastBracketIndex = buffer.lastIndexOf('<');
    if (-1 === lastBracketIndex) return false;
    const afterBracket = buffer.substring(lastBracketIndex);
    if (afterBracket.includes('>')) return false;
    return expectedTags.some((tag)=>tag.startsWith(afterBracket) || afterBracket.startsWith(tag.substring(0, afterBracket.length)));
}
let testCallCounter = 0;
function generateToolCallId() {
    if ('test' === process.env.NODE_ENV) return `random_id_${++testCallCounter}`;
    return `call_${Date.now()}_${Math.random().toString(36).substring(2, 11)}`;
}
export { extractToolCallT5 };
