// @omni-tars/core/dist/utils/streamingParserT5/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { extractThinkT5 } from "./extractThinkT5.mjs";
import { extractToolCallT5 } from "./extractToolCallT5.mjs";
import { extractChatContentT5 } from "./extractChatContentT5.mjs";
function createT5InitState() {
    return {
        contentBuffer: '',
        toolCalls: [],
        reasoningBuffer: '',
        finishReason: null,
        thinkBuffer: '',
        accumulatedChatContentBuffer: '',
        toolCallBuffer: '',
        currentTag: null,
        insideThink: false,
        thinkParseCompleted: false,
        insideChatContent: false,
        insideToolCall: false,
        currentToolName: '',
        currentParameters: {},
        parameterBuffer: '',
        currentParameterName: '',
        insideParameter: false,
        insideFunction: false,
        functionBuffer: '',
        currentToolCallId: '',
        functionCount: 0
    };
}
function processT5StreamingChunk(chunk, state) {
    var _chunk_choices_, _chunk_choices_1;
    const delta = null == (_chunk_choices_ = chunk.choices[0]) ? void 0 : _chunk_choices_.delta;
    let content = '';
    let reasoningContent = '';
    let hasToolCallUpdate = false;
    let streamingToolCallUpdates = [];
    if (null == (_chunk_choices_1 = chunk.choices[0]) ? void 0 : _chunk_choices_1.finish_reason) state.finishReason = chunk.choices[0].finish_reason;
    if (null == delta ? void 0 : delta.reasoning_content) {
        reasoningContent = delta.reasoning_content;
        state.reasoningBuffer += reasoningContent;
    }
    if (null == delta ? void 0 : delta.content) {
        state.contentBuffer += delta.content;
        reasoningContent = extractThinkT5(delta.content, state);
        const toolParsed = extractToolCallT5(delta.content, state);
        hasToolCallUpdate = toolParsed.hasToolCallUpdate;
        streamingToolCallUpdates = toolParsed.streamingToolCallUpdates || [];
        content = extractChatContentT5(delta.content, state);
    }
    return {
        content,
        reasoningContent,
        hasToolCallUpdate,
        toolCalls: state.toolCalls,
        streamingToolCallUpdates
    };
}
export { createT5InitState, processT5StreamingChunk };
