// @agent-infra/media-utils/dist/base64/image-parse.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
const IMAGE_TYPE_MAP = new Map([
    [
        '/9j/',
        'jpeg'
    ],
    [
        'iVBORw',
        'png'
    ],
    [
        'UklGR',
        'webp'
    ],
    [
        'R0lGOD',
        'gif'
    ],
    [
        'Qk',
        'bmp'
    ]
]);
function base64String2Uint8Array(base64) {
    if ('undefined' != typeof Buffer) return new Uint8Array(Buffer.from(base64, 'base64'));
    {
        const binaryString = atob(base64);
        const bytes = new Uint8Array(binaryString.length);
        for(let i = 0; i < binaryString.length; i++)bytes[i] = binaryString.charCodeAt(i);
        return bytes;
    }
}
function parsePngDimensions(bytes) {
    const width = bytes[16] << 24 | bytes[17] << 16 | bytes[18] << 8 | bytes[19];
    const height = bytes[20] << 24 | bytes[21] << 16 | bytes[22] << 8 | bytes[23];
    return {
        width,
        height
    };
}
function parseJpegDimensions(bytes) {
    let i = 2;
    while(i < bytes.length - 1)if (0xff === bytes[i]) {
        const marker = bytes[i + 1];
        if (marker >= 0xc0 && marker <= 0xc3) {
            if (i + 8 < bytes.length) {
                const height = bytes[i + 5] << 8 | bytes[i + 6];
                const width = bytes[i + 7] << 8 | bytes[i + 8];
                return {
                    width,
                    height
                };
            }
        }
        if (i + 3 < bytes.length) {
            const segmentLength = bytes[i + 2] << 8 | bytes[i + 3];
            i += 2 + segmentLength;
        } else break;
    } else i++;
    throw new Error('Unable to find JPEG dimensions');
}
function parseGifDimensions(bytes) {
    const width = bytes[6] | bytes[7] << 8;
    const height = bytes[8] | bytes[9] << 8;
    return {
        width,
        height
    };
}
function parseWebpDimensions(bytes) {
    const fourCC = String.fromCharCode(bytes[12], bytes[13], bytes[14], bytes[15]);
    if ('VP8 ' === fourCC) {
        const startByte = 23;
        const width = bytes[startByte + 3] | bytes[startByte + 4] << 8;
        const height = bytes[startByte + 5] | bytes[startByte + 6] << 8;
        return {
            width: 0x3fff & width,
            height: 0x3fff & height
        };
    }
    if ('VP8L' === fourCC) {
        const startByte = 21;
        const bits = bytes[startByte] | bytes[startByte + 1] << 8 | bytes[startByte + 2] << 16 | bytes[startByte + 3] << 24;
        const width = (0x3fff & bits) + 1;
        const height = (bits >> 14 & 0x3fff) + 1;
        return {
            width,
            height
        };
    }
    if ('VP8X' === fourCC) {
        const startByte = 24;
        const width = bytes[startByte] | bytes[startByte + 1] << 8 | bytes[startByte + 2] << 16;
        const height = bytes[startByte + 3] | bytes[startByte + 4] << 8 | bytes[startByte + 5] << 16;
        return {
            width: width + 1,
            height: height + 1
        };
    }
    throw new Error('Unsupported WebP format');
}
function parseBmpDimensions(bytes) {
    const width = bytes[18] | bytes[19] << 8 | bytes[20] << 16 | bytes[21] << 24;
    const height = bytes[22] | bytes[23] << 8 | bytes[24] << 16 | bytes[25] << 24;
    return {
        width,
        height
    };
}
export { IMAGE_TYPE_MAP, base64String2Uint8Array, parseBmpDimensions, parseGifDimensions, parseJpegDimensions, parsePngDimensions, parseWebpDimensions };
