// @agent-infra/media-utils/dist/base64/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { IMAGE_TYPE_MAP, base64String2Uint8Array, parseBmpDimensions, parseGifDimensions, parseJpegDimensions, parsePngDimensions, parseWebpDimensions } from "./image-parse.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class Base64ImageParser {
    getPureBase64Image() {
        return this.pureBase64;
    }
    getBuffer() {
        this.buffer = base64String2Uint8Array(this.pureBase64);
        return this.buffer;
    }
    getImageType() {
        if (this.imageType) return this.imageType;
        const prefix = this.pureBase64.substring(0, 8);
        for (const [signature, type] of IMAGE_TYPE_MAP)if (prefix.startsWith(signature)) {
            this.imageType = type;
            break;
        }
        return this.imageType;
    }
    getDimensions() {
        if (this.dimensions) return this.dimensions;
        const imageType = this.getImageType();
        if (!imageType) return null;
        try {
            switch(imageType){
                case 'png':
                    {
                        const bytes = this.getHeaderBuffer(32);
                        this.dimensions = parsePngDimensions(bytes);
                        break;
                    }
                case 'jpeg':
                    try {
                        const headerBytes = this.getHeaderBuffer(1024);
                        this.dimensions = parseJpegDimensions(headerBytes);
                    } catch (e) {
                        const fullBuffer = this.getBuffer();
                        this.dimensions = parseJpegDimensions(fullBuffer);
                    }
                    break;
                case 'webp':
                    {
                        const bytes = this.getHeaderBuffer(32);
                        this.dimensions = parseWebpDimensions(bytes);
                        break;
                    }
                case 'gif':
                    {
                        const bytes = this.getHeaderBuffer(24);
                        this.dimensions = parseGifDimensions(bytes);
                        break;
                    }
                case 'bmp':
                    {
                        const bytes = this.getHeaderBuffer(40);
                        this.dimensions = parseBmpDimensions(bytes);
                        break;
                    }
                default:
                    return null;
            }
            return this.dimensions;
        } catch (error) {
            console.warn('Failed to parse image dimensions:', error);
            return null;
        }
    }
    getDataUri() {
        const imageType = this.getImageType();
        if (imageType) return `data:image/${this.imageType};base64,${this.pureBase64}`;
        return null;
    }
    getHeaderBuffer(maxBytes) {
        if (this.buffer) return this.buffer;
        const base64CharsNeeded = Math.ceil(4 * maxBytes / 3);
        const alignedChars = 4 * Math.ceil(base64CharsNeeded / 4);
        const headerBase64 = this.pureBase64.substring(0, alignedChars);
        return base64String2Uint8Array(headerBase64);
    }
    constructor(base64Image){
        _define_property(this, "pureBase64", void 0);
        _define_property(this, "buffer", void 0);
        _define_property(this, "imageType", null);
        _define_property(this, "dimensions", null);
        this.pureBase64 = base64Image.replace(/^data:image\/\w+;base64,/, '');
    }
}
export { Base64ImageParser };
