// @agent-infra/sandbox/dist/aio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AioClient = void 0;
const base_1 = require("./base");
const shell_1 = require("./modules/shell");
const file_1 = require("./modules/file");
const browser_1 = require("./modules/browser");
class AioClient extends base_1.BaseClient {
    constructor(config) {
        super(config);
        const deps = {
            request: this.request.bind(this),
            logger: this.logger,
            sleep: this.sleep.bind(this),
        };
        this.shellModule = new shell_1.ShellModule(deps);
        this.fileModule = new file_1.FileModule(deps);
        this.browserModule = new browser_1.BrowserModule(deps);
    }
    /**
     * Execute shell command
     */
    async shellExec(params, option) {
        return this.shellModule.exec(params, option);
    }
    /**
     * Execute shell command with async polling and timeout management
     * @param params Shell execution parameters
     * @param maxWaitTime Maximum wait time in milliseconds (default: 10 minutes)
     * @param pollInterval Polling interval in milliseconds (default: 2 seconds)
     */
    async shellExecWithPolling(params, maxWaitTime = 10 * 60 * 1000, // 10 minutes
    pollInterval = 1000 // 1s
    ) {
        return this.shellModule.execWithPolling(params, maxWaitTime, pollInterval);
    }
    /**
     * View shell session
     */
    async shellView(params, option) {
        return this.shellModule.view(params, option);
    }
    /**
     * Kill shell session
     */
    async shellKill(params, option) {
        return this.shellModule.kill(params, option);
    }
    /**
     * Execute Jupyter code
     */
    async jupyterExecute(params, option) {
        return this.fileModule.jupyterExecute(params, option);
    }
    /**
     * File editor operations
     */
    async fileEditor(params, option) {
        return this.fileModule.editor(params, option);
    }
    /**
     * Download file
     */
    async fileDownload(params, option) {
        return this.fileModule.download(params, option);
    }
    async fileList(params, option) {
        return this.fileModule.list(params, option);
    }
    async cdpVersion(option) {
        return this.browserModule.cdpVersion(option);
    }
    /**
     * Get browser information
     */
    async browserInfo(option) {
        return this.browserModule.info(option);
    }
    /**
     * Get browser information
     */
    async browserScreenshot(option) {
        return this.browserModule.screenshot(option);
    }
    /**
     * Get browser information
     */
    async browserActions(params, option) {
        return this.browserModule.actions(params, option);
    }
}
exports.AioClient = AioClient;
