// @gui-agent/shared/dist/utils/standardizeNames.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
const actionTypeMap = {
    snapshot: 'screenshot',
    screenshot: 'screenshot',
    take_screenshot: 'screenshot',
    takescreenshot: 'screenshot',
    click: 'click',
    left_click: 'click',
    left_single: 'click',
    leftclick: 'click',
    leftsingle: 'click',
    double_click: 'double_click',
    left_double: 'double_click',
    doubleclick: 'double_click',
    leftdouble: 'double_click',
    right_click: 'right_click',
    right_single: 'right_click',
    rightclick: 'right_click',
    rightsingle: 'right_click',
    middle_click: 'middle_click',
    middle_single: 'middle_click',
    middleclick: 'middle_click',
    middlesingle: 'middle_click',
    move: 'mouse_move',
    move_to: 'mouse_move',
    mouse_move: 'mouse_move',
    moveto: 'mouse_move',
    mousemove: 'mouse_move',
    hover: 'mouse_move',
    mouse_down: 'mouse_down',
    mousedown: 'mouse_down',
    mouse_up: 'mouse_up',
    mouseup: 'mouse_up',
    drag: 'drag',
    select: 'drag',
    left_click_drag: 'drag',
    leftclickdrag: 'drag',
    swipe: 'swipe',
    scroll: 'scroll',
    type: 'type',
    hotkey: 'hotkey',
    press: 'press',
    release: 'release',
    navigate: 'navigate',
    navigate_back: 'navigate_back',
    navigateback: 'navigate_back',
    long_press: 'long_press',
    longpress: 'long_press',
    home: 'press_home',
    press_home: 'press_home',
    presshome: 'press_home',
    back: 'press_back',
    press_back: 'press_back',
    pressback: 'press_back',
    open: 'open_app',
    open_app: 'open_app',
    openapp: 'open_app',
    wait: 'wait',
    finished: 'finished',
    call_user: 'call_user',
    calluser: 'call_user'
};
function standardizeActionType(name) {
    name = name.toLowerCase();
    return actionTypeMap[name] || name;
}
const actionInputNameMap = {
    start: 'start',
    start_box: 'start',
    startbox: 'start',
    start_point: 'start',
    start_position: 'start',
    start_coordinate: 'start',
    start_coordinates: 'start',
    end: 'end',
    end_box: 'end',
    endbox: 'end',
    end_point: 'end',
    end_position: 'end',
    end_coordinate: 'end',
    end_coordinates: 'end',
    point: 'point',
    position: 'point',
    coordinate: 'point',
    coordinates: 'point',
    button: 'button',
    mouse_button: 'button',
    mousebutton: 'button',
    direction: 'direction',
    dir: 'direction',
    scroll_direction: 'direction',
    scrolldirection: 'direction',
    content: 'content',
    text: 'content',
    input_text: 'content',
    inputtext: 'content',
    type: 'content',
    key: 'key',
    keyname: 'key',
    hotkey: 'key',
    keyboard_key: 'key',
    keyboardkey: 'key',
    url: 'url',
    link: 'url',
    website: 'url',
    name: 'name',
    appname: 'name',
    app_name: 'name',
    application: 'name',
    time: 'time',
    duration: 'time',
    wait_time: 'time',
    waittime: 'time',
    delay: 'time'
};
const actionTypeSpecificMappings = {
    navigate: {
        content: 'url',
        url: 'url'
    },
    open_app: {
        content: 'name',
        appname: 'name',
        app_name: 'name',
        name: 'name'
    }
};
function standardizeActionInputName(actionType, inputName) {
    actionType = actionType.toLowerCase();
    inputName = inputName.toLowerCase();
    actionType = standardizeActionType(actionType);
    const typeSpecificMap = actionTypeSpecificMappings[actionType];
    if (typeSpecificMap && typeSpecificMap[inputName]) return typeSpecificMap[inputName];
    return actionInputNameMap[inputName] || inputName;
}
export { standardizeActionInputName, standardizeActionType };
