// @tarko/mcp-agent/dist/mcp-tool-adapter.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Tool } from "@tarko/agent";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class MCPToolAdapter {
    createTools() {
        const mcpTools = this.mcpClient.getTools();
        return mcpTools.map((mcpTool)=>new Tool({
                id: mcpTool.name,
                description: `[${this.serverName}] ${mcpTool.description}`,
                parameters: mcpTool.inputSchema || {
                    type: 'object',
                    properties: {}
                },
                function: async (args)=>{
                    const result = await this.mcpClient.callTool(mcpTool.name, args);
                    return result.content;
                }
            }));
    }
    constructor(mcpClient, serverName){
        _define_property(this, "mcpClient", void 0);
        _define_property(this, "serverName", void 0);
        this.mcpClient = mcpClient;
        this.serverName = serverName;
    }
}
export { MCPToolAdapter };
