// @tarko/mcp-agent/dist/mcp-client-v2.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { MCPClient } from "@agent-infra/mcp-client";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class MCPClientV2 {
    async initialize() {
        if (this.isInitialized) return this.tools;
        try {
            this.logger.info(`Initializing MCP client v2 for ${this.serverName}`);
            await this.v2Client.init();
            this.tools = await this.v2Client.listTools(this.serverName);
            this.isInitialized = true;
            this.logger.success(`MCP client v2 initialized successfully for ${this.serverName}, found ${this.tools.length} tools`);
            return this.tools;
        } catch (error) {
            this.logger.error(`Error initializing v2 MCP client for ${this.serverName}:`, error);
            throw error;
        }
    }
    async callTool(toolName, args) {
        if (!this.isInitialized) {
            this.logger.info(`Client not initialized, initializing before tool call: ${toolName}`);
            await this.initialize();
        }
        try {
            this.logger.info(`Calling tool ${toolName} with client ${this.serverName}`);
            const result = await this.v2Client.callTool({
                client: this.serverName,
                name: toolName,
                args
            });
            return result.content;
        } catch (error) {
            this.logger.error(`Error calling MCP tool ${toolName}:`, error);
            throw new Error(`Failed to execute tool ${toolName}: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async close() {
        if (this.isInitialized) {
            this.logger.info(`Closing MCP client v2 for ${this.serverName}`);
            await this.v2Client.cleanup();
            this.isInitialized = false;
            this.tools = [];
            this.logger.success(`MCP client v2 closed successfully for ${this.serverName}`);
        }
    }
    getTools() {
        return [
            ...this.tools
        ];
    }
    constructor(serverName, config, logger, defaultTimeout = 60){
        _define_property(this, "logger", void 0);
        _define_property(this, "v2Client", void 0);
        _define_property(this, "serverName", void 0);
        _define_property(this, "tools", void 0);
        _define_property(this, "isInitialized", void 0);
        this.logger = logger;
        this.tools = [];
        this.isInitialized = false;
        this.serverName = serverName;
        this.v2Client = new MCPClient([
            {
                name: serverName,
                ...config,
                status: 'activate'
            }
        ], {
            isDebug: false,
            defaultTimeout
        });
    }
}
export { MCPClientV2 };
