// @tarko/mcp-agent/dist/mcp-agent.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Agent, Tool } from "@tarko/agent";
import { MCPClientV2 } from "./mcp-client-v2.mjs";
import { filterItems } from "@tarko/shared-utils";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class MCPAgent extends Agent {
    filterMCPServers(mcpServerConfig, filterOptions) {
        if (!filterOptions || !filterOptions.include && !filterOptions.exclude) return mcpServerConfig;
        const serverItems = Object.entries(mcpServerConfig).map(([name, config])=>({
                name,
                config
            }));
        const filteredServers = filterItems(serverItems, filterOptions, 'MCP servers');
        const filteredRegistry = {};
        filteredServers.forEach(({ name, config })=>{
            filteredRegistry[name] = config;
        });
        return filteredRegistry;
    }
    async initialize() {
        const filteredMcpServerConfig = this.filterMCPServers(this.mcpServerConfig, this.options.mcpServer);
        for (const [serverName, config] of Object.entries(filteredMcpServerConfig))try {
            this.logger.info(`\u{1F50C} Connecting to MCP server: ${serverName}`);
            const defaultTimeout = this.options.defaultConnectionTimeout ?? 60;
            const mcpClient = new MCPClientV2(serverName, config, this.logger, defaultTimeout);
            await mcpClient.initialize();
            this.mcpClients.set(serverName, mcpClient);
            const mcpTools = mcpClient.getTools();
            for (const mcpTool of mcpTools){
                const tool = new Tool({
                    id: mcpTool.name,
                    description: `[${serverName}] ${mcpTool.description}`,
                    parameters: mcpTool.inputSchema || {
                        type: 'object',
                        properties: {}
                    },
                    function: async (args)=>await mcpClient.callTool(mcpTool.name, args)
                });
                this.registerTool(tool);
            }
            const toolCount = mcpTools.length;
            this.logger.success(`\u{2705} Connected to MCP server ${serverName} with ${toolCount} tools`);
        } catch (error) {
            this.logger.error(`\u{274C} Failed to connect to MCP server ${serverName}: ${error instanceof Error ? error.message : JSON.stringify(error)}`);
            throw new Error(`\u{274C} Failed to connect to MCP server ${serverName}: ${error instanceof Error ? error.message : JSON.stringify(error)}`);
        }
        super.initialize();
    }
    async cleanup() {
        for (const [serverName, client] of Array.from(this.mcpClients.entries()))try {
            await client.close();
            this.logger.info(`\u{2705} Closed connection to MCP server: ${serverName}`);
        } catch (error) {
            this.logger.error(`\u{274C} Error closing MCP client ${serverName}: ${error}`);
        }
        this.mcpClients.clear();
    }
    constructor(options){
        super(options), _define_property(this, "mcpClients", new Map()), _define_property(this, "mcpServerConfig", void 0);
        this.mcpServerConfig = options.mcpServers ?? {};
    }
}
_define_property(MCPAgent, "label", '@tarko/mcp-agent');
export { MCPAgent };
