// @tarko/agent-snapshot/dist/agent-generate-snapshot-hook.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import path from "path";
import fs from "fs";
import { logger } from "./utils/logger.mjs";
import { AgentHookBase } from "./agent-hook-base.mjs";
class AgentGenerateSnapshotHook extends AgentHookBase {
    llmRequests = {};
    llmResponses = {};
    toolCallsByLoop = {};
    startTimeByToolCall = {};
    constructor(agent, options){
        super(agent, options);
    }
    onEachAgentLoopStart(id) {
        logger.info(`Starting agent loop ${this.agent.getCurrentLoopIteration()}`);
        const currentLoop = this.agent.getCurrentLoopIteration();
        if (!this.toolCallsByLoop[currentLoop]) this.toolCallsByLoop[currentLoop] = [];
        if (this.originalEachLoopStartHook) return this.originalEachLoopStartHook.call(this.agent, id);
    }
    onLLMRequest(id, payload) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        this.llmRequests[currentLoop] = payload;
        const loopDir = path.join(this.snapshotPath, `loop-${currentLoop}`);
        if (!fs.existsSync(loopDir)) fs.mkdirSync(loopDir, {
            recursive: true
        });
        fs.writeFileSync(path.join(loopDir, 'llm-request.jsonl'), JSON.stringify(payload, null, 2), 'utf-8');
        const events = this.agent.getEventStream().getEvents();
        fs.writeFileSync(path.join(loopDir, 'event-stream.jsonl'), JSON.stringify(events, null, 2), 'utf-8');
        if (this.originalRequestHook) return this.originalRequestHook.call(this.agent, id, payload);
    }
    onLLMResponse(id, payload) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        this.llmResponses[currentLoop] = payload;
        if (this.originalResponseHook) return this.originalResponseHook.call(this.agent, id, payload);
    }
    onLLMStreamingResponse(id, payload) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        const loopDir = `loop-${currentLoop}`;
        try {
            const responsePath = path.join(this.snapshotPath, loopDir, 'llm-response.jsonl');
            this.writeStreamingChunks(responsePath, payload.chunks);
            logger.info(`Saved ${payload.chunks.length} streaming chunks for ${loopDir}`);
        } catch (error) {
            logger.error(`Failed to save streaming chunks: ${error}`);
        }
        if (this.originalStreamingResponseHook) this.originalStreamingResponseHook.call(this.agent, id, payload);
    }
    onBeforeToolCall(id, toolCall, args) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        this.startTimeByToolCall[toolCall.toolCallId] = Date.now();
        if (!this.toolCallsByLoop[currentLoop]) this.toolCallsByLoop[currentLoop] = [];
        this.toolCallsByLoop[currentLoop].push({
            toolCallId: toolCall.toolCallId,
            name: toolCall.name,
            args
        });
        logger.debug(`Tool call captured for ${toolCall.name} (${toolCall.toolCallId}) in loop ${currentLoop}`);
        if (this.originalBeforeToolCallHook) return this.originalBeforeToolCallHook.call(this.agent, id, toolCall, args);
        return args;
    }
    onAfterToolCall(id, toolCall, result) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        const executionTime = Date.now() - (this.startTimeByToolCall[toolCall.toolCallId] || Date.now());
        if (this.toolCallsByLoop[currentLoop]) {
            const toolCallData = this.toolCallsByLoop[currentLoop].find((tc)=>tc.toolCallId === toolCall.toolCallId);
            if (toolCallData) {
                toolCallData.result = result;
                toolCallData.executionTime = executionTime;
            }
        }
        logger.debug(`Tool call result captured for ${toolCall.name} (${toolCall.toolCallId}) in loop ${currentLoop}`);
        this.saveToolCalls(currentLoop);
        if (this.originalAfterToolCallHook) return this.originalAfterToolCallHook.call(this.agent, id, toolCall, result);
        return result;
    }
    onToolCallError(id, toolCall, error) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        const executionTime = Date.now() - (this.startTimeByToolCall[toolCall.toolCallId] || Date.now());
        if (this.toolCallsByLoop[currentLoop]) {
            const toolCallData = this.toolCallsByLoop[currentLoop].find((tc)=>tc.toolCallId === toolCall.toolCallId);
            if (toolCallData) {
                toolCallData.error = error;
                toolCallData.executionTime = executionTime;
            }
        }
        logger.debug(`Tool call error captured for ${toolCall.name} (${toolCall.toolCallId}) in loop ${currentLoop}`);
        this.saveToolCalls(currentLoop);
        if (this.originalToolCallErrorHook) return this.originalToolCallErrorHook.call(this.agent, id, toolCall, error);
        return `Error: ${error}`;
    }
    saveToolCalls(loopNumber) {
        if (!this.toolCallsByLoop[loopNumber] || 0 === this.toolCallsByLoop[loopNumber].length) return;
        try {
            const loopDir = path.join(this.snapshotPath, `loop-${loopNumber}`);
            if (!fs.existsSync(loopDir)) fs.mkdirSync(loopDir, {
                recursive: true
            });
            fs.writeFileSync(path.join(loopDir, 'tool-calls.jsonl'), JSON.stringify(this.toolCallsByLoop[loopNumber], null, 2), 'utf-8');
            logger.info(`Saved ${this.toolCallsByLoop[loopNumber].length} tool calls for loop ${loopNumber}`);
        } catch (error) {
            logger.error(`Failed to save tool calls for loop ${loopNumber}: ${error}`);
        }
    }
    onAgentLoopEnd(id) {
        const finalEvents = this.agent.getEventStream().getEvents();
        fs.writeFileSync(path.join(this.snapshotPath, 'event-stream.jsonl'), JSON.stringify(finalEvents, null, 2), 'utf-8');
        logger.info(`Snapshot generation completed: ${this.snapshotPath}`);
        if (this.originalLoopEndHook) return this.originalLoopEndHook.call(this.agent, id);
    }
    onProcessToolCalls(id, toolCalls) {}
}
export { AgentGenerateSnapshotHook };
