// @tarko/agent-snapshot/dist/agent-snapshot.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import path from "path";
import fs from "fs";
import { isAgentRunObjectOptions, isStreamingOptions } from "@tarko/agent-interface";
import { SnapshotManager } from "./snapshot-manager.mjs";
import { AgentGenerateSnapshotHook } from "./agent-generate-snapshot-hook.mjs";
import { AgentReplaySnapshotHook } from "./agent-replay-snapshot-hook.mjs";
import { logger } from "./utils/logger.mjs";
class AgentSnapshot {
    hostedAgent;
    options;
    snapshotPath;
    snapshotName;
    snapshotManager;
    replayHook;
    generateHook = null;
    constructor(agent, options){
        this.hostedAgent = agent;
        this.options = options;
        this.snapshotPath = options.snapshotPath || path.join(process.cwd(), 'fixtures');
        this.snapshotName = options.snapshotName ?? path.basename(options.snapshotPath);
        this.snapshotManager = new SnapshotManager(this.snapshotPath, options.normalizerConfig);
        this.replayHook = new AgentReplaySnapshotHook(agent, {
            snapshotPath: this.options.snapshotPath || path.join(process.cwd(), 'fixtures'),
            snapshotName: this.snapshotName
        });
        if (!fs.existsSync(this.snapshotPath)) fs.mkdirSync(this.snapshotPath, {
            recursive: true
        });
        const agentSnapshotProto = Object.getPrototypeOf(this);
        const methodsToPreserve = {};
        Object.getOwnPropertyNames(agentSnapshotProto).forEach((key)=>{
            const descriptor = Object.getOwnPropertyDescriptor(agentSnapshotProto, key);
            if ('function' == typeof descriptor?.value && 'constructor' !== key) methodsToPreserve[key] = this[key].bind(this);
        });
        Object.setPrototypeOf(this, Object.getPrototypeOf(agent));
        Object.getOwnPropertyNames(agent).forEach((prop)=>{
            if (!(prop in this)) Object.defineProperty(this, prop, {
                get: ()=>agent[prop],
                set: (value)=>{
                    agent[prop] = value;
                },
                configurable: true
            });
        });
        Object.entries(methodsToPreserve).forEach(([key, method])=>{
            this[key] = method;
        });
    }
    async run(runOptions) {
        logger.info(`AgentSnapshot.run called with ${'string' == typeof runOptions ? 'string' : 'options object'}`);
        if (!this.generateHook) this.generateHook = new AgentGenerateSnapshotHook(this.hostedAgent, {
            snapshotPath: this.options.snapshotPath,
            snapshotName: this.snapshotName
        });
        this.generateHook.setCurrentRunOptions(runOptions);
        this.generateHook.hookAgent();
        try {
            const isStreaming = 'object' == typeof runOptions && isAgentRunObjectOptions(runOptions) && isStreamingOptions(runOptions);
            logger.info(`Executing agent with ${isStreaming ? 'streaming' : 'non-streaming'} mode`);
            const response = await this.hostedAgent.run(runOptions);
            return response;
        } catch (error) {
            logger.error(`Error during AgentSnapshot.run: ${error}`);
            throw error;
        } finally{
            if (this.generateHook) this.generateHook.clearError();
        }
    }
    async generate(runOptions) {
        const snapshotName = this.snapshotName || `agent-snapshot-${Date.now()}`;
        this.generateHook = new AgentGenerateSnapshotHook(this.hostedAgent, {
            snapshotPath: this.options.snapshotPath || path.join(process.cwd(), 'fixtures'),
            snapshotName: snapshotName
        });
        if (this.snapshotPath) {
            if (!fs.existsSync(this.snapshotPath)) fs.mkdirSync(this.snapshotPath, {
                recursive: true
            });
        }
        logger.info(`Starting snapshot generation for '${snapshotName}'`);
        const startTime = Date.now();
        this.generateHook.setCurrentRunOptions(runOptions);
        this.generateHook.hookAgent();
        try {
            const response = await this.hostedAgent.run(runOptions);
            if (this.generateHook.hasError()) {
                const error = this.generateHook.getLastError();
                logger.error(`Error occurred during snapshot generation: ${error?.message}`);
                throw error;
            }
            const events = this.hostedAgent.getEventStream().getEvents();
            const snapshotPath = path.join(this.options.snapshotPath);
            const loopCount = this.countLoops(snapshotPath);
            logger.success(`Successfully generated snapshot with ${loopCount} loops`);
            return {
                snapshotPath,
                loopCount,
                response,
                events,
                meta: {
                    snapshotName: this.snapshotName,
                    executionTime: Date.now() - startTime
                }
            };
        } catch (error) {
            logger.error(`Snapshot generation failed: ${error}`);
            throw error;
        } finally{
            if (this.generateHook) this.generateHook.clearError();
        }
    }
    async replay(runOptions, config) {
        const snapshotName = this.options.snapshotName || path.basename(this.options.snapshotPath);
        const updateSnapshots = config?.updateSnapshots || this.options.updateSnapshots || false;
        if (config?.normalizerConfig) this.snapshotManager.updateAgentNormalizerConfig(config.normalizerConfig);
        const verification = {
            verifyLLMRequests: config?.verification?.verifyLLMRequests !== void 0 ? config.verification.verifyLLMRequests : this.options.verification?.verifyLLMRequests !== false,
            verifyEventStreams: config?.verification?.verifyEventStreams !== void 0 ? config.verification.verifyEventStreams : this.options.verification?.verifyEventStreams !== false,
            verifyToolCalls: config?.verification?.verifyToolCalls !== void 0 ? config.verification.verifyToolCalls : this.options.verification?.verifyToolCalls !== false
        };
        if (!fs.existsSync(this.snapshotPath)) throw new Error(`Snapshot directory not found: ${this.snapshotPath}. Generate snapshots first using .generate()`);
        logger.info(`Running test against snapshot '${snapshotName}'${updateSnapshots ? ' (update mode)' : ''}`);
        logger.info(`Verification settings: 
      LLM requests: ${verification.verifyLLMRequests ? 'enabled' : 'disabled'}, 
      Event streams: ${verification.verifyEventStreams ? 'enabled' : 'disabled'},
      Tool calls: ${verification.verifyToolCalls ? 'enabled' : 'disabled'}`);
        const loopCount = this.countLoops(this.snapshotPath);
        logger.info(`Found ${loopCount} loops in test case`);
        const startTime = Date.now();
        try {
            await this.replayHook.setup(this.hostedAgent, this.snapshotPath, loopCount, {
                updateSnapshots,
                normalizerConfig: config?.normalizerConfig || this.options.normalizerConfig,
                verification
            });
            if (this.replayHook.hasError()) {
                const error = this.replayHook.getLastError();
                logger.error(`Error occurred during test setup: ${error?.message}`);
                throw error;
            }
            const mockLLMClient = this.replayHook.getMockLLMClient();
            this.hostedAgent.setCustomLLMClient(mockLLMClient);
            const isStreaming = 'object' == typeof runOptions && isStreamingOptions(runOptions);
            let response;
            let events = [];
            this.hostedAgent._setIsReplay();
            if (isStreaming) {
                const asyncIterable = await this.hostedAgent.run(runOptions);
                const streamEvents = [];
                logger.info("Processing streaming response...");
                for await (const event of asyncIterable){
                    if (this.replayHook.hasError()) {
                        const error = this.replayHook.getLastError();
                        logger.error(`Error occurred during streaming: ${error?.message}`);
                        throw error;
                    }
                    streamEvents.push(event);
                }
                response = asyncIterable;
                events = this.hostedAgent.getEventStream().getEvents();
                logger.success(`Streaming execution completed with ${streamEvents.length} events`);
            } else {
                response = await this.hostedAgent.run(runOptions);
                if (this.replayHook.hasError()) {
                    const error = this.replayHook.getLastError();
                    logger.error(`Error occurred during execution: ${error?.message}`);
                    throw error;
                }
                events = this.hostedAgent.getEventStream().getEvents();
                logger.success("Execution completed successfully");
            }
            const executedLoops = this.hostedAgent.getCurrentLoopIteration();
            logger.info(`Executed ${executedLoops} agent loops out of ${loopCount} expected loops: ${JSON.stringify(this.options)}`);
            if (executedLoops !== loopCount) throw new Error(`Loop count mismatch: Agent executed ${executedLoops} loops, but fixture has ${loopCount} loop directories`);
            if (this.snapshotManager) await this.snapshotManager.cleanupAllActualFiles(this.snapshotName);
            return {
                response,
                events,
                meta: {
                    snapshotName,
                    executionTime: Date.now() - startTime,
                    loopCount: executedLoops
                }
            };
        } catch (error) {
            logger.error(`Test execution failed: ${error}`);
            throw error;
        } finally{
            this.replayHook.clearError();
        }
    }
    countLoops(casePath) {
        if (!fs.existsSync(casePath)) return 0;
        const loopDirs = fs.readdirSync(casePath).filter((dir)=>dir.startsWith('loop-') && fs.statSync(path.join(casePath, dir)).isDirectory()).sort((a, b)=>{
            const numA = parseInt(a.split('-')[1], 10);
            const numB = parseInt(b.split('-')[1], 10);
            return numA - numB;
        });
        return loopDirs.length;
    }
    getAgent() {
        return this.hostedAgent;
    }
    getCurrentLoop() {
        return this.hostedAgent.getCurrentLoopIteration();
    }
    updateAgentNormalizerConfig(config) {
        this.snapshotManager.updateAgentNormalizerConfig(config);
    }
}
export { AgentSnapshot };
