// @tarko/agent-snapshot/dist/agent-snapshot-runner.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { AgentSnapshot } from "./agent-snapshot.mjs";
class AgentSnapshotRunner {
    examples;
    constructor(examples){
        console.log(JSON.stringify(examples, null, 2));
        this.examples = examples;
    }
    shouldUpdateSnapshots() {
        return process.argv.includes('-u') || process.argv.includes('--updateSnapshot');
    }
    async cli() {
        {
            const args = process.argv.slice(2);
            const command = args[0];
            const exampleName = args[1];
            console.log(args, command, exampleName);
            const updateSnapshots = this.shouldUpdateSnapshots();
            if (updateSnapshots) console.log('Update snapshots mode enabled (-u flag detected)');
            if ('generate' === command) if (exampleName) if ('all' === exampleName) await this.generateAll();
            else {
                const example = this.getCaseByName(exampleName);
                if (example) await this.generateSnapshot(example);
                else {
                    console.error(`Example "${exampleName}" not found.`);
                    process.exit(1);
                }
            }
            else await this.generateAll();
            else if ('replay' === command) if (exampleName) if ('all' === exampleName) await this.replayAll(updateSnapshots);
            else {
                const example = this.getCaseByName(exampleName);
                if (example) await this.replaySnapshot(example, updateSnapshots);
                else {
                    console.error(`Example "${exampleName}" not found.`);
                    process.exit(1);
                }
            }
            else await this.replayAll(updateSnapshots);
            else {
                console.log('Usage: cli.ts [generate|replay] [example-name] [-u|--updateSnapshot]');
                console.log('Options:');
                console.log('  -u, --updateSnapshot    Update snapshots when replaying (skips verification and updates files directly)');
                console.log('Available examples:');
                this.examples.forEach((e)=>console.log(`- ${e.name}`));
                console.log('- all  (all examples)');
            }
        }
    }
    getCaseByName(name) {
        return this.examples.find((e)=>e.name === name);
    }
    async loadSnapshotCase(exampleConfig) {
        const importedModule = await import(exampleConfig.path);
        if (importedModule.agent && importedModule.runOptions) return importedModule;
        if (importedModule.default && importedModule.default.agent && importedModule.default.runOptions) return importedModule.default;
        throw new Error(`Invalid agent case module: ${exampleConfig.path}, required an "agent" instance and "runOptiond" exported`);
    }
    async generateSnapshot(exampleConfig) {
        console.log(`Generating snapshot for ${exampleConfig.name}...`);
        const { agent, runOptions } = await this.loadSnapshotCase(exampleConfig);
        const agentSnapshot = new AgentSnapshot(agent, {
            updateSnapshots: true,
            snapshotPath: exampleConfig.snapshotPath
        });
        await agentSnapshot.generate(runOptions);
        console.log(`Snapshot generated at ${exampleConfig.snapshotPath}`);
    }
    async replaySnapshot(exampleConfig, updateSnapshots = false) {
        console.log(`Testing snapshot for ${exampleConfig.name}...`);
        if (updateSnapshots) console.log("Update mode enabled: will update snapshots if they don't match");
        const { agent, runOptions } = await this.loadSnapshotCase(exampleConfig);
        console.log("Testing agent instance", agent);
        console.log("Testing agent run options", runOptions);
        if (!agent || !runOptions) throw new Error(`Invalid agent case module: ${exampleConfig.path}, required an "agent" instance and "runOptiond" exported`);
        const agentSnapshot = new AgentSnapshot(agent, {
            snapshotPath: exampleConfig.snapshotPath,
            updateSnapshots
        });
        const response = await agentSnapshot.replay(runOptions);
        console.log(`Snapshot test result for ${exampleConfig.name}:`, response);
        return response;
    }
    async generateAll() {
        for (const example of this.examples)await this.generateSnapshot(example);
    }
    async replayAll(updateSnapshots = false) {
        const results = {};
        for (const example of this.examples)results[example.name] = await this.replaySnapshot(example, updateSnapshots);
        return results;
    }
}
export { AgentSnapshotRunner };
