// @tarko/agent-snapshot/dist/agent-hook-base.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import fs from "fs";
import { logger } from "./utils/logger.mjs";
class AgentHookBase {
    agent;
    snapshotPath;
    snapshotName;
    originalRequestHook = null;
    originalResponseHook = null;
    originalLoopEndHook = null;
    originalEachLoopStartHook = null;
    originalStreamingResponseHook = null;
    originalBeforeToolCallHook = null;
    originalAfterToolCallHook = null;
    originalToolCallErrorHook = null;
    originalProcessToolCallsHook = null;
    isHooked = false;
    currentRunOptions;
    snapshotManager;
    lastError = null;
    constructor(agent, options){
        this.agent = agent;
        this.snapshotPath = options.snapshotPath;
        this.snapshotName = options.snapshotName;
        if (!fs.existsSync(this.snapshotPath)) fs.mkdirSync(this.snapshotPath, {
            recursive: true
        });
    }
    setCurrentRunOptions(options) {
        this.currentRunOptions = options;
    }
    hookAgent() {
        if (this.isHooked) return false;
        this.originalRequestHook = this.agent.onLLMRequest;
        this.originalResponseHook = this.agent.onLLMResponse;
        this.originalStreamingResponseHook = this.agent.onLLMStreamingResponse;
        this.originalLoopEndHook = this.agent.onAgentLoopEnd;
        this.originalEachLoopStartHook = this.agent.onEachAgentLoopStart;
        this.originalBeforeToolCallHook = this.agent.onBeforeToolCall;
        this.originalAfterToolCallHook = this.agent.onAfterToolCall;
        this.originalToolCallErrorHook = this.agent.onToolCallError;
        this.originalProcessToolCallsHook = this.agent.onProcessToolCalls;
        this.agent.onLLMRequest = (id, payload)=>this.safeExecuteHook(()=>this.onLLMRequest(id, payload));
        this.agent.onLLMResponse = (id, payload)=>this.safeExecuteHook(()=>this.onLLMResponse(id, payload));
        this.agent.onLLMStreamingResponse = (id, payload)=>this.safeExecuteHook(()=>this.onLLMStreamingResponse(id, payload));
        this.agent.onAgentLoopEnd = (id)=>this.safeExecuteHook(()=>this.onAgentLoopEnd(id));
        this.agent.onEachAgentLoopStart = (id)=>this.safeExecuteHook(()=>this.onEachAgentLoopStart(id));
        this.agent.onBeforeToolCall = (id, toolCall, args)=>this.safeExecuteHook(()=>this.onBeforeToolCall(id, toolCall, args));
        this.agent.onAfterToolCall = (id, toolCall, result)=>this.safeExecuteHook(()=>this.onAfterToolCall(id, toolCall, result));
        this.agent.onToolCallError = (id, toolCall, error)=>this.safeExecuteHook(()=>this.onToolCallError(id, toolCall, error));
        this.agent.onProcessToolCalls = (id, toolCalls)=>this.safeExecuteHook(()=>this.onProcessToolCalls(id, toolCalls));
        this.isHooked = true;
        logger.info(`Hooked into agent: ${this.snapshotName}`);
        return true;
    }
    unhookAgent(force = false) {
        if (!this.isHooked && !force) return false;
        if (this.originalRequestHook) this.agent.onLLMRequest = this.originalRequestHook;
        if (this.originalResponseHook) this.agent.onLLMResponse = this.originalResponseHook;
        if (this.originalStreamingResponseHook) this.agent.onLLMStreamingResponse = this.originalStreamingResponseHook;
        if (this.originalLoopEndHook) this.agent.onAgentLoopEnd = this.originalLoopEndHook;
        if (this.originalEachLoopStartHook) this.agent.onEachAgentLoopStart = this.originalEachLoopStartHook;
        if (this.originalBeforeToolCallHook) this.agent.onBeforeToolCall = this.originalBeforeToolCallHook;
        if (this.originalAfterToolCallHook) this.agent.onAfterToolCall = this.originalAfterToolCallHook;
        if (this.originalToolCallErrorHook) this.agent.onToolCallError = this.originalToolCallErrorHook;
        if (this.originalProcessToolCallsHook) this.agent.onProcessToolCalls = this.originalProcessToolCallsHook;
        this.isHooked = false;
        logger.info(`Unhooked from agent: ${this.snapshotName}`);
        return true;
    }
    async safeExecuteHook(hookFn) {
        try {
            const result = await hookFn();
            if (result instanceof Promise) return result.catch((error)=>{
                this.lastError = error;
                logger.error(`Hook execution error: ${error.message}`);
                throw error;
            });
            return result;
        } catch (error) {
            this.lastError = error;
            logger.error(`Hook execution error: ${error.message}`);
        }
    }
    hasError() {
        return null !== this.lastError;
    }
    getLastError() {
        return this.lastError;
    }
    clearError() {
        this.lastError = null;
    }
    writeStreamingChunks(filePath, chunks) {
        if (!chunks || 0 === chunks.length) return;
        try {
            const chunksAsJsonLines = chunks.map((chunk)=>JSON.stringify(chunk)).join('\n');
            fs.writeFileSync(filePath, chunksAsJsonLines, 'utf-8');
            logger.debug(`${chunks.length} chunks written to ${filePath}`);
        } catch (error) {
            logger.error(`Error writing streaming chunks: ${error}`);
            this.lastError = error;
            throw error;
        }
    }
}
export { AgentHookBase };
