// @tarko/agent-snapshot/dist/agent-replay-snapshot-hook.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import path from "path";
import { SnapshotManager } from "./snapshot-manager.mjs";
import { logger } from "./utils/logger.mjs";
import { AgentHookBase } from "./agent-hook-base.mjs";
class AgentReplaySnapshotHook extends AgentHookBase {
    totalLoops = 0;
    updateSnapshots = false;
    eventStreamStatesByLoop = new Map();
    finalEventStreamState = [];
    mockLLMClient = void 0;
    verifyLLMRequests = true;
    verifyEventStreams = true;
    verifyToolCalls = true;
    toolCallsByLoop = {};
    startTimeByToolCall = {};
    async setup(agent, casePath, totalLoops, options = {}) {
        this.agent = agent;
        this.snapshotPath = casePath;
        this.snapshotName = path.basename(casePath);
        this.totalLoops = totalLoops;
        this.updateSnapshots = options.updateSnapshots || false;
        this.verifyLLMRequests = options.verification?.verifyLLMRequests !== false;
        this.verifyEventStreams = options.verification?.verifyEventStreams !== false;
        this.verifyToolCalls = options.verification?.verifyToolCalls !== false;
        this.snapshotManager = new SnapshotManager(path.dirname(casePath), options.normalizerConfig);
        this.hookAgent();
        this.mockLLMClient = this.createMockLLMClient();
        logger.info(`LLM mocker set up for ${this.snapshotName} with ${totalLoops} loops`);
        logger.info(`Verification settings: LLM requests: ${this.verifyLLMRequests ? 'enabled' : 'disabled'}, Event streams: ${this.verifyEventStreams ? 'enabled' : 'disabled'}, Tool calls: ${this.verifyToolCalls ? 'enabled' : 'disabled'}`);
        if (this.verifyEventStreams) await this.verifyInitialEventStreamState();
    }
    storeFinalEventStreamState(events) {
        this.finalEventStreamState = [
            ...events
        ];
    }
    getFinalEventStreamState() {
        return this.finalEventStreamState;
    }
    getMockLLMClient() {
        return this.mockLLMClient;
    }
    restore() {
        this.unhookAgent();
        this.mockLLMClient = void 0;
        logger.info('Restored original LLM hooks and client');
    }
    createMockLLMClient() {
        return {
            chat: {
                completions: {
                    create: async (request)=>{
                        const currentLoop = this.agent?.getCurrentLoopIteration();
                        logger.info(`[Mock LLM Client] Creating chat completion for loop ${currentLoop} with args: ` + JSON.stringify(request, null, 2));
                        const loopDir = `loop-${currentLoop}`;
                        const mockResponse = await this.snapshotManager?.readSnapshot(path.basename(this.snapshotPath), loopDir, 'llm-response.jsonl');
                        if (!mockResponse) {
                            const error = new Error(`No mock response found for ${loopDir}`);
                            this.lastError = error;
                        }
                        logger.info(`[Mock LLM Response] Loop ${currentLoop}: Type: ${Array.isArray(mockResponse) ? 'array' : 'object'}, Length: ${Array.isArray(mockResponse) ? mockResponse.length : 1}`);
                        logger.success(`\u{2705} Using mock LLM response from snapshot for ${loopDir}`);
                        if (!request.stream) return mockResponse;
                        {
                            const streamResponse = Array.isArray(mockResponse) ? mockResponse : [
                                mockResponse
                            ];
                            logger.info(`Creating streaming response with ${streamResponse.length} chunks for loop ${currentLoop}`);
                            streamResponse.forEach((chunk, idx)=>{
                                if (!chunk.id || !chunk.object || !chunk.choices) logger.warn(`Chunk ${idx} may have invalid structure: ${JSON.stringify(chunk)}`);
                            });
                            return this.createAsyncIterable(streamResponse);
                        }
                    }
                }
            }
        };
    }
    async verifyInitialEventStreamState() {
        if (!this.snapshotPath || !this.snapshotManager || !this.agent) throw new Error('LLMMocker not properly set up');
        logger.info(`\u{1F50D} Verifying initial event stream state before first loop`);
        const events = this.agent.getEventStream().getEvents();
        if (events.length > 0) try {
            await this.snapshotManager.verifyEventStreamSnapshot(path.basename(this.snapshotPath), 'initial', events, this.updateSnapshots);
            logger.success(`\u{2705} Initial event stream verification succeeded`);
        } catch (error) {
            logger.error(`\u{274C} Initial event stream verification failed: ${error}`);
            if (!this.updateSnapshots) throw error;
        }
    }
    createAsyncIterable(chunks) {
        logger.info(`Creating AsyncIterable with ${chunks.length} chunks`);
        return {
            [Symbol.asyncIterator] () {
                let index = 0;
                let iteratorClosed = false;
                logger.info(`AsyncIterator created for ${chunks.length} chunks`);
                return {
                    async next () {
                        if (iteratorClosed) {
                            logger.info("Iterator already closed, returning done");
                            return {
                                done: true,
                                value: void 0
                            };
                        }
                        if (index < chunks.length) {
                            const chunk = chunks[index];
                            logger.info(`Yielding chunk ${index + 1}/${chunks.length}`);
                            index++;
                            return {
                                done: false,
                                value: chunk
                            };
                        }
                        logger.info(`Iterator completed after yielding ${index} chunks`);
                        iteratorClosed = true;
                        return {
                            done: true,
                            value: void 0
                        };
                    },
                    async return () {
                        logger.info(`Iterator return() called early at index ${index}/${chunks.length}`);
                        iteratorClosed = true;
                        return {
                            done: true,
                            value: void 0
                        };
                    },
                    async throw (error) {
                        logger.error(`Error in streaming response iterator: ${error}`);
                        iteratorClosed = true;
                        return {
                            done: true,
                            value: void 0
                        };
                    }
                };
            }
        };
    }
    onEachAgentLoopStart(id) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        if (!this.toolCallsByLoop[currentLoop]) this.toolCallsByLoop[currentLoop] = [];
        if (this.originalEachLoopStartHook) return this.originalEachLoopStartHook.call(this.agent, id);
    }
    async onLLMRequest(id, payload) {
        if (!this.snapshotPath || !this.snapshotManager) throw new Error('LLMMocker not properly set up');
        const currentLoop = this.agent.getCurrentLoopIteration();
        const loopDir = `loop-${currentLoop}`;
        logger.info(`\u{1F504} Intercepted LLM request for loop ${currentLoop}`);
        const events = this.agent.getEventStream().getEvents();
        this.eventStreamStatesByLoop.set(currentLoop, [
            ...events
        ]);
        if (this.verifyEventStreams) try {
            logger.info(`\u{1F50D} Verifying event stream state at the beginning of ${loopDir}`);
            if (this.updateSnapshots) logger.warn(`\u{26A0}\u{FE0F} Update mode enabled: directly updating event stream snapshot for ${loopDir}`);
            await this.snapshotManager.verifyEventStreamSnapshot(path.basename(this.snapshotPath), loopDir, events, this.updateSnapshots);
        } catch (error) {
            logger.error(`\u{274C} Event stream verification failed for ${loopDir}: ${error}`);
            if (!this.updateSnapshots) throw error;
        }
        else logger.info(`Event stream verification skipped for ${loopDir} (disabled in config)`);
        if (this.verifyLLMRequests) try {
            if (this.updateSnapshots) logger.warn(`\u{26A0}\u{FE0F} Update mode enabled: directly updating LLM request snapshot for ${loopDir}`);
            await this.snapshotManager.verifyRequestSnapshot(path.basename(this.snapshotPath), loopDir, payload, this.updateSnapshots);
        } catch (error) {
            logger.error(`\u{274C} Request verification failed for ${loopDir}: ${error}`);
            if (!this.updateSnapshots) throw error;
        }
        else logger.info(`LLM request verification skipped for ${loopDir} (disabled in config)`);
        if (this.originalRequestHook) await this.originalRequestHook.call(this.agent, id, payload);
    }
    async onLLMResponse(id, payload) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        logger.debug(`LLM response hook called for loop ${currentLoop}`);
        if (this.originalResponseHook) await this.originalResponseHook.call(this.agent, id, payload);
    }
    onLLMStreamingResponse(id, payload) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        logger.debug(`LLM onStreamingResponseHook called for loop ${currentLoop}`);
        if (this.originalStreamingResponseHook) this.originalStreamingResponseHook.call(this.agent, id, payload);
    }
    onBeforeToolCall(id, toolCall, args) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        this.startTimeByToolCall[toolCall.toolCallId] = Date.now();
        if (this.verifyToolCalls) this.loadToolCallsFromSnapshot(currentLoop).catch((error)=>{
            logger.error(`Error loading tool calls from snapshot: ${error}`);
            if (!this.updateSnapshots) this.lastError = error instanceof Error ? error : new Error(String(error));
        });
        if (!this.toolCallsByLoop[currentLoop]) this.toolCallsByLoop[currentLoop] = [];
        this.toolCallsByLoop[currentLoop].push({
            toolCallId: toolCall.toolCallId,
            name: toolCall.name,
            args
        });
        logger.debug(`Tool call intercepted for ${toolCall.name} (${toolCall.toolCallId}) in loop ${currentLoop}`);
        if (this.originalBeforeToolCallHook) return this.originalBeforeToolCallHook.call(this.agent, id, toolCall, args);
        return args;
    }
    onAfterToolCall(id, toolCall, result) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        const executionTime = Date.now() - (this.startTimeByToolCall[toolCall.toolCallId] || Date.now());
        if (this.toolCallsByLoop[currentLoop]) {
            const toolCallData = this.toolCallsByLoop[currentLoop].find((tc)=>tc.toolCallId === toolCall.toolCallId);
            if (toolCallData) {
                toolCallData.result = result;
                toolCallData.executionTime = executionTime;
            }
        }
        logger.debug(`Tool call result intercepted for ${toolCall.name} (${toolCall.toolCallId}) in loop ${currentLoop}`);
        if (this.verifyToolCalls) this.verifyToolCallsForLoop(currentLoop).catch((error)=>{
            logger.error(`Error verifying tool calls: ${error}`);
            if (!this.updateSnapshots) this.lastError = error instanceof Error ? error : new Error(String(error));
        });
        if (this.originalAfterToolCallHook) return this.originalAfterToolCallHook.call(this.agent, id, toolCall, result);
        return result;
    }
    onToolCallError(id, toolCall, error) {
        const currentLoop = this.agent.getCurrentLoopIteration();
        const executionTime = Date.now() - (this.startTimeByToolCall[toolCall.toolCallId] || Date.now());
        if (this.toolCallsByLoop[currentLoop]) {
            const toolCallData = this.toolCallsByLoop[currentLoop].find((tc)=>tc.toolCallId === toolCall.toolCallId);
            if (toolCallData) {
                toolCallData.error = error;
                toolCallData.executionTime = executionTime;
            }
        }
        logger.debug(`Tool call error intercepted for ${toolCall.name} (${toolCall.toolCallId}) in loop ${currentLoop}`);
        if (this.verifyToolCalls) this.verifyToolCallsForLoop(currentLoop).catch((error)=>{
            logger.error(`Error verifying tool calls: ${error}`);
            if (!this.updateSnapshots) this.lastError = error instanceof Error ? error : new Error(String(error));
        });
        if (this.originalToolCallErrorHook) return this.originalToolCallErrorHook.call(this.agent, id, toolCall, error);
        return `Error: ${error}`;
    }
    async onProcessToolCalls(id, toolCalls) {
        if (!this.verifyToolCalls) return;
        const currentLoop = this.agent.getCurrentLoopIteration();
        const loopDir = `loop-${currentLoop}`;
        try {
            const savedToolCalls = await this.snapshotManager?.readSnapshot(path.basename(this.snapshotPath), loopDir, 'tool-calls.jsonl');
            if (!savedToolCalls || 0 === savedToolCalls.length) return void logger.warn(`No saved tool calls found for ${loopDir}, executing real tools`);
            if (savedToolCalls.length !== toolCalls.length) logger.warn(`Tool call count mismatch in ${loopDir}: expected ${toolCalls.length} but found ${savedToolCalls.length} in snapshot`);
            const results = [];
            for(let i = 0; i < toolCalls.length; i++){
                const toolCall = toolCalls[i];
                const savedToolCall = savedToolCalls.find((stc)=>stc.name === toolCall.function.name);
                if (savedToolCall) results.push({
                    toolCallId: toolCall.id,
                    toolName: toolCall.function.name,
                    content: savedToolCall.result || savedToolCall.error || 'No result in snapshot'
                });
                else {
                    logger.warn(`No matching saved tool call found for ${toolCall.function.name} in ${loopDir}`);
                    results.push({
                        toolCallId: toolCall.id,
                        toolName: toolCall.function.name,
                        content: "Mock result: No saved result found for this tool in snapshot"
                    });
                }
            }
            logger.info(`Replaying ${results.length} tool call results from snapshot`);
            return results;
        } catch (error) {
            logger.error(`Error replaying tool calls: ${error}`);
            return;
        }
    }
    async loadToolCallsFromSnapshot(loopNumber) {
        if (!this.snapshotManager) return;
        const loopDir = `loop-${loopNumber}`;
        const toolCalls = await this.snapshotManager.readSnapshot(path.basename(this.snapshotPath), loopDir, 'tool-calls.jsonl');
        if (!toolCalls || 0 === toolCalls.length) return void logger.debug(`No tool calls found in snapshot for ${loopDir}`);
        logger.debug(`Loaded ${toolCalls.length} tool calls from snapshot for ${loopDir}`);
    }
    async verifyToolCallsForLoop(loopNumber) {
        if (!this.snapshotManager || !this.toolCallsByLoop[loopNumber]) return;
        const loopDir = `loop-${loopNumber}`;
        try {
            await this.snapshotManager.verifyToolCallsSnapshot(path.basename(this.snapshotPath), loopDir, this.toolCallsByLoop[loopNumber], this.updateSnapshots);
            logger.success(`\u{2705} Tool calls verification succeeded for ${loopDir}`);
        } catch (error) {
            logger.error(`\u{274C} Tool calls verification failed for ${loopDir}: ${error}`);
            if (!this.updateSnapshots) throw error;
        }
    }
    async onAgentLoopEnd(id) {
        if (!this.snapshotPath || !this.snapshotManager || !this.agent) throw new Error('LLMMocker not properly set up');
        logger.info(`\u{1F504} Agent loop execution completed`);
        const finalEvents = this.agent.getEventStream().getEvents();
        this.finalEventStreamState = finalEvents;
        if (this.verifyEventStreams) try {
            logger.info(`\u{1F50D} Verifying final event stream state after agent completion`);
            await this.snapshotManager.verifyEventStreamSnapshot(path.basename(this.snapshotPath), '', JSON.parse(JSON.stringify(finalEvents)), this.updateSnapshots);
            logger.success(`\u{2705} Final event stream verification succeeded`);
        } catch (error) {
            logger.error(`\u{274C} Final event stream verification failed: ${error}`);
            if (!this.updateSnapshots) throw error;
        }
        else logger.info("Final event stream verification skipped (disabled in config)");
        if (!this.hasError() && this.snapshotManager) await this.snapshotManager.cleanupAllActualFiles(path.basename(this.snapshotPath));
        if (this.originalLoopEndHook) await this.originalLoopEndHook.call(this.agent, id);
    }
    getEventStreamStateAfterLoop(loopNumber) {
        const events = this.eventStreamStatesByLoop.get(loopNumber);
        if (!events) throw new Error(`No event stream state found for loop ${loopNumber}`);
        return events;
    }
}
export { AgentReplaySnapshotHook };
