// @tarko/agent-snapshot/dist/snapshot-manager.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import fs from "fs";
import path from "path";
import { logger } from "./utils/logger.mjs";
import { AgentSnapshotNormalizer } from "./utils/snapshot-normalizer.mjs";
class SnapshotManager {
    fixturesRoot;
    normalizer;
    constructor(fixturesRoot, normalizerConfig){
        this.fixturesRoot = fixturesRoot;
        this.normalizer = new AgentSnapshotNormalizer(normalizerConfig);
    }
    getSnapshotPath(caseName, loopDir, filename) {
        if ('' === loopDir) return path.join(this.fixturesRoot, caseName, filename);
        return path.join(this.fixturesRoot, caseName, loopDir, filename);
    }
    async readSnapshot(caseName, loopDir, filename) {
        const filePath = this.getSnapshotPath(caseName, loopDir, filename);
        if (!fs.existsSync(filePath)) return null;
        try {
            const content = fs.readFileSync(filePath, 'utf-8');
            if ('llm-response.jsonl' === filename) try {
                return JSON.parse(content);
            } catch (parseError) {
                const lines = content.split('\n').filter((line)=>line.trim());
                if (lines.length > 0) try {
                    const chunks = lines.map((line)=>JSON.parse(line));
                    return chunks;
                } catch (lineParseError) {
                    logger.error(`Error parsing LLM response as streaming format: ${lineParseError}`);
                    throw lineParseError;
                }
                throw parseError;
            }
            return JSON.parse(content);
        } catch (error) {
            logger.error(`Error reading snapshot from ${filePath}: ${error}`);
            return null;
        }
    }
    async writeSnapshot(caseName, loopDir, filename, data) {
        const filePath = this.getSnapshotPath(caseName, loopDir, filename);
        const dirPath = path.dirname(filePath);
        if (!fs.existsSync(dirPath)) await fs.promises.mkdir(dirPath, {
            recursive: true
        });
        try {
            await fs.promises.writeFile(filePath, JSON.stringify(data, null, 2), 'utf-8');
            logger.info(`Snapshot written to ${filePath}`);
        } catch (error) {
            logger.error(`Error writing snapshot to ${filePath}: ${error}`);
            throw error;
        }
    }
    async writeActualData(caseName, loopDir, filename, data) {
        const actualFilename = filename.replace(/(\.[^.]+)$/, '.actual$1');
        const actualFilePath = this.getSnapshotPath(caseName, loopDir, actualFilename);
        await this.writeSnapshot(caseName, loopDir, actualFilename, data);
        logger.info(`Actual data written to ${actualFilePath}`);
        return actualFilePath;
    }
    async deleteActualDataIfExists(caseName, loopDir, filename) {
        const actualFilename = filename.replace(/(\.[^.]+)$/, '.actual$1');
        const actualFilePath = this.getSnapshotPath(caseName, loopDir, actualFilename);
        if (fs.existsSync(actualFilePath)) try {
            await fs.promises.unlink(actualFilePath);
            logger.info(`Deleted actual data file: ${actualFilePath}`);
        } catch (error) {
            logger.warn(`Failed to delete actual data file ${actualFilePath}: ${error}`);
        }
    }
    async cleanupAllActualFiles(caseName) {
        const casePath = path.join(this.fixturesRoot, caseName);
        if (!fs.existsSync(casePath)) return 0;
        try {
            const findActualFiles = (dir)=>{
                const results = [];
                const files = fs.readdirSync(dir);
                for (const file of files){
                    const fullPath = path.join(dir, file);
                    if (fs.statSync(fullPath).isDirectory()) results.push(...findActualFiles(fullPath));
                    else if (file.includes('.actual.jsonl')) results.push(fullPath);
                }
                return results;
            };
            const actualFiles = findActualFiles(casePath);
            for (const file of actualFiles)try {
                await fs.promises.unlink(file);
                logger.info(`Cleanup: Deleted leftover temporary file ${file}`);
            } catch (err) {
                logger.warn(`Failed to delete temporary file ${file}: ${err}`);
            }
            if (actualFiles.length > 0) logger.success(`\u{2705} Successfully cleaned up ${actualFiles.length} temporary .actual.jsonl files`);
            return actualFiles.length;
        } catch (error) {
            logger.warn(`Failed to perform cleanup of actual files: ${error}`);
            return 0;
        }
    }
    async verifyEventStreamSnapshot(caseName, loopDir, actualEventStream, updateSnapshots = false) {
        const filename = 'event-stream.jsonl';
        const expectedEventStream = await this.readSnapshot(caseName, loopDir, filename);
        if (!expectedEventStream) {
            if (updateSnapshots) {
                await this.writeSnapshot(caseName, loopDir, filename, actualEventStream);
                logger.success(`\u{2705} Created new event stream snapshot for ${caseName}/${loopDir}`);
                return true;
            }
            throw new Error(`No event stream snapshot found for ${caseName}/${loopDir}`);
        }
        if (updateSnapshots) {
            await this.writeSnapshot(caseName, loopDir, filename, actualEventStream);
            logger.warn(`\u{26A0}\u{FE0F} Skipping event stream verification for ${caseName}/${loopDir}, updating snapshot directly`);
            return true;
        }
        const result = this.normalizer.compare(expectedEventStream, actualEventStream);
        if (!result.equal) {
            await this.writeActualData(caseName, loopDir, filename, actualEventStream);
            logger.error(`\u{274C} Event stream comparison failed for ${caseName}/${loopDir}:\n${result.diff}`);
            throw new Error(`Event stream doesn't match for ${caseName}/${loopDir}. Actual data saved to ${loopDir ? `${loopDir}/` : ''}event-stream.actual.jsonl`);
        }
        await this.deleteActualDataIfExists(caseName, loopDir, filename);
        logger.success(`\u{2705} Event stream comparison passed for ${caseName}/${loopDir}`);
        return true;
    }
    async verifyRequestSnapshot(caseName, loopDir, actualRequest, updateSnapshots = false) {
        actualRequest = JSON.parse(JSON.stringify(actualRequest));
        const filename = 'llm-request.jsonl';
        const expectedRequest = await this.readSnapshot(caseName, loopDir, filename);
        if (!expectedRequest) {
            if (updateSnapshots) {
                await this.writeSnapshot(caseName, loopDir, filename, actualRequest);
                logger.success(`\u{2705} Created new request snapshot for ${caseName}/${loopDir}`);
                return true;
            }
            throw new Error(`No request snapshot found for ${caseName}/${loopDir}`);
        }
        if (updateSnapshots) {
            await this.writeSnapshot(caseName, loopDir, filename, actualRequest);
            logger.warn(`\u{26A0}\u{FE0F} Skipping request verification for ${caseName}/${loopDir}, updating snapshot directly`);
            return true;
        }
        const result = this.normalizer.compare(expectedRequest, actualRequest);
        if (!result.equal) {
            await this.writeActualData(caseName, loopDir, filename, actualRequest);
            logger.error(`\u{274C} Request comparison failed for ${caseName}/${loopDir}:\n${result.diff}`);
            throw new Error(`Request doesn't match for ${caseName}/${loopDir}. Actual data saved to ${loopDir}/llm-request.actual.jsonl`);
        }
        await this.deleteActualDataIfExists(caseName, loopDir, filename);
        logger.success(`\u{2705} LLM request comparison passed for ${caseName}/${loopDir}`);
        return true;
    }
    async verifyToolCallsSnapshot(caseName, loopDir, actualToolCalls, updateSnapshots = false) {
        actualToolCalls = JSON.parse(JSON.stringify(actualToolCalls));
        const filename = 'tool-calls.jsonl';
        const expectedToolCalls = await this.readSnapshot(caseName, loopDir, filename);
        if (!expectedToolCalls) {
            if (updateSnapshots) {
                await this.writeSnapshot(caseName, loopDir, filename, actualToolCalls);
                logger.success(`\u{2705} Created new tool calls snapshot for ${caseName}/${loopDir}`);
                return true;
            }
            throw new Error(`No tool calls snapshot found for ${caseName}/${loopDir}`);
        }
        if (updateSnapshots) {
            await this.writeSnapshot(caseName, loopDir, filename, actualToolCalls);
            logger.warn(`\u{26A0}\u{FE0F} Skipping tool calls verification for ${caseName}/${loopDir}, updating snapshot directly`);
            return true;
        }
        const result = this.normalizer.compare(expectedToolCalls, actualToolCalls);
        if (!result.equal) {
            await this.writeActualData(caseName, loopDir, filename, actualToolCalls);
            logger.error(`\u{274C} Tool calls comparison failed for ${caseName}/${loopDir}:\n${result.diff}`);
            throw new Error(`Tool calls don't match for ${caseName}/${loopDir}. Actual data saved to ${loopDir}/tool-calls.actual.jsonl`);
        }
        await this.deleteActualDataIfExists(caseName, loopDir, filename);
        logger.success(`\u{2705} Tool calls comparison passed for ${caseName}/${loopDir}`);
        return true;
    }
    async createTestCaseStructure(caseName, numLoops) {
        const caseDir = path.join(this.fixturesRoot, caseName);
        if (!fs.existsSync(caseDir)) await fs.promises.mkdir(caseDir, {
            recursive: true
        });
        for(let i = 1; i <= numLoops; i++){
            const loopDir = path.join(caseDir, `loop-${i}`);
            if (!fs.existsSync(loopDir)) await fs.promises.mkdir(loopDir, {
                recursive: true
            });
        }
        const initialDir = path.join(caseDir, 'initial');
        if (!fs.existsSync(initialDir)) await fs.promises.mkdir(initialDir, {
            recursive: true
        });
        return caseDir;
    }
    async writeStreamingChunks(caseName, loopDir, filename, chunks, updateIfExists = false) {
        const filePath = this.getSnapshotPath(caseName, loopDir, filename);
        const dirPath = path.dirname(filePath);
        if (!fs.existsSync(dirPath)) await fs.promises.mkdir(dirPath, {
            recursive: true
        });
        if (fs.existsSync(filePath) && !updateIfExists) return void logger.info(`Skipping write to existing file: ${filePath}`);
        try {
            const chunksAsJsonLines = chunks.map((chunk)=>JSON.stringify(chunk)).join('\n');
            await fs.promises.writeFile(filePath, chunksAsJsonLines, 'utf-8');
            logger.info(`Stream chunks written to ${filePath} (${chunks.length} chunks)`);
        } catch (error) {
            logger.error(`Error writing stream chunks to ${filePath}: ${error}`);
            throw error;
        }
    }
    async readStreamingChunks(caseName, loopDir, filename) {
        const filePath = this.getSnapshotPath(caseName, loopDir, filename);
        if (!fs.existsSync(filePath)) return [];
        try {
            const content = await fs.promises.readFile(filePath, 'utf-8');
            const lines = content.split('\n').filter((line)=>line.trim());
            if (0 === lines.length) return [];
            try {
                return lines.map((line)=>JSON.parse(line));
            } catch (lineParseError) {
                logger.error(`Error parsing streaming chunks: ${lineParseError}`);
                throw lineParseError;
            }
        } catch (error) {
            logger.error(`Error reading streaming chunks from ${filePath}: ${error}`);
            return [];
        }
    }
    updateAgentNormalizerConfig(config) {
        this.normalizer = new AgentSnapshotNormalizer(config);
    }
}
export { SnapshotManager };
