// @gui-agent/action-parser/dist/DefaultActionParser.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { BaseActionParser } from "@gui-agent/shared/base";
import { ConsoleLogger, LogLevel } from "@agent-infra/logger";
import { FormatParserChain } from "./FomatParsers.mjs";
import { ActionParserHelper } from "./ActionParserHelper.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const defaultLogger = new ConsoleLogger(void 0, LogLevel.DEBUG);
class DefaultActionParser extends BaseActionParser {
    parsePrediction(input) {
        this.logger.debug('[parsePrediction] starting:', input.length <= 30 ? input : input.substring(0, 30) + '...');
        const originInput = input;
        input = input.trim();
        let reasoningContent = null;
        let rawActionStrings;
        let actions;
        try {
            ({ reasoningContent, rawActionStrings, actions } = this.extractActionStrings(input));
        } catch (error) {
            return this.createErrorResponse(error.message, originInput);
        }
        if (actions && actions.length > 0) return {
            rawContent: originInput,
            reasoningContent,
            rawActionStrings,
            actions
        };
        if (!rawActionStrings || rawActionStrings.length <= 0) return this.createErrorResponse('There is no GUI action detected', originInput);
        actions = [];
        try {
            for (const actionString of rawActionStrings){
                const action = this.helper.parseActionCallString(actionString);
                if (action) actions.push(action);
            }
        } catch (error) {
            return this.createErrorResponse(error.message, originInput);
        }
        this.logger.debug('[parsePrediction] reasoningContent:', reasoningContent);
        this.logger.debug('[parsePrediction] actions lenth:', actions.length);
        return {
            rawContent: originInput,
            reasoningContent,
            rawActionStrings,
            actions
        };
    }
    extractActionStrings(input) {
        const parserChain = new FormatParserChain(this.logger);
        const { reasoningContent, rawActionStrings, actions } = parserChain.parse(input);
        this.logger.debug('[extractActionStrings] result of chains:', {
            reasoningContent,
            rawActionStrings
        });
        return {
            reasoningContent: reasoningContent || void 0,
            rawActionStrings: rawActionStrings || void 0,
            actions: actions || void 0
        };
    }
    createErrorResponse(errorMessage, rawContent) {
        return {
            errorMessage,
            rawContent,
            actions: []
        };
    }
    constructor(logger = defaultLogger){
        super(), _define_property(this, "logger", void 0), _define_property(this, "helper", void 0);
        this.logger = logger.spawn('[DefaultActionParser]');
        this.helper = new ActionParserHelper(this.logger);
    }
}
export { DefaultActionParser };
