// @gui-agent/action-parser/dist/FomatParsers.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ConsoleLogger, LogLevel } from "@agent-infra/logger";
import { XMLBuilder, XMLParser } from "fast-xml-parser";
import { ActionParserHelper } from "./ActionParserHelper.mjs";
import { serializeAction } from "@gui-agent/shared/utils";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class XMLFormatParser {
    parse(text) {
        if (text.includes('computer_env')) {
            this.logger.debug('[XMLFormatParser] canParse:', false);
            return null;
        }
        const builder = new XMLBuilder();
        const parser = new XMLParser({
            ignoreAttributes: false
        });
        const object = parser.parse(text);
        if (!object || 'object' != typeof object) {
            this.logger.debug('[XMLFormatParser] canParse:', false);
            return null;
        }
        let canParse = false;
        let reasoningContent = null;
        const actions = [];
        for (const [key, value] of Object.entries(object)){
            this.logger.debug('[XMLFormatParser] key:', key);
            this.logger.debug('[XMLFormatParser] value:', value);
            if (/^think.*$/.test(key)) {
                reasoningContent = 'string' == typeof value ? value : builder.build(value);
                continue;
            }
            if ('answer' === key) {
                canParse = true;
                actions.push({
                    type: 'finished',
                    inputs: {
                        content: value
                    }
                });
                continue;
            }
            if ('seed:tool_call' === key) {
                canParse = true;
                actions.push(...this.helper.standardizeGUIActionsFromXMLObject(value));
                continue;
            }
        }
        this.logger.debug('[XMLFormatParser] canParse:', canParse);
        if (!canParse) return null;
        if (actions.length <= 0) throw Error('No valid GUI action string was detected');
        const rawActionStrings = [];
        for (const action of actions)rawActionStrings.push(serializeAction(action));
        return {
            reasoningContent,
            rawActionStrings,
            actions
        };
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        _define_property(this, "helper", void 0);
        this.logger = logger;
        this.helper = new ActionParserHelper(this.logger);
    }
}
class OmniFormatParser {
    canParse(text) {
        const canParse = text.includes('<computer_env>') || text.includes('<answer>') && text.includes('</answer>');
        this.logger.debug('[OmniFormatParser] canParse:', canParse);
        return canParse;
    }
    parse(text) {
        if (!this.canParse(text)) return null;
        const thinkMatch = text.match(/<think[^>]*>([\s\S]*?)<\/think[^>]*>/i);
        const reasoningContent = thinkMatch ? thinkMatch[1].trim() : null;
        let actionStr = '';
        const computerEnvMatch = text.match(/<computer_env>([\s\S]*?)<\/computer_env>/i);
        if (computerEnvMatch) {
            actionStr = computerEnvMatch[1].trim();
            actionStr = actionStr.replace(/^Action:\s*/i, '');
        } else {
            var _answerMatch_;
            const answerMatch = text.match(/<answer>([\s\S]*?)<\/answer>/i);
            const finishContent = null == answerMatch ? void 0 : null == (_answerMatch_ = answerMatch[1]) ? void 0 : _answerMatch_.trim();
            actionStr = `finished(content='${finishContent}')`;
        }
        const result = {
            reasoningContent,
            rawActionStrings: actionStr.split('\n\n').filter((action)=>'' !== action.trim()),
            actions: void 0
        };
        return result;
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        this.logger = logger;
    }
}
class UnifiedBCFormatParser {
    canParse(text) {
        const hasBasicStructure = text.includes('Thought:') && text.includes('Action:') && !text.includes('Reflection:') && !text.includes('Action_Summary:');
        this.logger.debug('[UnifiedBCFormatParser] canParse:', hasBasicStructure);
        return hasBasicStructure;
    }
    parse(text) {
        if (!this.canParse(text)) return null;
        const thoughtMatch = text.match(/Thought:\s*([\s\S]+?)(?=\s*Action:|$)/);
        const reasoningContent = thoughtMatch ? thoughtMatch[1].trim() : null;
        let actionStr = '';
        if (text.includes('Action:')) {
            const actionParts = text.split(/Action:/);
            actionStr = actionParts[actionParts.length - 1].trim();
        } else actionStr = text;
        return {
            reasoningContent,
            rawActionStrings: actionStr.split('\n\n').filter((action)=>'' !== action.trim()),
            actions: void 0
        };
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        this.logger = logger;
    }
}
class BCComplexFormatParser {
    canParse(text) {
        const canParse = text.includes('Reflection:') && text.includes('Action_Summary:') || text.startsWith('Action_Summary:');
        this.logger.debug('[BCComplexFormatParser] canParse:', canParse);
        return canParse;
    }
    parse(text) {
        if (!this.canParse(text)) return null;
        let thought = null;
        let reflection = null;
        let actionStr = '';
        if (text.startsWith('Reflection:')) {
            const reflectionMatch = text.match(/Reflection:\s*([\s\S]+?)Action_Summary:\s*([\s\S]+?)(?=\s*Action[：:]|$)/);
            if (reflectionMatch) {
                reflection = reflectionMatch[1].trim();
                thought = reflectionMatch[2].trim();
                this.logger.debug('[BCComplexFormatParser] Reflection and Action_Summary');
            }
        } else if (text.startsWith('Action_Summary:')) {
            const summaryMatch = text.match(/Action_Summary:\s*([\s\S]+?)(?=\s*Action[：:]|$)/);
            if (summaryMatch) {
                thought = summaryMatch[1].trim();
                this.logger.debug('[BCComplexFormatParser] Only Action_Summary');
            }
        }
        if (text.includes('Action:') || text.includes("Action\uFF1A")) {
            const actionParts = text.split(/Action[：:]/);
            actionStr = actionParts[actionParts.length - 1].trim();
        }
        return {
            reasoningContent: reflection && thought ? `${reflection}, ${thought}` : thought ?? reflection,
            rawActionStrings: actionStr.split('\n\n').filter((action)=>'' !== action.trim()),
            actions: void 0
        };
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        this.logger = logger;
    }
}
class O1FormatParser {
    canParse(text) {
        const canParse = text.includes('<Thought>') && text.includes('</Thought>');
        this.logger.debug('[O1FormatParser] canParse:', canParse);
        return canParse;
    }
    parse(text) {
        if (!this.canParse(text)) return null;
        const thoughtMatch = text.match(/<Thought>\s*([\s\S]*?)\s*<\/Thought>/s);
        const actionSummaryMatch = text.match(/Action_Summary:\s*([\s\S]*?)\s*Action:/s);
        const actionMatch = text.match(/Action:\s*([\s\S]*?)\s*<\/Output>/s);
        const thoughtContent = thoughtMatch ? thoughtMatch[1].trim() : null;
        const actionSummaryContent = actionSummaryMatch ? actionSummaryMatch[1].trim() : null;
        const actionContent = actionMatch ? actionMatch[1].trim() : '';
        const thought = actionSummaryContent ? `${thoughtContent}, ${actionSummaryContent}` : thoughtContent;
        return {
            reasoningContent: thought,
            rawActionStrings: actionContent.split('\n\n').filter((action)=>'' !== action.trim()),
            actions: void 0
        };
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        this.logger = logger;
    }
}
class FallbackFormatParser {
    parse(text) {
        this.logger.debug('[FallbackFormatParser] canParse: always true');
        const regex = /\w+\((?:[^()"']|"[^"]*"|'[^']*'|\([^()]*\))*\)/;
        const actionMatch = text.match(regex);
        if (!actionMatch) {
            this.logger.error('[FallbackFormatParser] there is no function call format in the text');
            throw Error('No valid GUI action string was detected');
        }
        const actionStr = actionMatch[0].trim();
        const thoughtMatch = text.match(/Thought:\s*([\s\S]+?)(?=\s*Action[：:]|$)/);
        const thoughtStr = thoughtMatch ? thoughtMatch[1].trim() : null;
        return {
            reasoningContent: thoughtStr,
            rawActionStrings: actionStr.split('\n\n').filter((action)=>'' !== action.trim()),
            actions: void 0
        };
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        this.logger = logger;
    }
}
const defaultLogger = new ConsoleLogger(void 0, LogLevel.DEBUG);
class FormatParserChain {
    parse(text) {
        this.logger.debug('[FormatParserChain] start...');
        for (const parser of this.parsers){
            const result = parser.parse(text);
            if (result) return result;
        }
        this.logger.warn('[FormatParserChain]', 'No appropriate parser found');
        throw Error('No valid GUI action string was detected');
    }
    constructor(logger = defaultLogger){
        _define_property(this, "logger", void 0);
        _define_property(this, "parsers", void 0);
        this.logger = logger.spawn('[FormatParserChain]');
        this.parsers = [
            new XMLFormatParser(this.logger),
            new OmniFormatParser(this.logger),
            new UnifiedBCFormatParser(this.logger),
            new BCComplexFormatParser(this.logger),
            new O1FormatParser(this.logger),
            new FallbackFormatParser(this.logger)
        ];
    }
}
export { FormatParserChain, OmniFormatParser, UnifiedBCFormatParser, XMLFormatParser };
