// @omni-tars/gui-agent/dist/GUIAgentToolCallEngine.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ToolCallEngine } from "@tarko/agent-interface";
import { createSystemPromptGroup, createT5InitState, processT5StreamingChunk } from "@omni-tars/core";
import { getLogger } from "@tarko/agent";
import { GUIAgentT5Adapter } from "./GUIAgentT5Adapter.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class GUIAgentToolCallEngine extends ToolCallEngine {
    preparePrompt(instructions, tools) {
        return createSystemPromptGroup(this.agentMode);
    }
    prepareRequest(context) {
        return {
            model: context.model,
            messages: context.messages,
            temperature: context.temperature || 1.0,
            top_p: context.top_p,
            stream: true
        };
    }
    initStreamProcessingState() {
        return createT5InitState();
    }
    processStreamingChunk(chunk, state) {
        return processT5StreamingChunk(chunk, state);
    }
    finalizeStreamProcessing(state) {
        const fullContent = state.contentBuffer;
        this.logger.debug('finalizeStreamProcessing fullContent:', fullContent);
        const toolCalls = state.toolCalls;
        this.logger.debug('finalizeStreamProcessing toolCalls:', toolCalls);
        const convertedToolCalls = this.t5Adapter.convertToBrowserVisionControlCalls(toolCalls, state.reasoningBuffer ?? '');
        return {
            content: state.accumulatedChatContentBuffer ?? fullContent,
            rawContent: fullContent,
            reasoningContent: state.reasoningBuffer ?? '',
            toolCalls: convertedToolCalls,
            finishReason: (toolCalls || []).length > 0 ? 'tool_calls' : 'stop'
        };
    }
    buildHistoricalAssistantMessage(currentLoopAssistantEvent) {
        return {
            role: 'assistant',
            content: currentLoopAssistantEvent.rawContent || currentLoopAssistantEvent.content
        };
    }
    buildHistoricalToolCallResultMessages(toolCallResults) {
        return toolCallResults.map((result)=>{
            const textContent = result.content.filter((part)=>'text' === part.type).map((part)=>part.text).join('');
            return {
                role: 'user',
                content: `Tool "${result.toolName}" result:\n${textContent}`
            };
        });
    }
    generateToolCallId() {
        return `call_${Date.now()}_${Math.random().toString(36).substring(2, 11)}`;
    }
    constructor(...args){
        super(), _define_property(this, "logger", getLogger('GUIAgentToolCallEngine')), _define_property(this, "t5Adapter", new GUIAgentT5Adapter(this.logger)), _define_property(this, "agentMode", void 0);
        const agentMode = args[0];
        this.agentMode = agentMode;
    }
}
export { GUIAgentToolCallEngine };
