// @omni-tars/gui-agent/dist/GuiAgentPlugin.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { AgentPlugin, COMPUTER_USE_ENVIRONMENT } from "@omni-tars/core";
import { Tool } from "@tarko/agent";
import { ConsoleLogger, LogLevel, createGUIErrorResponse } from "@tarko/shared-utils";
import { Base64ImageParser } from "@agent-infra/media-utils";
import { ImageCompressor, formatBytes } from "./utils/ImageCompressor.mjs";
import { sleep } from "@gui-agent/shared/utils";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const guiLogger = new ConsoleLogger('[O-GUIAgent]', LogLevel.DEBUG);
const LOW_DETAIL_THRESHOLD = 1048576;
class GuiAgentPlugin extends AgentPlugin {
    async initialize() {
        if (this.agentMode && 'game' === this.agentMode.id && !this.initializedWithMode) {
            if (this.agentMode.link) {
                await this.emitPresetUserQuey();
                await this.emitScreenshotEvent(4000);
                await this.emitPresetAssistantNavMsg();
            } else await this.emitPresetAssistantWaitMsg();
            this.initializedWithMode = true;
        }
        this.agent.registerTool(new Tool({
            id: 'browser_vision_control',
            description: 'operator tool',
            parameters: {},
            function: async (input)=>{
                try {
                    guiLogger.info('browser_vision_control', input);
                    const op = await this.operatorManager.getInstance();
                    const rawResult = await (null == op ? void 0 : op.doExecute({
                        rawContent: '',
                        rawActionStrings: [],
                        actions: input.operator_action
                    }));
                    if (null == rawResult ? void 0 : rawResult.errorMessage) {
                        guiLogger.error('execute error', null == rawResult ? void 0 : rawResult.errorMessage);
                        return createGUIErrorResponse(input.action, null == rawResult ? void 0 : rawResult.errorMessage);
                    }
                    return {
                        success: true,
                        action: input.action,
                        normalizedAction: input.action_for_gui,
                        observation: void 0
                    };
                } catch (error) {
                    guiLogger.error('execute error from trycatch', error instanceof Error ? error.message : 'Unknown error');
                    return createGUIErrorResponse(input.action, error);
                }
            }
        }));
    }
    async onLLMRequest(id, payload) {}
    async onAfterToolCall(id, toolCall, result) {
        guiLogger.info('onAfterToolCall toolCall', JSON.stringify(toolCall));
        if ('browser_vision_control' !== toolCall.name) return void guiLogger.info('onAfterToolCall: skipping screenshot');
        await this.emitScreenshotEvent();
    }
    async emitScreenshotEvent(interval = 0) {
        const operator = await this.operatorManager.getInstance();
        if (interval > 0) await sleep(interval);
        const output = await (null == operator ? void 0 : operator.doScreenshot());
        if (!(null == output ? void 0 : output.base64)) return void guiLogger.error('emitScreenshotEvent: Failed to get screenshot');
        const base64Tool = new Base64ImageParser(output.base64);
        const originalBuffer = Buffer.from(output.base64, 'base64');
        const originalSize = originalBuffer.byteLength;
        const { width, height } = base64Tool.getDimensions() || {
            width: -1,
            height: -1
        };
        guiLogger.debug(`emitScreenshotEvent: original image size: ${width}x${height}`);
        const compressor = new ImageCompressor({
            quality: 80,
            format: 'webp'
        });
        const compressedBuffer = await compressor.compressToBuffer(originalBuffer);
        const compressedBase64 = `data:image/webp;base64,${compressedBuffer.toString('base64')}`;
        const compressedSize = compressedBuffer.byteLength;
        const compressionRatio = ((originalSize - compressedSize) / originalSize * 100).toFixed(2);
        const compressedBase64Tool = new Base64ImageParser(compressedBase64);
        const { width: compressedWidth, height: compressedHeight } = compressedBase64Tool.getDimensions() || {
            width: -1,
            height: -1
        };
        guiLogger.debug(`emitScreenshotEvent: compressed image size: ${compressedWidth}x${compressedHeight}`);
        guiLogger.debug("emitScreenshotEvent compression stat: ", {
            originalSize: formatBytes(originalSize),
            compressedSize: formatBytes(compressedSize),
            compressionRatio: `${compressionRatio}% reduction`
        });
        const pixelCount = compressedWidth * compressedHeight;
        let detailLevel;
        detailLevel = pixelCount <= LOW_DETAIL_THRESHOLD ? 'low' : 'high';
        guiLogger.debug("emitScreenshotEvent: detail parameter selection", {
            selectedDetail: detailLevel
        });
        const content = [
            {
                type: 'image_url',
                image_url: {
                    url: compressedBase64,
                    detail: detailLevel
                }
            }
        ];
        if (null == output ? void 0 : output.url) content.push({
            type: 'text',
            text: `The current page's url: ${output.url}`
        });
        const eventStream = this.agent.getEventStream();
        const events = eventStream.getEvents();
        guiLogger.info('emitScreenshotEvent events length:', events.length);
        const event = eventStream.createEvent('environment_input', {
            description: 'Browser Screenshot',
            content,
            metadata: {
                type: 'screenshot',
                url: null == output ? void 0 : output.url
            }
        });
        eventStream.sendEvent(event);
    }
    async emitPresetUserQuey() {
        const eventStream = this.agent.getEventStream();
        const events = eventStream.getEvents();
        const hasUserMessage = events.some((event)=>'user_message' === event.type);
        if (!hasUserMessage) {
            const event = eventStream.createEvent('user_message', {
                content: `Goto: ${this.agentMode.link}`
            });
            eventStream.sendEvent(event);
        }
    }
    async emitPresetAssistantNavMsg() {
        const eventStream = this.agent.getEventStream();
        const events = eventStream.getEvents();
        const hasAssistantMessage = events.some((event)=>'assistant_message' === event.type);
        if (!hasAssistantMessage) {
            const event = eventStream.createEvent('assistant_message', {
                content: `Successfully navigated to ${this.agentMode.link}, and the page has loaded.`
            });
            eventStream.sendEvent(event);
        }
    }
    async emitPresetAssistantWaitMsg() {
        const eventStream = this.agent.getEventStream();
        const events = eventStream.getEvents();
        const hasAssistantMessage = events.some((event)=>'assistant_message' === event.type);
        if (!hasAssistantMessage) {
            const event = eventStream.createEvent('assistant_message', {
                content: `Just tell me which game you want me to play, and I\u{2019}ll begin!`
            });
            eventStream.sendEvent(event);
        }
    }
    constructor(option){
        super(), _define_property(this, "name", 'gui-agent'), _define_property(this, "environmentSection", COMPUTER_USE_ENVIRONMENT), _define_property(this, "agentMode", void 0), _define_property(this, "operatorManager", void 0), _define_property(this, "initializedWithMode", false);
        this.agentMode = option.agentMode;
        this.operatorManager = option.operatorManager;
        if (this.agentMode) guiLogger.info('AgentMode:', JSON.stringify(this.agentMode));
    }
}
export { GuiAgentPlugin };
