// @omni-tars/gui-agent/dist/GUIAgentT5Adapter.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ConsoleLogger } from "@agent-infra/logger";
import { LogLevel } from "@tarko/agent-interface";
import { ActionParserHelper } from "@gui-agent/action-parser";
import { serializeAction } from "@gui-agent/shared/utils";
import { convertToAgentUIAction } from "./utils.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class GUIAgentT5Adapter {
    convertToBrowserVisionControlCalls(tools, reasoningBuffer) {
        return tools.map((tool)=>{
            this.logger.debug(`[convertToBVCCalls] Processing tool: ${JSON.stringify(tool)}`);
            const { roughType, roughInputs } = this.helper.parseRoughFromFunctionCall(tool.function);
            const operator_action = this.helper.standardizeAction(roughType, roughInputs);
            this.logger.debug(`[convertToBVCCalls] Standardized action: ${JSON.stringify(operator_action)}`);
            this.normalizeActionCoordinates(operator_action.inputs);
            this.logger.debug(`[convertToBVCCalls] Normalized action: ${JSON.stringify(operator_action)}`);
            const actionString = serializeAction(operator_action);
            this.logger.debug(`[convertToBVCCalls] Serialized action: ${actionString}`);
            const action_for_gui_render = convertToAgentUIAction(operator_action);
            this.logger.debug(`[convertToBVCCalls] action_for_gui: ${JSON.stringify(action_for_gui_render)}`);
            const browserCall = {
                id: tool.id,
                type: 'function',
                function: {
                    name: 'browser_vision_control',
                    arguments: JSON.stringify({
                        action: actionString,
                        step: '',
                        thought: reasoningBuffer,
                        operator_action: [
                            operator_action
                        ],
                        action_for_gui: action_for_gui_render
                    })
                }
            };
            this.logger.debug(`[convertToBVCCalls] Generated browser call for tool ${tool.id}`);
            return browserCall;
        });
    }
    normalizeActionCoordinates(actionInputs) {
        if (!actionInputs || 'object' != typeof actionInputs) return;
        if (actionInputs.point && 'object' == typeof actionInputs.point) actionInputs.point = this.normalizeCoordinates(actionInputs.point);
        if (actionInputs.start && 'object' == typeof actionInputs.start) actionInputs.start = this.normalizeCoordinates(actionInputs.start);
        if (actionInputs.end && 'object' == typeof actionInputs.end) actionInputs.end = this.normalizeCoordinates(actionInputs.end);
    }
    normalizeCoordinates(coords) {
        if (coords.raw) coords.normalized = {
            x: coords.raw.x / 1000,
            y: coords.raw.y / 1000
        };
        return coords;
    }
    constructor(logger){
        _define_property(this, "logger", void 0);
        _define_property(this, "helper", void 0);
        this.logger = logger || new ConsoleLogger('[GUIAgentT5Adapter]', LogLevel.DEBUG);
        this.helper = new ActionParserHelper(this.logger);
    }
}
export { GUIAgentT5Adapter };
