// @omni-tars/gui-agent/dist/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ComposableAgent } from "@omni-tars/core";
import { GuiAgentPlugin } from "./GuiAgentPlugin.mjs";
import { GUIAgentToolCallEngine } from "./GUIAgentToolCallEngine.mjs";
import { OperatorManager } from "./OperatorManager.mjs";
import { GuiToolCallEngineProvider } from "./GuiToolCallEngineProvider.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class GUIAgent extends ComposableAgent {
    constructor(options){
        super({
            ...options,
            plugins: [
                new GuiAgentPlugin({
                    operatorManager: OperatorManager.create(options.agentMode, options.sandboxUrl),
                    agentMode: options.agentMode
                })
            ],
            toolCallEngine: GUIAgentToolCallEngine
        });
    }
}
_define_property(GUIAgent, "label", void 0);
export { GuiAgentPlugin, GuiToolCallEngineProvider, OperatorManager, GUIAgent as default };
