// @omni-tars/gui-agent/dist/OperatorManager.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { AIOGameOperator, AIOHybridOperator } from "@gui-agent/operator-aio";
import { getAioUrl } from "@omni-tars/core";
import { defaultLogger } from "@agent-infra/logger";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class OperatorManager {
    async init() {
        var _this_operator;
        await (null == (_this_operator = this.operator) ? void 0 : _this_operator.doInitialize());
        this.initialized = true;
    }
    async getInstance() {
        if (!this.initialized) await this.init();
        return this.operator;
    }
    getMode() {
        return this.agentMode;
    }
    static create(agentMode, sandboxUrl) {
        return new OperatorManager(agentMode ?? {
            id: 'gui',
            browserMode: 'hybrid'
        }, sandboxUrl);
    }
    constructor(agentMode, sandboxUrl){
        _define_property(this, "agentMode", void 0);
        _define_property(this, "aioClient", null);
        _define_property(this, "remoteOperator", null);
        _define_property(this, "remoteBrowser", null);
        _define_property(this, "operator", null);
        _define_property(this, "browser", null);
        _define_property(this, "sandboxUrl", void 0);
        _define_property(this, "initialized", false);
        this.agentMode = agentMode;
        this.sandboxUrl = sandboxUrl ?? getAioUrl();
        if ('game' === this.agentMode.id) {
            const targetUrl = this.agentMode.link;
            if (!targetUrl) defaultLogger.warn('Game agent mode link is null');
            this.operator = new AIOGameOperator({
                baseURL: this.sandboxUrl,
                timeout: 10000,
                targetUrl
            });
            this.initialized = true;
        } else this.operator = new AIOHybridOperator({
            baseURL: this.sandboxUrl,
            timeout: 10000
        });
    }
}
export { OperatorManager };
