// @omni-tars/gui-agent/dist/utils.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
function convertToAgentUIAction(action) {
    const { type: action_type, inputs: action_inputs } = action;
    switch(action_type){
        case 'click':
        case 'left_click':
        case 'left_single':
            return {
                type: 'click',
                inputs: {
                    startX: action_inputs.point.normalized.x || 0,
                    startY: action_inputs.point.normalized.y || 0
                }
            };
        case 'double_click':
        case 'left_double':
            return {
                type: 'double_click',
                inputs: {
                    startX: action_inputs.point.normalized.x || 0,
                    startY: action_inputs.point.normalized.y || 0
                }
            };
        case 'right_click':
        case 'right_single':
            return {
                type: 'right_click',
                inputs: {
                    startX: action_inputs.point.normalized.x || 0,
                    startY: action_inputs.point.normalized.y || 0
                }
            };
        case 'drag':
            return {
                type: 'drag',
                inputs: {
                    startX: action_inputs.start.normalized.x || 0,
                    startY: action_inputs.start.normalized.y || 0,
                    endX: action_inputs.end.normalized.x || 0,
                    endY: action_inputs.end.normalized.y || 0
                }
            };
        case 'scroll':
            var _action_inputs_point_normalized, _action_inputs_point, _action_inputs_point_normalized1, _action_inputs_point1;
            return {
                type: 'scroll',
                inputs: {
                    startX: (null == (_action_inputs_point = action_inputs.point) ? void 0 : null == (_action_inputs_point_normalized = _action_inputs_point.normalized) ? void 0 : _action_inputs_point_normalized.x) || 0,
                    startY: (null == (_action_inputs_point1 = action_inputs.point) ? void 0 : null == (_action_inputs_point_normalized1 = _action_inputs_point1.normalized) ? void 0 : _action_inputs_point_normalized1.y) || 0,
                    direction: action_inputs.direction || 'down'
                }
            };
        case 'navigate':
            return {
                type: 'navigate',
                inputs: {
                    url: action_inputs.url || action_inputs.content || ''
                }
            };
        default:
            return action;
    }
}
function createErrorAction() {
    return {
        type: 'wait',
        inputs: {}
    };
}
export { convertToAgentUIAction, createErrorAction };
