// @omni-tars/mcp-agent/dist/tools/linkReader.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Tool, z } from "@tarko/agent";
import { McpManager } from "./mcp.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class LinkReaderToolProvider {
    getTool() {
        return new Tool({
            id: 'LinkReader',
            description: '',
            parameters: z.object({
                description: z.string().describe("A detailed description of the content to be extracted from the current URL."),
                url: z.string().describe('The target link, which should be a complete URL (starting with http).')
            }),
            function: async ({ description, url })=>this.mcpManager.client.callTool({
                    client: McpManager.McpClientType.LinkReader,
                    name: 'text_browser_view',
                    args: {
                        url,
                        description
                    }
                })
        });
    }
    constructor(mcpManager){
        _define_property(this, "mcpManager", void 0);
        this.mcpManager = mcpManager;
    }
}
export { LinkReaderToolProvider };
