// @omni-tars/mcp-agent/dist/tools/search.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Tool, z } from "@tarko/agent";
import { McpManager } from "./mcp.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class SearchToolProvider {
    getTool() {
        return new Tool({
            id: 'Search',
            description: '',
            parameters: z.object({
                query: z.string().describe('The search query to research')
            }),
            function: async ({ query })=>this.mcpManager.client.callTool({
                    client: McpManager.McpClientType.Google,
                    name: 'google_search',
                    args: {
                        q: query
                    }
                })
        });
    }
    constructor(mcpManager){
        _define_property(this, "mcpManager", void 0);
        this.mcpManager = mcpManager;
    }
}
export { SearchToolProvider };
