// @tarko/agent-snapshot/dist/utils/snapshot-normalizer.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import fast_json_stable_stringify from "fast-json-stable-stringify";
const DEFAULT_CONFIG = {
    fieldsToNormalize: [
        {
            pattern: /id$/,
            replacement: '<<ID>>'
        },
        {
            pattern: 'timestamp',
            replacement: '<<TIMESTAMP>>'
        },
        {
            pattern: 'created',
            replacement: '<<TIMESTAMP>>'
        },
        {
            pattern: 'startTime',
            replacement: '<<TIMESTAMP>>'
        },
        {
            pattern: 'elapsedMs',
            replacement: '<<elapsedMs>>'
        },
        {
            pattern: 'image_url',
            replacement: '<<image_url>>'
        },
        {
            pattern: 'toolCallId',
            replacement: '<<toolCallId>>'
        },
        {
            pattern: 'sessionId',
            replacement: '<<sessionId>>'
        },
        {
            pattern: 'messageId',
            replacement: '<<messageId>>'
        },
        {
            pattern: 'ttftMs',
            replacement: '<<ttftMs>>'
        },
        {
            pattern: 'ttltMs',
            replacement: '<<ttltMs>>'
        },
        {
            pattern: /Time$/,
            replacement: '<<TIMESTAMP>>'
        }
    ],
    fieldsToIgnore: []
};
class SimpleDiffer {
    contextLines;
    constructor(contextLines = 3){
        this.contextLines = contextLines;
    }
    diff(expected, actual, expectedLabel = 'Expected', actualLabel = 'Actual') {
        const expectedLines = expected.split('\n');
        const actualLines = actual.split('\n');
        const diffLines = [];
        diffLines.push(`--- ${expectedLabel}`);
        diffLines.push(`+++ ${actualLabel}`);
        const lcs = this.longestCommonSubsequence(expectedLines, actualLines);
        const changes = this.generateChanges(expectedLines, actualLines, lcs);
        const hunks = this.groupChangesIntoHunks(changes, expectedLines.length, actualLines.length);
        for (const hunk of hunks){
            diffLines.push(`@@ -${hunk.oldStart},${hunk.oldCount} +${hunk.newStart},${hunk.newCount} @@`);
            diffLines.push(...hunk.lines);
        }
        return diffLines.join('\n');
    }
    longestCommonSubsequence(a, b) {
        const m = a.length;
        const n = b.length;
        const dp = Array(m + 1).fill(null).map(()=>Array(n + 1).fill(0));
        for(let i = 1; i <= m; i++)for(let j = 1; j <= n; j++)if (a[i - 1] === b[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
        else dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
        return dp;
    }
    generateChanges(expected, actual, lcs) {
        const changes = [];
        let i = expected.length;
        let j = actual.length;
        while(i > 0 || j > 0)if (i > 0 && j > 0 && expected[i - 1] === actual[j - 1]) {
            changes.unshift({
                type: 'equal',
                expectedIndex: i - 1,
                actualIndex: j - 1,
                line: expected[i - 1]
            });
            i--;
            j--;
        } else if (j > 0 && (0 === i || lcs[i][j - 1] >= lcs[i - 1][j])) {
            changes.unshift({
                type: 'add',
                expectedIndex: -1,
                actualIndex: j - 1,
                line: actual[j - 1]
            });
            j--;
        } else if (i > 0) {
            changes.unshift({
                type: 'remove',
                expectedIndex: i - 1,
                actualIndex: -1,
                line: expected[i - 1]
            });
            i--;
        }
        return changes;
    }
    groupChangesIntoHunks(changes, expectedLength, actualLength) {
        const hunks = [];
        let currentHunk = null;
        for(let i = 0; i < changes.length; i++){
            const change = changes[i];
            if ('equal' !== change.type) {
                if (!currentHunk) {
                    const contextStart = Math.max(0, i - this.contextLines);
                    currentHunk = {
                        oldStart: changes[contextStart]?.expectedIndex + 1 || 1,
                        oldCount: 0,
                        newStart: changes[contextStart]?.actualIndex + 1 || 1,
                        newCount: 0,
                        lines: []
                    };
                    for(let j = contextStart; j < i; j++)if ('equal' === changes[j].type) {
                        currentHunk.lines.push(` ${changes[j].line}`);
                        currentHunk.oldCount++;
                        currentHunk.newCount++;
                    }
                }
                if ('remove' === change.type) {
                    currentHunk.lines.push(`-${change.line}`);
                    currentHunk.oldCount++;
                } else if ('add' === change.type) {
                    currentHunk.lines.push(`+${change.line}`);
                    currentHunk.newCount++;
                }
            } else if (currentHunk) {
                currentHunk.lines.push(` ${change.line}`);
                currentHunk.oldCount++;
                currentHunk.newCount++;
                const nextChanges = changes.slice(i + 1, i + 1 + 2 * this.contextLines);
                const hasMoreChanges = nextChanges.some((c)=>'equal' !== c.type);
                if (!hasMoreChanges || i === changes.length - 1) {
                    const contextEnd = Math.min(i + this.contextLines, changes.length - 1);
                    for(let j = i + 1; j <= contextEnd; j++)if (changes[j]?.type === 'equal') {
                        currentHunk.lines.push(` ${changes[j].line}`);
                        currentHunk.oldCount++;
                        currentHunk.newCount++;
                    }
                    hunks.push(currentHunk);
                    currentHunk = null;
                }
            }
        }
        return hunks;
    }
}
class AgentSnapshotNormalizer {
    config;
    seenObjects = new WeakMap();
    differ = new SimpleDiffer(3);
    constructor(config){
        this.config = {
            ...DEFAULT_CONFIG,
            ...config,
            fieldsToNormalize: [
                ...DEFAULT_CONFIG.fieldsToNormalize || [],
                ...config?.fieldsToNormalize || []
            ],
            fieldsToIgnore: [
                ...DEFAULT_CONFIG.fieldsToIgnore || [],
                ...config?.fieldsToIgnore || []
            ],
            customNormalizers: [
                ...DEFAULT_CONFIG.customNormalizers || [],
                ...config?.customNormalizers || []
            ]
        };
    }
    normalize(obj, path = '') {
        if ('' === path) this.seenObjects = new WeakMap();
        if (null == obj) return obj;
        if ('object' == typeof obj) {
            if (this.seenObjects.has(obj)) return '[Circular Reference]';
            this.seenObjects.set(obj, true);
        }
        if (Array.isArray(obj)) return obj.map((item, index)=>this.normalize(item, `${path}[${index}]`));
        if ('object' == typeof obj) {
            const result = {};
            for (const [key, value] of Object.entries(obj)){
                const currentPath = path ? `${path}.${key}` : key;
                if (this.shouldIgnoreField(key, currentPath)) continue;
                const normalized = this.normalizeField(key, value, currentPath);
                if (void 0 !== normalized) result[key] = normalized;
                else if ('object' == typeof value && null !== value) result[key] = this.normalize(value, currentPath);
                else result[key] = value;
            }
            return result;
        }
        return obj;
    }
    shouldIgnoreField(key, path) {
        return this.config.fieldsToIgnore?.some((pattern)=>{
            if (pattern instanceof RegExp) return pattern.test(key) || pattern.test(path);
            return key === pattern || path === pattern;
        }) || false;
    }
    normalizeField(key, value, path) {
        if (this.config.customNormalizers) {
            for (const { pattern, normalizer } of this.config.customNormalizers)if (pattern instanceof RegExp && (pattern.test(key) || pattern.test(path)) || key === pattern || path === pattern) return normalizer(value, path);
        }
        if (this.config.fieldsToNormalize) {
            for (const { pattern, replacement, deep = true } of this.config.fieldsToNormalize)if (pattern instanceof RegExp && (pattern.test(key) || pattern.test(path)) || key === pattern || path === pattern) return replacement;
        }
    }
    compare(expected, actual) {
        const normalizedExpected = this.normalize(expected);
        const normalizedActual = this.normalize(actual);
        const expectedString = fast_json_stable_stringify(normalizedExpected);
        const actualString = fast_json_stable_stringify(normalizedActual);
        if (expectedString === actualString) return {
            equal: true,
            diff: null
        };
        const diff = this.differ.diff(JSON.stringify(normalizedExpected, null, 2), JSON.stringify(normalizedActual, null, 2), 'Created Agent Snapshot', 'Runtime Agent State');
        return {
            equal: false,
            diff
        };
    }
    createSnapshotSerializer() {
        return {
            test (val) {
                return 'object' == typeof val && null !== val;
            },
            serialize: (val)=>JSON.stringify(this.normalize(val), null, 2)
        };
    }
}
export { AgentSnapshotNormalizer };
