// @tarko/shared-utils/dist/gui-agent.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
function convertToGUIResponse(actionStr, parsed, result) {
    const normalizedAction = convertToNormalizedAction(parsed, result);
    return {
        success: true,
        action: actionStr,
        normalizedAction,
        observation: void 0
    };
}
function convertToNormalizedAction(parsed, result) {
    const { action_type, action_inputs } = parsed;
    const { startXPercent, startYPercent } = result;
    switch(action_type){
        case 'click':
        case 'left_click':
        case 'left_single':
            {
                const clickAction = {
                    type: 'click',
                    inputs: {
                        startX: startXPercent || 0,
                        startY: startYPercent || 0
                    }
                };
                return clickAction;
            }
        case 'double_click':
        case 'left_double':
            {
                const doubleClickAction = {
                    type: 'double_click',
                    inputs: {
                        startX: startXPercent || 0,
                        startY: startYPercent || 0
                    }
                };
                return doubleClickAction;
            }
        case 'right_click':
        case 'right_single':
            {
                const rightClickAction = {
                    type: 'right_click',
                    inputs: {
                        startX: startXPercent || 0,
                        startY: startYPercent || 0
                    }
                };
                return rightClickAction;
            }
        case 'drag':
            {
                const endBox = action_inputs.end_box;
                let endXPercent = 0;
                let endYPercent = 0;
                if (endBox) try {
                    const coords = JSON.parse(endBox);
                    if (Array.isArray(coords) && coords.length >= 2) {
                        endXPercent = coords[0];
                        endYPercent = coords[1];
                    }
                } catch (e) {
                    console.warn('Failed to parse end_box coordinates:', endBox);
                }
                const dragAction = {
                    type: 'drag',
                    inputs: {
                        startX: startXPercent || 0,
                        startY: startYPercent || 0,
                        endX: endXPercent,
                        endY: endYPercent
                    }
                };
                return dragAction;
            }
        case 'type':
            {
                const typeAction = {
                    type: 'type',
                    inputs: {
                        content: action_inputs.content || ''
                    }
                };
                return typeAction;
            }
        case 'hotkey':
            {
                const hotkeyAction = {
                    type: 'hotkey',
                    inputs: {
                        key: action_inputs.key || action_inputs.hotkey || ''
                    }
                };
                return hotkeyAction;
            }
        case 'scroll':
            {
                const scrollAction = {
                    type: 'scroll',
                    inputs: {
                        startX: startXPercent || 0,
                        startY: startYPercent || 0,
                        direction: action_inputs.direction || 'down'
                    }
                };
                return scrollAction;
            }
        case 'wait':
            {
                const waitAction = {
                    type: 'wait',
                    inputs: {}
                };
                return waitAction;
            }
        case 'navigate':
            {
                const navigateAction = {
                    type: 'navigate',
                    inputs: {
                        url: action_inputs.content || ''
                    }
                };
                return navigateAction;
            }
        case 'navigate_back':
            {
                const navigateBackAction = {
                    type: 'navigate_back',
                    inputs: {}
                };
                return navigateBackAction;
            }
        default:
            {
                console.warn(`Unknown action type: ${action_type}, returning raw type`);
                const fallbackAction = {
                    type: action_type,
                    inputs: {}
                };
                return fallbackAction;
            }
    }
}
function convertToAgentUIAction(action) {
    const { type: action_type, inputs: action_inputs } = action;
    switch(action_type){
        case 'click':
        case 'left_click':
        case 'left_single':
            return {
                type: 'click',
                inputs: {
                    startX: action_inputs.point.normalized.x || 0,
                    startY: action_inputs.point.normalized.y || 0
                }
            };
        case 'double_click':
        case 'left_double':
            return {
                type: 'double_click',
                inputs: {
                    startX: action_inputs.point.normalized.x || 0,
                    startY: action_inputs.point.normalized.y || 0
                }
            };
        case 'right_click':
        case 'right_single':
            return {
                type: 'right_click',
                inputs: {
                    startX: action_inputs.point.normalized.x || 0,
                    startY: action_inputs.point.normalized.y || 0
                }
            };
        case 'drag':
            return {
                type: 'drag',
                inputs: {
                    startX: action_inputs.start.normalized.x || 0,
                    startY: action_inputs.start.normalized.y || 0,
                    endX: action_inputs.end.normalized.x || 0,
                    endY: action_inputs.end.normalized.y || 0
                }
            };
        case 'scroll':
            var _action_inputs_point_normalized, _action_inputs_point, _action_inputs_point_normalized1, _action_inputs_point1;
            return {
                type: 'scroll',
                inputs: {
                    startX: (null == (_action_inputs_point = action_inputs.point) ? void 0 : null == (_action_inputs_point_normalized = _action_inputs_point.normalized) ? void 0 : _action_inputs_point_normalized.x) || 0,
                    startY: (null == (_action_inputs_point1 = action_inputs.point) ? void 0 : null == (_action_inputs_point_normalized1 = _action_inputs_point1.normalized) ? void 0 : _action_inputs_point_normalized1.y) || 0,
                    direction: action_inputs.direction || 'down'
                }
            };
        case 'navigate':
            return {
                type: 'navigate',
                inputs: {
                    url: action_inputs.url || action_inputs.content || ''
                }
            };
        default:
            return action;
    }
}
function createErrorAction() {
    const errorAction = {
        type: 'wait',
        inputs: {}
    };
    return errorAction;
}
function createGUIErrorResponse(actionStr, error) {
    return {
        success: false,
        action: actionStr,
        normalizedAction: createErrorAction(),
        error: error instanceof Error ? error.message : String(error)
    };
}
export { convertToAgentUIAction, convertToGUIResponse, convertToNormalizedAction, createErrorAction, createGUIErrorResponse };
