// @tarko/context-engineer/dist/node/context-reference-processor.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import fs from "fs";
import path from "path";
import { WorkspacePack } from "./workspace-pack.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ContextReferenceProcessor {
    async processContextualReferences(query, workspacePath) {
        if ('string' != typeof query) return null;
        const contextualReferencePattern = /@(file|dir):([^\s]+)/g;
        const matches = Array.from(query.matchAll(contextualReferencePattern));
        if (0 === matches.length) return null;
        const fileReferences = [];
        const dirReferences = [];
        for (const match of matches){
            const [, type, relativePath] = match;
            if ('file' === type) fileReferences.push(relativePath);
            else if ('dir' === type) dirReferences.push(relativePath);
        }
        const expandedContents = [];
        for (const fileRef of fileReferences)try {
            const absolutePath = path.resolve(workspacePath, fileRef);
            const normalizedWorkspace = path.resolve(workspacePath);
            const normalizedTarget = path.resolve(absolutePath);
            if (!normalizedTarget.startsWith(normalizedWorkspace)) {
                console.warn(`File reference outside workspace: ${fileRef}`);
                expandedContents.push(`<file path="${fileRef}">\nError: File reference outside workspace\n</file>`);
                continue;
            }
            if (!fs.existsSync(absolutePath)) {
                console.warn(`File reference not found: ${fileRef}`);
                expandedContents.push(`<file path="${fileRef}">\nError: File not found\n</file>`);
                continue;
            }
            const stats = fs.statSync(absolutePath);
            if (stats.isFile()) try {
                const fileContent = fs.readFileSync(absolutePath, 'utf8');
                expandedContents.push(`<file path="${fileRef}">\n${fileContent}\n</file>`);
            } catch (error) {
                console.error(`Failed to read file ${fileRef}:`, error);
                expandedContents.push(`<file path="${fileRef}">\nError: Failed to read file\n</file>`);
            }
        } catch (error) {
            console.error(`Failed to process file reference ${fileRef}:`, error);
            expandedContents.push(`<file path="${fileRef}">\nError: Failed to process file reference\n</file>`);
        }
        if (dirReferences.length > 0) try {
            const absoluteDirPaths = dirReferences.map((dirRef)=>{
                const absolutePath = path.resolve(workspacePath, dirRef);
                const normalizedWorkspace = path.resolve(workspacePath);
                const normalizedTarget = path.resolve(absolutePath);
                if (!normalizedTarget.startsWith(normalizedWorkspace)) {
                    console.warn(`Directory reference outside workspace: ${dirRef}`);
                    return null;
                }
                return {
                    relativePath: dirRef,
                    absolutePath
                };
            }).filter((p)=>null !== p);
            if (absoluteDirPaths.length > 0) {
                const packResult = await this.workspacePack.packPaths(absoluteDirPaths.map((p)=>p.absolutePath), workspacePath);
                for (const { relativePath } of absoluteDirPaths)expandedContents.push(`<directory path="${relativePath}">\n${packResult.packedContent}\n</directory>`);
                console.log('Workspace packing completed:', {
                    paths: packResult.processedPaths.length,
                    files: packResult.stats.totalFiles,
                    totalSize: packResult.stats.totalSize,
                    errors: packResult.stats.errorCount
                });
            }
        } catch (error) {
            console.error('Failed to pack workspace paths:', error);
            for (const dirRef of dirReferences)expandedContents.push(`<directory path="${dirRef}">\nError: Failed to pack directory\n</directory>`);
        }
        if (expandedContents.length > 0) return expandedContents.join('\n\n');
        return null;
    }
    constructor(options = {}){
        _define_property(this, "workspacePack", void 0);
        this.workspacePack = new WorkspacePack({
            maxFileSize: options.maxFileSize ?? 2097152,
            ignoreExtensions: options.ignoreExtensions ?? [
                '.jpg',
                '.jpeg',
                '.png',
                '.gif',
                '.webp',
                '.svg',
                '.pdf',
                '.zip',
                '.tar',
                '.gz',
                '.exe',
                '.dll'
            ],
            ignoreDirs: options.ignoreDirs ?? [
                'node_modules',
                '.git',
                '.next',
                'dist',
                'build',
                'coverage',
                '.vscode',
                '.idea'
            ],
            maxDepth: options.maxDepth ?? 8
        });
    }
}
export { ContextReferenceProcessor };
