// @tarko/context-engineer/dist/node/image-processor.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ImageCompressor, formatBytes } from "@tarko/shared-media-utils";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ImageProcessor {
    async compressImagesInQuery(query) {
        try {
            if ('string' == typeof query) return query;
            if (Array.isArray(query)) {
                const compressedQuery = await Promise.all(query.map(async (part)=>{
                    var _part_image_url;
                    if ('image_url' === part.type && (null == (_part_image_url = part.image_url) ? void 0 : _part_image_url.url)) return await this.compressImageUrl(part);
                    return part;
                }));
                return compressedQuery;
            }
            return query;
        } catch (error) {
            console.error('Error compressing images in query:', error);
            return query;
        }
    }
    async compressImageUrl(imagePart) {
        try {
            const imageUrl = imagePart.image_url.url;
            if (!imageUrl.startsWith('data:image/')) return imagePart;
            const base64Data = imageUrl.replace(/^data:image\/\w+;base64,/, '');
            const originalBuffer = Buffer.from(base64Data, 'base64');
            const originalSize = originalBuffer.length;
            const compressedBuffer = await this.compressor.compressToBuffer(originalBuffer);
            const compressedSize = compressedBuffer.length;
            const compressedBase64 = `data:image/webp;base64,${compressedBuffer.toString('base64')}`;
            const compressionRatio = originalSize / compressedSize;
            const compressionPercentage = ((1 - compressedSize / originalSize) * 100).toFixed(2);
            console.log('Image compression stats:', {
                original: formatBytes(originalSize),
                compressed: formatBytes(compressedSize),
                ratio: `${compressionRatio.toFixed(2)}x (${compressionPercentage}% smaller)`,
                format: 'webp',
                quality: 80
            });
            return {
                ...imagePart,
                image_url: {
                    url: compressedBase64
                }
            };
        } catch (error) {
            console.error('Error compressing individual image:', error);
            return imagePart;
        }
    }
    constructor(options = {}){
        _define_property(this, "compressor", void 0);
        this.compressor = new ImageCompressor({
            quality: options.quality ?? 5,
            format: options.format ?? 'webp'
        });
    }
}
export { ImageProcessor };
