// @tarko/context-engineer/dist/node/workspace-pack.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import promises from "fs/promises";
import fs from "fs";
import path from "path";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const DEFAULT_OPTIONS = {
    maxFileSize: 1048576,
    ignoreExtensions: [
        '.jpg',
        '.jpeg',
        '.png',
        '.gif',
        '.webp',
        '.svg',
        '.pdf',
        '.zip',
        '.tar',
        '.gz'
    ],
    ignoreDirs: [
        'node_modules',
        '.git',
        '.next',
        'dist',
        'build',
        'coverage'
    ],
    maxDepth: 10
};
class WorkspacePack {
    async packPaths(paths, cwd) {
        const validatedPaths = this.validateAndDeduplicatePaths(paths);
        if (0 === validatedPaths.length) return {
            processedPaths: [],
            files: [],
            packedContent: '',
            stats: {
                totalFiles: 0,
                totalSize: 0,
                errorCount: 0
            }
        };
        const fileCollectionPromises = validatedPaths.map((targetPath)=>this.collectFilesFromPath(targetPath, 0));
        const fileArrays = await Promise.all(fileCollectionPromises);
        const allFilePaths = fileArrays.flat();
        const uniqueFilePaths = [
            ...new Set(allFilePaths)
        ];
        const fileReadPromises = uniqueFilePaths.map((filePath)=>this.readFileInfo(filePath));
        const files = await Promise.all(fileReadPromises);
        const stats = {
            totalFiles: files.length,
            totalSize: files.reduce((sum, file)=>sum + file.size, 0),
            errorCount: files.filter((file)=>file.hasError).length
        };
        const packedContent = this.formatForLLM(validatedPaths, files, cwd);
        return {
            processedPaths: validatedPaths,
            files,
            packedContent,
            stats
        };
    }
    validateAndDeduplicatePaths(paths) {
        const validPaths = [];
        const seenPaths = new Set();
        for (const targetPath of paths)try {
            const absolutePath = path.resolve(targetPath);
            if (seenPaths.has(absolutePath)) continue;
            seenPaths.add(absolutePath);
            if (!fs.existsSync(absolutePath)) {
                console.warn(`Path not found: ${targetPath}`);
                continue;
            }
            validPaths.push(absolutePath);
        } catch (error) {
            console.warn(`Failed to validate path ${targetPath}:`, error);
        }
        return validPaths;
    }
    async collectFilesFromPath(targetPath, depth) {
        try {
            const stat = await promises.stat(targetPath);
            if (stat.isFile()) {
                const ext = path.extname(targetPath).toLowerCase();
                if (!this.options.ignoreExtensions.includes(ext)) return [
                    targetPath
                ];
                return [];
            }
            if (stat.isDirectory()) return await this.collectFilesRecursively(targetPath, depth);
            return [];
        } catch (error) {
            console.warn(`Failed to access path ${targetPath}:`, error);
            return [];
        }
    }
    async collectFilesRecursively(directoryPath, depth) {
        if (depth > this.options.maxDepth) return [];
        try {
            const entries = await promises.readdir(directoryPath, {
                withFileTypes: true
            });
            const filePaths = [];
            for (const entry of entries){
                const fullPath = path.join(directoryPath, entry.name);
                if (entry.isFile()) {
                    const ext = path.extname(entry.name).toLowerCase();
                    if (!this.options.ignoreExtensions.includes(ext)) filePaths.push(fullPath);
                } else if (entry.isDirectory()) {
                    if (!this.options.ignoreDirs.includes(entry.name)) {
                        const subFiles = await this.collectFilesRecursively(fullPath, depth + 1);
                        filePaths.push(...subFiles);
                    }
                }
            }
            return filePaths;
        } catch (error) {
            console.warn(`Failed to read directory ${directoryPath}:`, error);
            return [];
        }
    }
    async readFileInfo(filePath) {
        try {
            const stats = await promises.stat(filePath);
            if (stats.size > this.options.maxFileSize) return {
                absolutePath: filePath,
                content: `[File too large: ${this.formatFileSize(stats.size)}]`,
                hasError: true,
                size: stats.size
            };
            const content = await promises.readFile(filePath, 'utf8');
            return {
                absolutePath: filePath,
                content,
                hasError: false,
                size: stats.size
            };
        } catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            return {
                absolutePath: filePath,
                content: `[Error reading file: ${errorMessage}]`,
                hasError: true,
                size: 0
            };
        }
    }
    formatForLLM(processedPaths, files, cwd) {
        const sections = [];
        const filesByProcessedPath = new Map();
        for (const processedPath of processedPaths){
            const pathFiles = files.filter((file)=>file.absolutePath.startsWith(processedPath));
            filesByProcessedPath.set(processedPath, pathFiles);
        }
        for (const processedPath of processedPaths){
            const pathFiles = filesByProcessedPath.get(processedPath) || [];
            if (0 === pathFiles.length) continue;
            const sortedFiles = pathFiles.sort((a, b)=>a.absolutePath.localeCompare(b.absolutePath));
            const displayPath = cwd ? path.relative(cwd, processedPath) || processedPath : processedPath;
            sections.push(`<directory path="${displayPath}">`);
            for (const file of sortedFiles){
                const fileRelativePath = path.relative(processedPath, file.absolutePath);
                const displayFilePath = fileRelativePath || path.basename(file.absolutePath);
                sections.push(`  <file path="${displayFilePath}">`);
                const indentedContent = file.content.split('\n').map((line)=>`    ${line}`).join('\n');
                sections.push(indentedContent);
                sections.push("  </file>");
            }
            sections.push("</directory>");
        }
        return sections.join('\n');
    }
    formatFileSize(bytes) {
        if (0 === bytes) return '0 B';
        const k = 1024;
        const sizes = [
            'B',
            'KB',
            'MB',
            'GB'
        ];
        const i = Math.floor(Math.log(bytes) / Math.log(k));
        return `${parseFloat((bytes / Math.pow(k, i)).toFixed(1))} ${sizes[i]}`;
    }
    constructor(options = {}){
        _define_property(this, "options", void 0);
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options
        };
    }
}
export { WorkspacePack };
