// @omni-tars/core/dist/utils/streamingParser/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { extractAnswer, extractThink } from "./extractors.mjs";
import { extractCodeEnv } from "./extractCode.mjs";
function createInitState() {
    return {
        contentBuffer: '',
        accumulatedAnswerBuffer: '',
        toolCalls: [],
        reasoningBuffer: '',
        finishReason: null,
        thinkBuffer: '',
        answerBuffer: '',
        currentTag: null,
        insideThink: false,
        thinkParseCompleted: false,
        insideAnswer: false,
        insideCodeEnv: false,
        codeEnvBuffer: '',
        currentToolName: '',
        currentParameters: {},
        parameterBuffer: '',
        currentParameterName: '',
        insideParameter: false,
        insideFunction: false,
        functionBuffer: '',
        currentToolCallId: '',
        insideMcp: false,
        insideGUI: false
    };
}
function processStreamingChunk(chunk, state) {
    var _chunk_choices_, _chunk_choices_1;
    const delta = null == (_chunk_choices_ = chunk.choices[0]) ? void 0 : _chunk_choices_.delta;
    let content = '';
    let reasoningContent = '';
    let hasToolCallUpdate = false;
    let streamingToolCallUpdates = [];
    if (null == (_chunk_choices_1 = chunk.choices[0]) ? void 0 : _chunk_choices_1.finish_reason) state.finishReason = chunk.choices[0].finish_reason;
    if (null == delta ? void 0 : delta.content) {
        state.contentBuffer += delta.content;
        reasoningContent = extractThink(delta.content, state);
        content = extractAnswer(delta.content, state);
        const parsed = extractCodeEnv(delta.content, state);
        hasToolCallUpdate = parsed.hasToolCallUpdate;
        streamingToolCallUpdates = parsed.streamingToolCallUpdates || [];
    }
    return {
        content,
        reasoningContent,
        hasToolCallUpdate,
        toolCalls: state.toolCalls,
        streamingToolCallUpdates
    };
}
export { createInitState, processStreamingChunk };
