// @agent-infra/puppeteer-enhance/dist/hotkey/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import delay from "delay";
import { ConsoleLogger } from "@agent-infra/logger";
import { KEY_ABBR_TO_STANDARD_MAP, KEY_LOW_TO_STANDARD_MAP, MAC_SYSTEM_HOTKEY_MAP } from "./key-map.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class Hotkey {
    formatHotkey(hotkey) {
        const lowerCaseHotkey = hotkey.toLowerCase();
        const keys = lowerCaseHotkey.split(/[\s+]+/);
        this.logger.info('lowerCase keys', keys);
        const formattedKeys = keys.map((key)=>{
            if (KEY_ABBR_TO_STANDARD_MAP[key]) return KEY_ABBR_TO_STANDARD_MAP[key];
            if (KEY_LOW_TO_STANDARD_MAP[key]) return KEY_LOW_TO_STANDARD_MAP[key];
            throw new Error('Unsupported key: ' + key);
        });
        this.logger.info('format keys', formattedKeys);
        return formattedKeys;
    }
    async macOSCDPHotKey(page, keys, options) {
        const hotkey = keys.map((key)=>{
            if ('Control' === key || 'Meta' === key) return 'CorM';
            return key;
        }).join('+');
        const command = MAC_SYSTEM_HOTKEY_MAP.get(hotkey);
        if (command) {
            await page.keyboard.down(command.key, {
                commands: [
                    command.commands
                ]
            });
            await delay(options.delay);
            await page.keyboard.up(command.key);
            return true;
        }
        return false;
    }
    async press(page, hotkey, options = {
        delay: 100
    }) {
        const formattedHotkey = this.formatHotkey(hotkey);
        if ('macOS' === this.osName && 'Chrome' === this.browserName) {
            const success = await this.macOSCDPHotKey(page, formattedHotkey, options);
            if (success) return;
        }
        for (const key of formattedHotkey)await page.keyboard.down(key);
        await delay(options.delay);
        for (const key of formattedHotkey.reverse())await page.keyboard.up(key);
    }
    async down(page, hotkey, options = {
        delay: 100
    }) {
        const formattedHotkey = this.formatHotkey(hotkey);
        if ('macOS' === this.osName && 'Chrome' === this.browserName) {
            const success = await this.macOSCDPHotKey(page, formattedHotkey, options);
            if (success) return;
        }
        for (const key of formattedHotkey)await page.keyboard.down(key);
    }
    async up(page, hotkey, options = {
        delay: 100
    }) {
        const formattedHotkey = this.formatHotkey(hotkey);
        for (const key of formattedHotkey.reverse())await page.keyboard.up(key);
    }
    constructor(envInfo, logger){
        _define_property(this, "osName", void 0);
        _define_property(this, "browserName", void 0);
        _define_property(this, "logger", void 0);
        this.osName = envInfo.osName;
        this.browserName = envInfo.browserName;
        this.logger = logger ?? new ConsoleLogger('Hotkey:');
    }
}
export { Hotkey };
