// @tarko/agent-ui-builder/dist/builder.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import fs from "fs";
import path from "path";
import { getStaticPath } from "./static-path.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class AgentUIBuilder {
    async dump(filePath) {
        const { events, sessionInfo, serverInfo, uiConfig, staticPath } = this.input;
        try {
            let htmlContent = await this.getHtmlContent(staticPath, uiConfig);
            const safeEventJson = AgentUIBuilder.safeJsonStringify(events);
            const safeSessionInfoJson = AgentUIBuilder.safeJsonStringify(sessionInfo);
            const safeVersionJson = serverInfo ? AgentUIBuilder.safeJsonStringify(serverInfo) : null;
            const safeUIConfigJson = uiConfig ? AgentUIBuilder.safeJsonStringify(uiConfig) : null;
            const scriptTag = `<script>
        window.AGENT_REPLAY_MODE = true;
        window.AGENT_SESSION_DATA = ${safeSessionInfoJson};
        window.AGENT_EVENT_STREAM = ${safeEventJson};${safeVersionJson ? `\n        window.AGENT_VERSION_INFO = ${safeVersionJson};` : ''}${safeUIConfigJson ? `\n        window.AGENT_WEB_UI_CONFIG = ${safeUIConfigJson};` : ''}
      </script>
      <script>
        // Add a fallback mechanism for when routes don't match in shared HTML files
        window.addEventListener('DOMContentLoaded', function() {
          // Give React time to attempt normal routing
          setTimeout(function() {
            const root = document.getElementById('root');
            if (root && (!root.children || root.children.length === 0)) {
              console.log('[ReplayMode] No content rendered, applying fallback');
              // Try to force the app to re-render if no content is displayed
              window.dispatchEvent(new Event('resize'));
            }
          }, 1000);
        });
      </script>`;
            htmlContent = htmlContent.replace('</head>', `${scriptTag}\n</head>`);
            if (filePath) {
                const dir = path.dirname(filePath);
                if (!fs.existsSync(dir)) fs.mkdirSync(dir, {
                    recursive: true
                });
                fs.writeFileSync(filePath, htmlContent, 'utf8');
            }
            return htmlContent;
        } catch (error) {
            console.error('Failed to generate HTML:', error);
            throw new Error(`Failed to generate HTML: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async getHtmlContent(staticPath, webui) {
        if ((null == webui ? void 0 : webui.type) === 'remote' && (null == webui ? void 0 : webui.remoteUrl)) {
            const url = webui.remoteUrl;
            const resp = await fetch(url, {
                method: 'GET',
                headers: {
                    Accept: "text/html"
                }
            });
            return await resp.text();
        }
        staticPath = staticPath || getStaticPath();
        const indexPath = path.join(staticPath, 'index.html');
        if (!fs.existsSync(indexPath)) throw new Error(`Static web UI not found at: ${indexPath}`);
        return fs.readFileSync(indexPath, 'utf8');
    }
    async upload(html, shareProviderUrl, options) {
        var _sessionInfo_metadata, _sessionInfo_metadata1;
        const sessionId = this.input.sessionInfo.id;
        const formData = new FormData();
        const fileName = `agent-replay-${sessionId}-${Date.now()}.html`;
        const file = new File([
            html
        ], fileName, {
            type: 'text/html'
        });
        formData.append('file', file);
        formData.append('sessionId', sessionId);
        formData.append('type', 'html');
        if (null == options ? void 0 : options.slug) formData.append('slug', options.slug);
        if (null == options ? void 0 : options.query) formData.append('query', options.query);
        const sessionInfo = this.input.sessionInfo;
        if (null == (_sessionInfo_metadata = sessionInfo.metadata) ? void 0 : _sessionInfo_metadata.name) formData.append('name', sessionInfo.metadata.name);
        if ((null == (_sessionInfo_metadata1 = sessionInfo.metadata) ? void 0 : _sessionInfo_metadata1.tags) && sessionInfo.metadata.tags.length > 0) {
            const tagsJson = JSON.stringify(sessionInfo.metadata.tags);
            formData.append('tags', tagsJson);
        }
        try {
            const response = await fetch(shareProviderUrl, {
                method: 'POST',
                body: formData
            });
            if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
            const responseData = await response.json();
            if (responseData && responseData.url) {
                const url = new URL(responseData.url);
                url.searchParams.set('replay', '1');
                return url.toString();
            }
            throw new Error('Invalid response from share provider');
        } catch (error) {
            console.error('Failed to upload to share provider:', error);
            throw error;
        }
    }
    static safeJsonStringify(data) {
        let jsonString = JSON.stringify(data);
        jsonString = jsonString.replace(/</g, '\\u003C');
        jsonString = jsonString.replace(/>/g, '\\u003E');
        jsonString = jsonString.replace(/\//g, '\\/');
        return jsonString;
    }
    constructor(input){
        _define_property(this, "input", void 0);
        this.input = input;
    }
}
export { AgentUIBuilder };
