// @omni-tars/mcp-agent/dist/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ComposableAgent } from "@omni-tars/core";
import { McpAgentPlugin } from "./McpAgentPlugin.mjs";
import { McpManager } from "./tools/mcp.mjs";
import { McpToolCallEngine } from "./McpToolCallEngine.mjs";
import { McpToolCallEngineProvider } from "./McpToolCallEngineProvider.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const mcpPluginBuilder = (option)=>new McpAgentPlugin({
        mcpServers: [
            {
                type: 'streamable-http',
                name: McpManager.McpClientType.Tavily,
                description: 'tavily search tool',
                url: `https://mcp.tavily.com/mcp/?tavilyApiKey=${option.tavilyApiKey}`,
                timeout: 60,
                enable: !!option.tavilyApiKey
            },
            {
                type: 'streamable-http',
                name: McpManager.McpClientType.Google,
                description: 'google search tool',
                url: option.googleMcpUrl,
                headers: {
                    'x-serper-api-key': option.googleApiKey
                },
                enable: true
            },
            {
                type: 'streamable-http',
                name: McpManager.McpClientType.LinkReader,
                description: 'Crawl, parse and summarize for web pages',
                url: option.linkReaderMcpUrl,
                headers: {
                    'x-text-browser-ak': option.linkReaderAK,
                    'x-text-browser-traffic-id': 'edge_agent_research',
                    'x-text-browser-traffic-group': 'Seed_Edge'
                },
                timeout: 60,
                enable: !!option.linkReaderMcpUrl
            }
        ]
    });
class McpAgent extends ComposableAgent {
    constructor(options){
        const { tavilyApiKey, googleApiKey, googleMcpUrl, linkReaderMcpUrl, ...restOptions } = options;
        super({
            ...restOptions,
            plugins: [
                mcpPluginBuilder({
                    tavilyApiKey,
                    googleMcpUrl,
                    googleApiKey,
                    linkReaderMcpUrl
                })
            ],
            toolCallEngine: McpToolCallEngine
        });
    }
}
_define_property(McpAgent, "label", void 0);
export { McpAgentPlugin, McpToolCallEngineProvider, McpAgent as default, mcpPluginBuilder };
