// @omni-tars/mcp-agent/dist/McpToolCallEngine.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ToolCallEngine, getLogger } from "@tarko/agent";
import { SYSTEM_PROMPT_GROUP, createT5InitState, parseMcpContent, processT5StreamingChunk } from "@omni-tars/core";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class McpToolCallEngine extends ToolCallEngine {
    preparePrompt(instructions, tools) {
        return SYSTEM_PROMPT_GROUP;
    }
    prepareRequest(context) {
        return {
            model: context.model,
            messages: context.messages,
            temperature: context.temperature || 1.0,
            top_p: context.top_p,
            stream: true
        };
    }
    initStreamProcessingState() {
        return createT5InitState();
    }
    processStreamingChunk(chunk, state) {
        return processT5StreamingChunk(chunk, state);
    }
    finalizeStreamProcessing(state) {
        this.logger.info('finalizeStreamProcessing state \n', state);
        let tools = state.toolCalls;
        const fullContent = state.contentBuffer;
        if (state.contentBuffer.includes('mcp_env')) {
            const extracted = parseMcpContent(fullContent);
            this.logger.info('extracted', JSON.stringify(extracted, null, 2));
            tools = extracted.tools;
        }
        return {
            content: state.accumulatedChatContentBuffer ?? fullContent,
            rawContent: fullContent,
            reasoningContent: state.reasoningBuffer ?? '',
            toolCalls: tools,
            finishReason: (tools || []).length > 0 ? 'tool_calls' : 'stop'
        };
    }
    generateToolCallId() {
        return `call_${Date.now()}_${Math.random().toString(36).substring(2, 11)}`;
    }
    parseContent(content) {
        let think = '';
        let answer = '';
        let tools = [];
        try {
            const thinkMatch = content.match(/<think>(.*?)<\/think>/s);
            if (thinkMatch) think = thinkMatch[1].trim();
            const answerMatch = content.match(/<answer>(.*?)<\/answer>/s);
            if (answerMatch) answer = answerMatch[1].trim();
            const fcResponseMatch = content.match(/<\|FCResponseBegin\|>(.*?)<\/answer>/s);
            if (fcResponseMatch) answer = fcResponseMatch[1].trim();
            const functionCallMatch = content.match(/<\|FunctionCallBegin\|>(.*?)<\|FunctionCallEnd\|>/s);
            if (functionCallMatch) {
                const functionCallContent = functionCallMatch[1];
                const jsonMatch = functionCallContent.match(/\[.*\]/s);
                if (jsonMatch) try {
                    const toolCallsData = JSON.parse(jsonMatch[0]);
                    tools = toolCallsData.map((toolCall)=>({
                            id: this.generateToolCallId(),
                            type: 'function',
                            function: {
                                name: toolCall.name,
                                arguments: JSON.stringify(toolCall.parameters || {})
                            }
                        }));
                } catch (parseError) {
                    this.logger.warn('Failed to parse tool calls JSON:', parseError);
                }
            } else {
                const newFormatMatch = content.match(/<\|FunctionCallBegin\|>(.*?)<\/think>\s*(\[.*?\])/s);
                if (newFormatMatch) {
                    if (!think) think = newFormatMatch[1].trim();
                    try {
                        const toolCallsData = JSON.parse(newFormatMatch[2]);
                        tools = toolCallsData.map((toolCall)=>({
                                id: this.generateToolCallId(),
                                type: 'function',
                                function: {
                                    name: toolCall.name,
                                    arguments: JSON.stringify(toolCall.parameters || {})
                                }
                            }));
                    } catch (parseError) {
                        this.logger.warn('Failed to parse new format tool calls JSON:', parseError);
                    }
                }
            }
            const noBeginMatch = content.match(/(\[.*?\])<\|FunctionCallEnd\|>/s);
            if (noBeginMatch) try {
                const toolCallsData = JSON.parse(noBeginMatch[1]);
                tools = toolCallsData.map((toolCall)=>({
                        id: this.generateToolCallId(),
                        type: 'function',
                        function: {
                            name: toolCall.name,
                            arguments: JSON.stringify(toolCall.parameters || {})
                        }
                    }));
            } catch (parseError) {
                this.logger.warn('Failed to parse no-begin format tool calls JSON:', parseError);
            }
            if (!answer && !tools.length && content.trim()) {
                let remainingContent = content;
                if (thinkMatch) remainingContent = content.replace(/<think>.*?<\/think>/s, '').trim();
                if (remainingContent) answer = remainingContent;
            }
            if (tools.length > 0) answer = '';
        } catch (error) {
            this.logger.error('Error parsing content:', error);
            answer = content;
        }
        return {
            think,
            tools,
            answer
        };
    }
    buildHistoricalAssistantMessage(currentLoopAssistantEvent) {
        return {
            role: 'assistant',
            content: currentLoopAssistantEvent.rawContent || currentLoopAssistantEvent.content
        };
    }
    buildHistoricalToolCallResultMessages(toolCallResults) {
        return toolCallResults.map((result)=>{
            const textContent = result.content.filter((part)=>'text' === part.type).map((part)=>part.text).join('');
            return {
                role: 'user',
                content: `Tool "${result.toolName}" result:\n${textContent}`
            };
        });
    }
    constructor(...args){
        super(...args), _define_property(this, "logger", getLogger('McpToolCallEngine'));
    }
}
export { McpToolCallEngine };
