// @omni-tars/core/dist/ComposableAgent.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Agent, getLogger } from "@tarko/agent";
import { AgentComposer } from "./AgentComposer.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ComposableAgent extends Agent {
    async initialize() {
        await this.composer.initialize();
        const tools = this.composer.getAllTools();
        for (const tool of tools)this.registerTool(tool);
        await super.initialize();
    }
    async onLLMRequest(id, payload) {
        await this.composer.executeOnLLMRequest(id, payload);
    }
    async onLLMResponse(id, payload) {
        await this.composer.executeOnLLMResponse(id, payload);
    }
    async onEachAgentLoopStart() {
        await this.composer.executeOnEachAgentLoopStart();
    }
    async onEachAgentLoopEnd() {
        await this.composer.executeOnEachAgentLoopEnd();
    }
    async onAgentLoopEnd(id) {
        await this.composer.executeOnAgentLoopEnd();
        await super.onAgentLoopEnd(id);
    }
    async onAfterToolCall(id, toolCall, result) {
        return await this.composer.executeOnAfterToolCall(id, toolCall, result);
    }
    constructor(options){
        const { plugins, ...optionsWithoutPlugins } = options;
        const composer = new AgentComposer({
            plugins
        });
        super({
            ...optionsWithoutPlugins
        }), _define_property(this, "composer", void 0);
        this.composer = composer;
        this.composer.setAgent(this);
        this.logger = getLogger('ComposableAgent');
        this.logger.info('load plugins success: ', plugins.map((p)=>p.name));
    }
}
export { ComposableAgent };
