// @omni-tars/core/dist/ComposableToolCallEngine.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ToolCallEngine, getLogger } from "@tarko/agent";
import { assert } from "console";
import { bypass_native_thinking } from "./environments/prompt_t5.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ComposableToolCallEngine extends ToolCallEngine {
    selectEngine(context) {
        for (const engineProvider of this.engines)if (!engineProvider.canHandle || engineProvider.canHandle(context)) return engineProvider.getEngine();
        this.logger.debug(`No engine matched, using default engine: ${this.defaultEngineProvider.name}`);
        return this.defaultEngineProvider.getEngine();
    }
    preparePrompt(instructions, tools) {
        return this.selectEngine({
            tools
        }).preparePrompt(instructions, tools);
    }
    prepareRequest(context) {
        if (bypass_native_thinking) context.messages.push({
            role: 'assistant',
            content: ''
        });
        return this.selectEngine({
            tools: context.tools || [],
            messageHistory: context.messages
        }).prepareRequest(context);
    }
    initStreamProcessingState() {
        try {
            return this.defaultEngineProvider.getEngine().initStreamProcessingState();
        } catch (e) {
            this.logger.error('initStreamProcessingState err: ', e);
            return {
                contentBuffer: '',
                toolCalls: [],
                reasoningBuffer: '',
                finishReason: null
            };
        }
    }
    processStreamingChunk(chunk, state) {
        return this.defaultEngineProvider.getEngine().processStreamingChunk(chunk, state);
    }
    finalizeStreamProcessing(state) {
        return this.selectEngine({
            toolCalls: state.toolCalls,
            latestAssistantMessage: state.contentBuffer
        }).finalizeStreamProcessing(state);
    }
    buildHistoricalAssistantMessage(currentLoopAssistantEvent) {
        return this.defaultEngineProvider.getEngine().buildHistoricalAssistantMessage(currentLoopAssistantEvent);
    }
    buildHistoricalToolCallResultMessages(toolCallResults) {
        return this.defaultEngineProvider.getEngine().buildHistoricalToolCallResultMessages(toolCallResults);
    }
    getEngineInfo() {
        return this.engines.map((engine)=>({
                name: engine.name,
                priority: engine.priority,
                description: engine.description
            }));
    }
    getActiveEngineName() {
        if (!this.activeEngine) return;
        for (const engineProvider of this.engines)if (engineProvider.getEngine() === this.activeEngine) return engineProvider.name;
        return 'unknown';
    }
    constructor(config){
        super(), _define_property(this, "logger", getLogger('ComposableToolCallEngine')), _define_property(this, "engines", void 0), _define_property(this, "defaultEngineProvider", void 0), _define_property(this, "activeEngine", void 0);
        this.engines = [
            ...config.engines
        ].sort((a, b)=>b.priority - a.priority);
        this.defaultEngineProvider = config.defaultEngine || this.engines[0];
        assert(this.engines.length > 0 || !!this.defaultEngineProvider, 'No tool call engines available');
        this.logger.info(`Initialized ComposableToolCallEngine with ${this.engines.length} engines`, {
            engines: this.engines.map((e)=>`${e.name}(${e.priority})`)
        });
    }
}
export { ComposableToolCallEngine };
