// @omni-tars/core/dist/ComposableToolCallEngineFactory.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { ToolCallEngine } from "@tarko/agent";
import { ComposableToolCallEngine } from "./ComposableToolCallEngine.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ComposableToolCallEngineFactory extends ToolCallEngine {
    preparePrompt(instructions, tools) {
        return this.composableEngine.preparePrompt(instructions, tools);
    }
    prepareRequest(context) {
        return this.composableEngine.prepareRequest(context);
    }
    processStreamingChunk(chunk, state) {
        return this.composableEngine.processStreamingChunk(chunk, state);
    }
    finalizeStreamProcessing(state) {
        return this.composableEngine.finalizeStreamProcessing(state);
    }
    initStreamProcessingState() {
        return this.composableEngine.initStreamProcessingState();
    }
    buildHistoricalAssistantMessage(currentLoopAssistantEvent) {
        return this.composableEngine.buildHistoricalAssistantMessage(currentLoopAssistantEvent);
    }
    buildHistoricalToolCallResultMessages(toolCallResults) {
        return this.composableEngine.buildHistoricalToolCallResultMessages(toolCallResults);
    }
    getComposableEngine() {
        return this.composableEngine;
    }
    constructor(config){
        super(), _define_property(this, "composableEngine", void 0);
        this.composableEngine = new ComposableToolCallEngine(config);
    }
}
function createComposableToolCallEngineFactory(config) {
    return class extends ComposableToolCallEngineFactory {
        constructor(){
            super(config);
        }
    };
}
export { ComposableToolCallEngineFactory, createComposableToolCallEngineFactory };
