// @omni-tars/code-agent/dist/tools/ExcuteBash.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Tool, z } from "@tarko/agent";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ExcuteBashProvider {
    getTool() {
        return new Tool({
            id: 'execute_bash',
            description: '',
            parameters: z.object({
                command: z.string().describe('Execute a bash command in the terminal.')
            }),
            function: async ({ command })=>(await this.client.shellExec({
                    command
                }, {
                    timeout: 120000
                })).data
        });
    }
    constructor(client){
        _define_property(this, "client", void 0);
        this.client = client;
    }
}
export { ExcuteBashProvider };
