// @omni-tars/code-agent/dist/tools/JupyterCI.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { Tool, z } from "@tarko/agent";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class JupyterCIProvider {
    getTool() {
        return new Tool({
            id: 'JupyterCI',
            description: '',
            parameters: z.object({
                code: z.string().describe('code'),
                timeout: z.number().describe('timeout in seconds').optional()
            }),
            function: async ({ code, timeout })=>{
                var _resp_data;
                const resp = await this.client.jupyterExecute({
                    code,
                    timeout,
                    session_id: this.sessionId,
                    kernel_name: 'python3'
                });
                if ((null == (_resp_data = resp.data) ? void 0 : _resp_data.session_id) !== this.sessionId) {
                    var _resp_data1;
                    this.sessionId = (null == (_resp_data1 = resp.data) ? void 0 : _resp_data1.session_id) ?? '';
                }
                return resp.data;
            }
        });
    }
    constructor(client){
        _define_property(this, "client", void 0);
        _define_property(this, "sessionId", '');
        this.client = client;
    }
}
export { JupyterCIProvider };
