// @tarko/llm-client/dist/index.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { lookup } from "mime-types";
import sdk from "@anthropic-ai/sdk";
import { FinishReason, FunctionCallingMode, GoogleGenerativeAI } from "@google/generative-ai";
import { nanoid } from "nanoid";
import openai_0, { AzureOpenAI } from "openai";
import mistralai from "@mistralai/mistralai";
const TableDisplayNames = {
    models: 'Model',
    supportsCompletion: 'Chat Completion',
    supportsStreaming: 'Streaming',
    supportsJSON: 'JSON Output',
    supportsImages: 'Image Input',
    supportsToolCalls: 'Function Calling',
    supportsN: 'N > 1'
};
const models_models = {
    openai: {
        models: [
            'gpt-image-1',
            'gpt-4.5-preview',
            'gpt-4.5-preview-2025-02-27',
            'gpt-4.1',
            'gpt-4.1-2025-04-14',
            'gpt-4o',
            'gpt-4o-mini',
            'gpt-4o-2024-05-13',
            'gpt-4o-2024-08-06',
            'gpt-4-turbo',
            'gpt-4-turbo-2024-04-09',
            'gpt-4-0125-preview',
            'gpt-4-turbo-preview',
            'gpt-4-1106-preview',
            'gpt-4-vision-preview',
            'gpt-4',
            'gpt-4-0314',
            'gpt-4-0613',
            'gpt-4-32k',
            'gpt-4-32k-0314',
            'gpt-4-32k-0613',
            'gpt-3.5-turbo',
            'gpt-3.5-turbo-16k',
            'gpt-3.5-turbo-0301',
            'gpt-3.5-turbo-0613',
            'gpt-3.5-turbo-1106',
            'gpt-3.5-turbo-0125',
            'gpt-3.5-turbo-16k-0613',
            'o3-mini',
            'o1-mini',
            'o1-mini-2024-09-12',
            'o1-preview',
            'o1-preview-2024-09-12'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'gpt-4.1',
            'gpt-4.1-2025-04-14',
            'gpt-4.5-preview',
            'gpt-4.5-preview-2025-02-27',
            'gpt-4o',
            'gpt-4o-mini',
            'gpt-4o-2024-05-13',
            'gpt-4o-2024-08-06',
            'gpt-4-turbo',
            'gpt-4-turbo-2024-04-09',
            'gpt-4-0125-preview',
            'gpt-4-turbo-preview',
            'gpt-4-1106-preview',
            'gpt-4-vision-preview',
            'gpt-4',
            'gpt-4-0314',
            'gpt-4-0613',
            'gpt-4-32k',
            'gpt-4-32k-0314',
            'gpt-4-32k-0613',
            'gpt-3.5-turbo',
            'gpt-3.5-turbo-16k',
            'gpt-3.5-turbo-0301',
            'gpt-3.5-turbo-0613',
            'gpt-3.5-turbo-1106',
            'gpt-3.5-turbo-0125',
            'gpt-3.5-turbo-16k-0613',
            'o3-mini'
        ],
        supportsJSON: [
            'gpt-4.1',
            'gpt-4.1-2025-04-14',
            'gpt-4.5-preview',
            'gpt-4.5-preview-2025-02-27',
            'gpt-4o',
            'gpt-4o-mini',
            'gpt-4o-2024-05-13',
            'gpt-4o-2024-08-06',
            'gpt-4-turbo',
            'gpt-4-turbo-2024-04-09',
            'gpt-4-0125-preview',
            'gpt-4-turbo-preview',
            'gpt-4-1106-preview',
            'gpt-4-vision-preview',
            'gpt-3.5-turbo',
            'gpt-3.5-turbo-1106',
            'gpt-3.5-turbo-0125',
            'o3-mini'
        ],
        supportsImages: [
            'gpt-4.1',
            'gpt-4.1-2025-04-14',
            'gpt-4.5-preview',
            'gpt-4.5-preview-2025-02-27',
            'gpt-4o',
            'gpt-4o-mini',
            'gpt-4o-2024-05-13',
            'gpt-4o-2024-08-06',
            'gpt-4-turbo',
            'gpt-4-turbo-2024-04-09',
            'gpt-4-0125-preview',
            'gpt-4-turbo-preview',
            'gpt-4-1106-preview',
            'gpt-4-vision-preview'
        ],
        supportsToolCalls: [
            'gpt-4.1',
            'gpt-4.1-2025-04-14',
            'gpt-4.5-preview',
            'gpt-4.5-preview-2025-02-27',
            'gpt-4o',
            'gpt-4o-mini',
            'gpt-4o-2024-05-13',
            'gpt-4o-2024-08-06',
            'gpt-4-turbo',
            'gpt-4-turbo-2024-04-09',
            'gpt-4-turbo-preview',
            'gpt-4-0125-preview',
            'gpt-4-1106-preview',
            'gpt-4',
            'gpt-4-0613',
            'gpt-3.5-turbo',
            'gpt-3.5-turbo-0125',
            'gpt-3.5-turbo-1106',
            'gpt-3.5-turbo-0613',
            'o3-mini'
        ],
        supportsN: true,
        generateDocs: true
    },
    'openai-non-streaming': {
        models: [
            'gpt-image-1',
            'o1-mini',
            'o1-mini-2024-09-12',
            'o1-preview',
            'o1-preview-2024-09-12'
        ],
        supportsCompletion: true,
        supportsStreaming: [],
        supportsJSON: [
            'o1-mini',
            'o1-mini-2024-09-12',
            'o1-preview',
            'o1-preview-2024-09-12'
        ],
        supportsImages: [
            'gpt-image-1'
        ],
        supportsToolCalls: [
            'o1-mini',
            'o1-mini-2024-09-12',
            'o1-preview',
            'o1-preview-2024-09-12'
        ],
        supportsN: true,
        generateDocs: true
    },
    ai21: {
        models: [
            'jamba-instruct'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'jamba-instruct'
        ],
        supportsJSON: [],
        supportsImages: [],
        supportsToolCalls: [],
        supportsN: true,
        generateDocs: true
    },
    anthropic: {
        models: [
            'claude-3-7-sonnet-latest',
            'claude-3-7-sonnet-20250219',
            'claude-3-5-sonnet-latest',
            'claude-3-5-sonnet-20240620',
            'claude-3-5-haiku-20241022',
            'claude-3-opus-20240229',
            'claude-3-sonnet-20240229',
            'claude-3-haiku-20240307',
            'claude-2.1',
            'claude-2.0',
            'claude-instant-1.2'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'claude-3-7-sonnet-latest',
            'claude-3-7-sonnet-20250219',
            'claude-3-5-sonnet-latest',
            'claude-3-5-sonnet-20240620',
            'claude-3-5-haiku-20241022',
            'claude-3-opus-20240229',
            'claude-3-sonnet-20240229',
            'claude-3-haiku-20240307',
            'claude-2.1',
            'claude-2.0',
            'claude-instant-1.2'
        ],
        supportsJSON: [],
        supportsImages: [
            'claude-3-7-sonnet-latest',
            'claude-3-7-sonnet-20250219',
            'claude-3-5-sonnet-latest',
            'claude-3-5-sonnet-20240620',
            'claude-3-opus-20240229',
            'claude-3-sonnet-20240229',
            'claude-3-haiku-20240307'
        ],
        supportsToolCalls: [
            'claude-3-7-sonnet-latest',
            'claude-3-7-sonnet-20250219',
            'claude-3-5-sonnet-latest',
            'claude-3-5-sonnet-20240620',
            'claude-3-5-haiku-20241022',
            'claude-3-opus-20240229',
            'claude-3-sonnet-20240229',
            'claude-3-haiku-20240307'
        ],
        supportsN: false,
        generateDocs: true
    },
    gemini: {
        models: [
            'gemini-2.0-flash-001',
            'gemini-2.0-flash-lite-preview-02-05',
            'gemini-1.5-pro',
            'gemini-1.5-flash',
            'gemini-1.5-flash-8b',
            'gemini-1.0-pro'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'gemini-2.0-flash-001',
            'gemini-2.0-flash-lite-preview-02-05',
            'gemini-1.5-pro',
            'gemini-1.5-flash',
            'gemini-1.5-flash-8b',
            'gemini-1.0-pro'
        ],
        supportsJSON: [
            'gemini-2.0-flash-001',
            'gemini-2.0-flash-lite-preview-02-05',
            'gemini-1.5-pro',
            'gemini-1.5-flash',
            'gemini-1.5-flash-8b'
        ],
        supportsImages: [
            'gemini-2.0-flash-001',
            'gemini-2.0-flash-lite-preview-02-05',
            'gemini-1.5-pro',
            'gemini-1.5-flash',
            'gemini-1.5-flash-8b'
        ],
        supportsToolCalls: [
            'gemini-2.0-flash-001',
            'gemini-1.5-pro',
            'gemini-1.5-flash',
            'gemini-1.5-flash-8b',
            'gemini-1.0-pro'
        ],
        supportsN: true,
        generateDocs: true
    },
    mistral: {
        models: [
            'open-mistral-7b',
            'mistral-tiny-2312',
            'open-mixtral-8x7b',
            'mistral-small-2312',
            'open-mixtral-8x22b',
            'open-mixtral-8x22b-2404',
            'mistral-small-latest',
            'mistral-small-2402',
            'mistral-medium-latest',
            'mistral-medium-2312',
            'mistral-large-latest',
            'mistral-large-2402',
            'codestral-latest',
            'codestral-2405',
            'codestral-mamba-2407'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'open-mistral-7b',
            'mistral-tiny-2312',
            'open-mixtral-8x7b',
            'mistral-small-2312',
            'open-mixtral-8x22b',
            'open-mixtral-8x22b-2404',
            'mistral-small-latest',
            'mistral-small-2402',
            'mistral-medium-latest',
            'mistral-medium-2312',
            'mistral-large-latest',
            'mistral-large-2402',
            'codestral-latest',
            'codestral-2405',
            'codestral-mamba-2407'
        ],
        supportsJSON: [
            'open-mistral-7b',
            'mistral-tiny-2312',
            'open-mixtral-8x22b',
            'open-mixtral-8x22b-2404',
            'mistral-large-latest',
            'mistral-large-2402',
            'codestral-latest',
            'codestral-2405',
            'codestral-mamba-2407'
        ],
        supportsImages: [],
        supportsToolCalls: [
            'open-mixtral-8x22b',
            'open-mixtral-8x22b-2404',
            'mistral-small-latest',
            'mistral-small-2402',
            'mistral-large-latest',
            'mistral-large-2402',
            'codestral-mamba-2407'
        ],
        supportsN: false,
        generateDocs: true
    },
    groq: {
        models: [
            'llama-3.3-70b-versatile',
            'llama-3.1-8b-instant',
            'llama3-8b-8192',
            'llama3-70b-8192',
            'mixtral-8x7b-32768',
            'gemma-7b-it',
            'gemma2-9b-it'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'llama-3.3-70b-versatile',
            'llama-3.1-8b-instant',
            'llama3-8b-8192',
            'llama3-70b-8192',
            'mixtral-8x7b-32768',
            'gemma-7b-it',
            'gemma2-9b-it'
        ],
        supportsJSON: [
            'llama-3.3-70b-versatile',
            'llama-3.1-8b-instant',
            'llama3-70b-8192',
            'gemma-7b-it',
            'gemma2-9b-it'
        ],
        supportsImages: [],
        supportsToolCalls: [],
        supportsN: false,
        generateDocs: true
    },
    perplexity: {
        models: [
            'llama-3.1-sonar-small-128k-online',
            'llama-3.1-sonar-large-128k-online',
            'llama-3.1-sonar-huge-128k-online'
        ],
        supportsCompletion: true,
        supportsStreaming: [
            'llama-3.1-sonar-small-128k-online',
            'llama-3.1-sonar-large-128k-online',
            'llama-3.1-sonar-huge-128k-online'
        ],
        supportsJSON: [],
        supportsImages: [],
        supportsToolCalls: [],
        supportsN: false,
        generateDocs: true
    },
    openrouter: {
        models: true,
        supportsCompletion: true,
        supportsStreaming: true,
        supportsJSON: true,
        supportsImages: true,
        supportsToolCalls: true,
        supportsN: true,
        generateDocs: false
    },
    'openai-compatible': {
        models: true,
        supportsCompletion: true,
        supportsStreaming: true,
        supportsJSON: true,
        supportsImages: true,
        supportsToolCalls: true,
        supportsN: true,
        generateDocs: false
    },
    'azure-openai': {
        models: true,
        supportsCompletion: true,
        supportsStreaming: true,
        supportsJSON: true,
        supportsImages: true,
        supportsToolCalls: true,
        supportsN: true,
        generateDocs: false
    }
};
class InputError extends Error {
    constructor(message){
        super(message);
        this.name = this.constructor.name;
        Error.captureStackTrace(this, this.constructor);
    }
}
class InvariantError extends Error {
    constructor(message){
        super(`${message}\nShould never happen. Please report this error to the developers.`);
        this.name = 'InvariantError';
        Error.captureStackTrace(this, this.constructor);
    }
}
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class BaseHandler {
    validateInputs(body) {
        var _body_response_format;
        delete body.provider;
        if (!this.isSupportedModel(body.model)) throw new InputError(`Invalid 'model' field: ${body.model}.`);
        if (body.stream && !this.supportsStreaming(body.model)) throw new Error(`Detected 'stream: true', but the following model does not support streaming: ${body.model}`);
        if (void 0 !== body.tools && !this.supportsTools(body.model)) throw new InputError(`Detected a 'tools' parameter, but the following model does not support tools: ${body.model}`);
        if (void 0 !== body.tool_choice && !this.supportsTools(body.model)) throw new InputError(`Detected a 'tool_choice' parameter, but the following model does not support tools: ${body.model}`);
        if ('number' == typeof body.temperature && body.temperature > 2) throw new InputError(`Expected a temperature less than or equal to 2, but got: ${body.temperature}`);
        if ('number' == typeof body.top_p && body.top_p > 1) throw new InputError(`Expected a top_p less than or equal to 1.0, but got: ${body.top_p}`);
        for (const message of body.messages){
            if ('function' === message.role) throw new InputError("The 'function' role is deprecated. Please use the 'tool' role instead.");
            if ('user' === message.role) {
                if (Array.isArray(message.content)) {
                    for (const content of message.content)if ('image_url' === content.type && !this.supportsImageMessages(body.model)) throw new InputError(`Detected an image in the 'messages' array, but the following model does not support images: ${body.model}`);
                }
            }
        }
        if ('number' == typeof body.n && body.n > 1 && !this.supportsNGreaterThanOne(body.model)) throw new InputError(`The model ${body.model} does not support setting 'n' greater than 1.`);
        if ((null == (_body_response_format = body.response_format) ? void 0 : _body_response_format.type) === 'json_object') {
            if (!this.supportsJSONMode(body.model)) throw new InputError(`The model ${body.model} does not support the 'response_format' type 'json_object'.`);
            let containsJSONString = false;
            for (const message of body.messages)if ('string' == typeof message.content) {
                if (message.content.toLowerCase().includes('json')) {
                    containsJSONString = true;
                    break;
                }
            } else if (Array.isArray(message.content)) {
                for (const e of message.content)if ('text' === e.type) {
                    if (e.text.toLowerCase().includes('json')) {
                        containsJSONString = true;
                        break;
                    }
                }
            }
            if (!containsJSONString) throw new InputError("You must include the string 'JSON' somewhere in your prompt when the 'response_format' type is 'json_object'.");
        }
    }
    isSupportedFeature(featureSupport, model) {
        if ('boolean' == typeof featureSupport) return featureSupport;
        return featureSupport.includes(model);
    }
    isSupportedModel(model) {
        return this.isSupportedFeature(this.models, model);
    }
    supportsJSONMode(model) {
        return this.isSupportedFeature(this.supportsJSON, model);
    }
    supportsImageMessages(model) {
        return this.isSupportedFeature(this.supportsImages, model);
    }
    supportsNGreaterThanOne(model) {
        return this.isSupportedFeature(this.supportsN, model);
    }
    supportsTools(model) {
        return this.isSupportedFeature(this.supportsToolCalls, model);
    }
    supportsStreaming(model) {
        return this.isSupportedFeature(this.supportsStreamingMessages, model);
    }
    constructor(opts, models, supportsJSON, supportsImages, supportsToolCalls, suportsN, supportsStreamingMessages){
        _define_property(this, "opts", void 0);
        _define_property(this, "models", void 0);
        _define_property(this, "supportsJSON", void 0);
        _define_property(this, "supportsImages", void 0);
        _define_property(this, "supportsToolCalls", void 0);
        _define_property(this, "supportsN", void 0);
        _define_property(this, "supportsStreamingMessages", void 0);
        this.opts = opts;
        this.models = models;
        this.supportsJSON = supportsJSON;
        this.supportsImages = supportsImages;
        this.supportsToolCalls = supportsToolCalls;
        this.supportsN = suportsN;
        this.supportsStreamingMessages = supportsStreamingMessages;
    }
}
const convertMessages = (messages)=>{
    const output = [];
    let previousRole = 'user';
    let currentMessages = [];
    for(let i = 0; i < messages.length; i++){
        const message = messages[i];
        if (0 === i && 'system' === message.role) output.push({
            role: 'system',
            content: convertMessageContentToString(message.content)
        });
        else if ('user' === message.role || 'assistant' === message.role || 'system' === message.role) {
            const newRole = 'user' === message.role || 'system' === message.role ? 'user' : 'assistant';
            if (previousRole !== newRole) {
                output.push({
                    role: previousRole,
                    content: currentMessages.join('\n')
                });
                currentMessages = [];
            }
            if ('string' == typeof message.content) {
                const text = 'system' === message.role ? `System: ${message.content}` : message.content;
                currentMessages.push(text);
            } else if (Array.isArray(message.content)) {
                const convertedContent = message.content.map((e)=>{
                    if ('text' === e.type) {
                        const text = 'system' === message.role ? `System: ${e.text}` : e.text;
                        return text;
                    }
                    throw new InputError("AI21 does not support images.");
                });
                currentMessages.push(...convertedContent);
            }
            previousRole = newRole;
        }
    }
    if (currentMessages.length > 0) output.push({
        role: previousRole,
        content: currentMessages.join('\n')
    });
    return output;
};
async function* createCompletionResponseStreaming(responseStream, model, created) {
    const reader = responseStream.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    try {
        while(true){
            const { done, value } = await reader.read();
            if (done) break;
            buffer += decoder.decode(value, {
                stream: true
            });
            const lines = buffer.split('\n');
            buffer = lines.pop() || '';
            for (const line of lines){
                if ('data: [DONE]' === line.trim()) return;
                if (line.trim().startsWith('data:')) {
                    const response = JSON.parse(line.replace(/^data:\s*/, '').trim());
                    let content = '';
                    for (const choice of response.choices)if ('string' == typeof choice.delta.content) content += choice.delta.content;
                    const { id, choices } = response;
                    const finish_reason = choices[0].finish_reason;
                    yield {
                        choices: [
                            {
                                index: 0,
                                finish_reason,
                                logprobs: null,
                                delta: {
                                    role: 'assistant',
                                    content
                                }
                            }
                        ],
                        id,
                        created,
                        model,
                        object: 'chat.completion.chunk'
                    };
                }
            }
        }
    } finally{
        reader.releaseLock();
    }
}
class AI21Handler extends BaseHandler {
    validateInputs(body) {
        super.validateInputs(body);
        if ('number' == typeof body.n && (body.n > 16 || body.n < 0)) throw new InputError(`AI21 requires that the 'n' parameter is a value between 0 and 16, inclusive. Instead, got: ${body.n}`);
        if ('number' == typeof body.n && true === body.stream && body.n > 1) throw new InputError(`AI21 requires that 'n' equals '1' when streaming is enabled. Received an 'n' value of: ${body.n}`);
    }
    async create(body) {
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.AI21_API_KEY;
        if (void 0 === apiKey) throw new InputError("No AI21 API key detected. Please define an 'AI21_API_KEY' environment variable or supply the API key using the 'apiKey' parameter.");
        const messages = convertMessages(body.messages);
        const params = {
            max_tokens: body.max_tokens ?? void 0,
            messages,
            model: body.model,
            n: body.n ?? void 0,
            stop: body.stop ?? void 0,
            temperature: body.temperature ?? void 0,
            top_p: body.top_p ?? void 0,
            stream: body.stream ?? void 0
        };
        const created = getTimestamp();
        const response = await fetch('https://api.ai21.com/studio/v1/chat/completions', {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${apiKey}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(params)
        });
        if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
        if (true === body.stream) return createCompletionResponseStreaming(response.body, body.model, created);
        {
            const data = await response.json();
            const convertedChoices = data.choices.map((choice)=>({
                    ...choice,
                    message: {
                        ...choice.message,
                        refusal: null
                    },
                    logprobs: null
                }));
            const convertedResponse = {
                choices: convertedChoices,
                id: data.id,
                usage: data.usage,
                created,
                model: body.model,
                object: 'chat.completion'
            };
            return convertedResponse;
        }
    }
}
const createCompletionResponseNonStreaming = (response, created, toolChoice)=>{
    const finishReason = toFinishReasonNonStreaming(response.stop_reason);
    const chatMessage = toChatCompletionChoiceMessage(response.content, response.role, toolChoice);
    const choice = {
        index: 0,
        logprobs: null,
        message: chatMessage,
        finish_reason: finishReason
    };
    const converted = {
        id: response.id,
        choices: [
            choice
        ],
        created,
        model: response.model,
        object: 'chat.completion',
        usage: {
            prompt_tokens: response.usage.input_tokens,
            completion_tokens: response.usage.output_tokens,
            total_tokens: response.usage.input_tokens + response.usage.output_tokens
        }
    };
    return converted;
};
async function* anthropic_createCompletionResponseStreaming(response, created) {
    let message;
    let initialToolCallIndex = null;
    for await (const chunk of response){
        if ('message_start' === chunk.type) {
            message = chunk.message;
            yield {
                choices: [
                    {
                        index: 0,
                        finish_reason: toFinishReasonStreaming(chunk.message.stop_reason),
                        logprobs: null,
                        delta: {
                            role: chunk.message.role
                        }
                    }
                ],
                created,
                model: message.model,
                id: message.id,
                object: 'chat.completion.chunk'
            };
        }
        if (void 0 === message) throw new InvariantError("Message is undefined.");
        let newStopReason;
        let delta = {};
        if ('content_block_start' === chunk.type) if ('text' === chunk.content_block.type) delta = {
            content: chunk.content_block.text
        };
        else {
            if (null === initialToolCallIndex) initialToolCallIndex = chunk.index;
            delta = {
                tool_calls: [
                    {
                        index: chunk.index - initialToolCallIndex,
                        id: chunk.content_block.id,
                        type: 'function',
                        function: {
                            name: chunk.content_block.name,
                            arguments: isEmptyObject(chunk.content_block.input) ? '' : JSON.stringify(chunk.content_block.input)
                        }
                    }
                ]
            };
        }
        else if ('content_block_delta' === chunk.type) if ('input_json_delta' === chunk.delta.type) {
            if (null === initialToolCallIndex) throw new InvariantError("Content block delta event came before a content block start event.");
            delta = {
                tool_calls: [
                    {
                        index: chunk.index - initialToolCallIndex,
                        function: {
                            arguments: chunk.delta.partial_json
                        }
                    }
                ]
            };
        } else delta = {
            content: chunk.delta.text
        };
        else if ('message_delta' === chunk.type) newStopReason = chunk.delta.stop_reason;
        const stopReason = void 0 !== newStopReason ? newStopReason : message.stop_reason;
        const finishReason = toFinishReasonStreaming(stopReason);
        const chunkChoice = {
            index: 0,
            finish_reason: finishReason,
            logprobs: null,
            delta
        };
        yield {
            choices: [
                chunkChoice
            ],
            created,
            model: message.model,
            id: message.id,
            object: 'chat.completion.chunk'
        };
    }
}
const isTextBlock = (contentBlock)=>'text' === contentBlock.type;
const isToolUseBlock = (contentBlock)=>'tool_use' === contentBlock.type;
const toChatCompletionChoiceMessage = (content, role, toolChoice)=>{
    const textBlocks = content.filter(isTextBlock);
    if (textBlocks.length > 1) consoleWarn("Received multiple text blocks from Anthropic, which is unexpected. Concatenating the text blocks into a single string.");
    let toolUseBlocks;
    if ('string' != typeof toolChoice && (null == toolChoice ? void 0 : toolChoice.type) === 'function') {
        const selected = content.filter(isToolUseBlock).find((block)=>block.name === toolChoice.function.name);
        if (!selected) throw new InvariantError(`Did not receive a tool use block from Anthropic for the function: ${toolChoice.function.name}`);
        toolUseBlocks = [
            selected
        ];
    } else toolUseBlocks = content.filter(isToolUseBlock);
    let toolCalls;
    if (toolUseBlocks.length > 0) toolCalls = toolUseBlocks.map((toolUse)=>({
            id: toolUse.id,
            function: {
                name: toolUse.name,
                arguments: JSON.stringify(toolUse.input)
            },
            type: 'function'
        }));
    if (0 !== textBlocks.length) return {
        role,
        refusal: null,
        content: textBlocks.map((textBlock)=>textBlock.text).join('\n'),
        tool_calls: toolCalls
    };
    {
        const messageContent = content.every(isToolUseBlock) ? null : '';
        return {
            role,
            refusal: null,
            content: messageContent,
            tool_calls: toolCalls
        };
    }
};
const toFinishReasonNonStreaming = (stopReason)=>{
    if (null === stopReason) throw new InvariantError("Detected a 'stop_reason' value of 'null' during a non-streaming call.");
    if ('end_turn' === stopReason || 'stop_sequence' === stopReason) return 'stop';
    if ('max_tokens' === stopReason) return 'length';
    if ('tool_use' === stopReason) return 'tool_calls';
    return 'unknown';
};
const convertToolParams = (toolChoice, tools)=>{
    if (void 0 === tools || 'none' === toolChoice) return {
        toolChoice: void 0,
        tools: void 0
    };
    const convertedTools = tools.map((tool)=>({
            name: tool.function.name,
            description: tool.function.description,
            input_schema: {
                type: 'object',
                ...tool.function.parameters
            }
        }));
    let convertedToolChoice;
    convertedToolChoice = void 0 === toolChoice || 'auto' === toolChoice ? {
        type: 'auto'
    } : 'required' === toolChoice ? {
        type: 'any'
    } : {
        type: 'tool',
        name: toolChoice.function.name
    };
    return {
        toolChoice: convertedToolChoice,
        tools: convertedTools
    };
};
const toFinishReasonStreaming = (stopReason)=>{
    if (null === stopReason) return null;
    if ('end_turn' === stopReason || 'stop_sequence' === stopReason) return 'stop';
    if ('max_tokens' === stopReason) return 'length';
    if ('tool_use' === stopReason) return 'tool_calls';
    return 'unknown';
};
const getDefaultMaxTokens = (model)=>{
    if ('claude-3-5-sonnet-20240620' === model || 'claude-3-opus-20240229' === model || 'claude-3-sonnet-20240229' === model || 'claude-3-haiku-20240307' === model || 'claude-2.1' === model || 'claude-2.0' === model || 'claude-instant-1.2' === model) return 4096;
    'claude-3-5-sonnet-latest' === model || 'claude-3-5-sonnet-20241022' === model || 'claude-3-7-sonnet-latest' === model;
    return 8192;
};
const anthropic_convertMessages = async (messages)=>{
    const output = [];
    const clonedMessages = structuredClone(messages);
    let systemMessage;
    if (clonedMessages.length > 0 && 'system' === clonedMessages[0].role) {
        systemMessage = convertMessageContentToString(clonedMessages[0].content);
        clonedMessages.shift();
    }
    if ('user' !== clonedMessages[0].role && 'system' !== clonedMessages[0].role) clonedMessages.unshift({
        role: 'user',
        content: 'Empty'
    });
    let previousRole = 'user';
    let currentParams = [];
    for (const message of clonedMessages){
        const newRole = 'user' === message.role || 'system' === message.role || 'tool' === message.role ? 'user' : 'assistant';
        if (previousRole !== newRole) {
            output.push({
                role: previousRole,
                content: currentParams
            });
            currentParams = [];
        }
        if ('tool' === message.role) {
            const toolResult = {
                tool_use_id: message.tool_call_id,
                content: message.content,
                type: 'tool_result'
            };
            currentParams.push(toolResult);
        } else if ('assistant' === message.role) {
            if ('string' == typeof message.content) currentParams.push({
                text: message.content,
                type: 'text'
            });
            if (Array.isArray(message.tool_calls)) {
                const convertedContent = message.tool_calls.map((toolCall)=>({
                        id: toolCall.id,
                        input: JSON.parse(toolCall.function.arguments),
                        name: toolCall.function.name,
                        type: 'tool_use'
                    }));
                currentParams.push(...convertedContent);
            }
        } else if ('string' == typeof message.content) {
            const text = 'system' === message.role ? `System: ${message.content}` : message.content;
            currentParams.push({
                type: 'text',
                text
            });
        } else if (Array.isArray(message.content)) {
            const convertedContent = await Promise.all(message.content.map(async (e)=>{
                if ('text' === e.type) {
                    const text = 'system' === message.role ? `System: ${e.text}` : e.text;
                    return {
                        type: 'text',
                        text
                    };
                }
                {
                    const parsedImage = await fetchThenParseImage(e.image_url.url);
                    return {
                        type: 'image',
                        source: {
                            data: parsedImage.content,
                            media_type: parsedImage.mimeType,
                            type: 'base64'
                        }
                    };
                }
            }));
            currentParams.push(...convertedContent);
        }
        previousRole = newRole;
    }
    if (currentParams.length > 0) output.push({
        role: previousRole,
        content: currentParams
    });
    return {
        messages: output,
        systemMessage
    };
};
const convertStopSequences = (stop)=>{
    if (null == stop) return;
    if ('string' == typeof stop) return [
        stop
    ];
    if (Array.isArray(stop) && stop.every((e)=>'string' == typeof e)) return stop;
    throw new Error(`Unknown stop sequence: ${stop}`);
};
const getApiKey = (apiKey)=>apiKey ?? process.env.ANTHROPIC_API_KEY;
class AnthropicHandler extends BaseHandler {
    validateInputs(body) {
        super.validateInputs(body);
        let logImageDetailWarning = false;
        for (const message of body.messages)if (Array.isArray(message.content)) {
            for (const e of message.content)if ('image_url' === e.type) {
                if (void 0 !== e.image_url.detail && 'auto' !== e.image_url.detail) logImageDetailWarning = true;
                if ('claude-instant-1.2' === body.model || 'claude-2.0' === body.model || 'claude-2.1' === body.model) throw new InputError(`Model '${body.model}' does not support images. Remove any images from the prompt or use Claude version 3 or later.`);
            }
        }
        if (logImageDetailWarning) consoleWarn("Anthropic does not support the 'detail' field for images. The default image quality will be used.");
    }
    async create(body) {
        this.validateInputs(body);
        const apiKey = getApiKey(this.opts.apiKey);
        if (void 0 === apiKey) throw new InputError("No Anthropic API key detected. Please define an 'ANTHROPIC_API_KEY' environment variable or supply the API key using the 'apiKey' parameter.");
        const stream = 'boolean' == typeof body.stream ? body.stream : void 0;
        const maxTokens = body.max_tokens ?? getDefaultMaxTokens(body.model);
        const client = new sdk({
            apiKey: getApiKey(this.opts.apiKey),
            defaultHeaders: this.opts.defaultHeaders
        });
        const stopSequences = convertStopSequences(body.stop);
        const topP = 'number' == typeof body.top_p ? body.top_p : void 0;
        const temperature = 'number' == typeof body.temperature ? body.temperature / 2 : void 0;
        const { messages, systemMessage } = await anthropic_convertMessages(body.messages);
        const { toolChoice, tools } = convertToolParams(body.tool_choice, body.tools);
        if (true === stream) {
            const convertedBody = {
                max_tokens: maxTokens,
                messages,
                model: body.model,
                stop_sequences: stopSequences,
                temperature,
                top_p: topP,
                stream,
                system: systemMessage,
                tools,
                tool_choice: toolChoice
            };
            const created = getTimestamp();
            const response = client.messages.stream(convertedBody);
            return anthropic_createCompletionResponseStreaming(response, created);
        }
        {
            const convertedBody = {
                max_tokens: maxTokens,
                messages,
                model: body.model,
                stop_sequences: stopSequences,
                temperature,
                top_p: topP,
                system: systemMessage,
                tools,
                tool_choice: toolChoice
            };
            const created = getTimestamp();
            const response = await client.messages.create(convertedBody);
            return createCompletionResponseNonStreaming(response, created, body.tool_choice);
        }
    }
}
const convertContentsToParts = async (contents, systemPrefix)=>{
    if (null == contents) return [];
    if ('string' == typeof contents) return [
        {
            text: `${systemPrefix}${contents}`
        }
    ];
    {
        const allParts = contents.map(async (part)=>{
            if ('text' === part.type) return {
                text: `${systemPrefix}${part.text}`
            };
            if ('image_url' === part.type) {
                const imageData = await fetchThenParseImage(part.image_url.url);
                return {
                    inlineData: {
                        mimeType: imageData.mimeType,
                        data: imageData.content
                    }
                };
            }
            throw new InputError(`Invalid content part type: ${part.type}. Must be "text" or "image_url".`);
        });
        return Promise.all(allParts);
    }
};
const convertRole = (role)=>{
    switch(role){
        case 'assistant':
            return 'model';
        case 'function':
        case 'tool':
        case 'user':
        case 'system':
            return 'user';
        default:
            throw new InputError(`Unexpected message role: ${role}`);
    }
};
const convertAssistantMessage = (message)=>{
    const parts = message.tool_calls ? message.tool_calls.map((call)=>({
            functionCall: {
                name: call.function.name,
                args: JSON.parse(call.function.arguments)
            }
        })) : [];
    if (null !== message.content) parts.push({
        text: message.content
    });
    return {
        role: convertRole(message.role),
        parts
    };
};
const convertMessageToContent = async (message, includeSystemPrefix)=>{
    switch(message.role){
        case 'tool':
            return {
                role: convertRole(message.role),
                parts: [
                    {
                        functionResponse: {
                            name: message.tool_call_id,
                            response: JSON.parse(convertMessageContentToString(message.content))
                        }
                    }
                ]
            };
        case 'assistant':
            return convertAssistantMessage(message);
        case 'user':
        case 'system':
            const systemPrefix = 'system' === message.role && includeSystemPrefix ? 'System:\n' : '';
            return {
                role: convertRole(message.role),
                parts: await convertContentsToParts(message.content, systemPrefix)
            };
        default:
            throw new InputError(`Unexpected message role: ${message.role}`);
    }
};
const convertMessagesToContents = async (messages)=>{
    const clonedMessages = structuredClone(messages);
    let systemInstruction;
    if (clonedMessages.length > 0 && 'system' === clonedMessages[0].role) {
        const systemMessage = clonedMessages.shift();
        systemInstruction = void 0 !== systemMessage ? await convertMessageToContent(systemMessage, false) : void 0;
    }
    const converted = [];
    for (const message of clonedMessages)if ('system' === message.role || 'user' === message.role) converted.push(await convertMessageToContent(message, true));
    else if ('assistant' === message.role) {
        converted.push(await convertMessageToContent(message, true));
        if (void 0 !== message.tool_calls) for (const assistantToolCall of message.tool_calls){
            const toolResult = clonedMessages.find((m)=>'tool' === m.role && m.tool_call_id === assistantToolCall.id);
            if (void 0 === toolResult) throw new Error(`Could not find tool message with the id: ${assistantToolCall.id}`);
            converted.push(await convertMessageToContent(toolResult, true));
        }
    }
    return {
        contents: converted,
        systemInstruction
    };
};
const convertFinishReason = (finishReason, parts)=>{
    if (null == parts ? void 0 : parts.some((part)=>void 0 !== part.functionCall)) return 'tool_calls';
    switch(finishReason){
        case FinishReason.STOP:
            return 'stop';
        case FinishReason.MAX_TOKENS:
            return 'length';
        case FinishReason.SAFETY:
            return 'content_filter';
        case FinishReason.OTHER:
        case FinishReason.FINISH_REASON_UNSPECIFIED:
        case FinishReason.RECITATION:
            return 'stop';
        default:
            return 'stop';
    }
};
const convertToolCalls = (candidate)=>{
    var _candidate_content;
    const toolCalls = null == (_candidate_content = candidate.content) ? void 0 : _candidate_content.parts.filter((part)=>void 0 !== part.functionCall).map((part, index)=>({
            id: nanoid(),
            index,
            function: {
                arguments: JSON.stringify(part.functionCall.args),
                name: part.functionCall.name
            },
            type: 'function'
        }));
    if (void 0 !== toolCalls && toolCalls.length > 0) return toolCalls;
};
const convertStreamToolCalls = (candidate)=>{
    var _convertToolCalls;
    return null == (_convertToolCalls = convertToolCalls(candidate)) ? void 0 : _convertToolCalls.map((toolCall, index)=>({
            ...toolCall,
            index
        }));
};
const convertResponseMessage = (candidate)=>{
    var _candidate_content;
    return {
        content: (null == (_candidate_content = candidate.content) ? void 0 : _candidate_content.parts.map((part)=>part.text).join('')) ?? null,
        role: 'assistant',
        tool_calls: convertToolCalls(candidate),
        refusal: null
    };
};
const convertUsageData = (usageMetadata)=>({
        completion_tokens: usageMetadata.candidatesTokenCount,
        prompt_tokens: usageMetadata.promptTokenCount,
        total_tokens: usageMetadata.totalTokenCount
    });
const convertToolConfig = (toolChoice, tools)=>{
    if ('object' == typeof toolChoice) return {
        functionCallingConfig: {
            mode: FunctionCallingMode.ANY,
            allowedFunctionNames: [
                toolChoice.function.name
            ]
        }
    };
    switch(toolChoice){
        case 'auto':
            return {
                functionCallingConfig: {
                    mode: FunctionCallingMode.AUTO
                }
            };
        case 'none':
            return {
                functionCallingConfig: {
                    mode: FunctionCallingMode.NONE
                }
            };
        case 'required':
            return {
                functionCallingConfig: {
                    mode: FunctionCallingMode.ANY
                }
            };
        default:
            return {
                functionCallingConfig: {
                    mode: tools && (null == tools ? void 0 : tools.length) > 0 ? FunctionCallingMode.AUTO : FunctionCallingMode.NONE
                }
            };
    }
};
const convertTools = (tools)=>{
    if (void 0 === tools) return;
    return tools.map((tool)=>({
            functionDeclarations: [
                {
                    name: tool.function.name,
                    description: tool.function.description,
                    parameters: tool.function.parameters
                }
            ]
        }));
};
const convertResponse = async (result, model, timestamp)=>{
    var _result_response_candidates;
    return {
        id: null,
        object: 'chat.completion',
        created: timestamp,
        model,
        choices: (null == (_result_response_candidates = result.response.candidates) ? void 0 : _result_response_candidates.map((candidate)=>{
            var _candidate_content;
            return {
                index: candidate.index,
                finish_reason: candidate.finishReason ? convertFinishReason(candidate.finishReason, null == (_candidate_content = candidate.content) ? void 0 : _candidate_content.parts) : 'stop',
                message: convertResponseMessage(candidate),
                logprobs: null
            };
        })) ?? [],
        usage: result.response.usageMetadata ? convertUsageData(result.response.usageMetadata) : void 0
    };
};
async function* convertStreamResponse(result, model, timestamp) {
    for await (const chunk of result.stream){
        var _chunk_candidates;
        const text = chunk.text();
        yield {
            id: null,
            object: 'chat.completion.chunk',
            created: timestamp,
            model,
            choices: (null == (_chunk_candidates = chunk.candidates) ? void 0 : _chunk_candidates.map((candidate)=>{
                var _candidate_content;
                return {
                    index: candidate.index,
                    finish_reason: candidate.finishReason ? convertFinishReason(candidate.finishReason, null == (_candidate_content = candidate.content) ? void 0 : _candidate_content.parts) : 'stop',
                    delta: {
                        content: text,
                        tool_calls: convertStreamToolCalls(candidate),
                        role: 'assistant'
                    },
                    logprobs: null
                };
            })) ?? [],
            usage: chunk.usageMetadata ? convertUsageData(chunk.usageMetadata) : void 0
        };
    }
}
class GeminiHandler extends BaseHandler {
    async create(body) {
        var _body_response_format;
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.GEMINI_API_KEY;
        if (void 0 === apiKey) throw new InputError('API key is required for Gemini, define GEMINI_API_KEY in your environment or specifty the apiKey option.');
        if (this.opts.baseURL) consoleWarn("The 'baseUrl' will be ignored by Gemini because it does not support this field.");
        const responseMimeType = (null == (_body_response_format = body.response_format) ? void 0 : _body_response_format.type) === 'json_object' ? 'application/json' : void 0;
        const stop = 'string' == typeof body.stop ? [
            body.stop
        ] : body.stop;
        const genAI = new GoogleGenerativeAI(apiKey);
        const model = genAI.getGenerativeModel({
            model: body.model,
            generationConfig: {
                maxOutputTokens: body.max_tokens ?? void 0,
                temperature: body.temperature ?? void 0,
                topP: body.top_p ?? void 0,
                stopSequences: stop ?? void 0,
                candidateCount: body.n ?? void 0,
                responseMimeType
            }
        });
        const { contents, systemInstruction } = await convertMessagesToContents(body.messages);
        const params = {
            contents,
            toolConfig: convertToolConfig(body.tool_choice, body.tools),
            tools: convertTools(body.tools),
            systemInstruction
        };
        const timestamp = getTimestamp();
        if (body.stream) {
            const result = await model.generateContentStream(params);
            return convertStreamResponse(result, body.model, timestamp);
        }
        {
            const result = await model.generateContent(params);
            return convertResponse(result, body.model, timestamp);
        }
    }
}
class GroqHandler extends BaseHandler {
    validateInputs(body) {
        var _body_response_format;
        super.validateInputs(body);
        if ((null == (_body_response_format = body.response_format) ? void 0 : _body_response_format.type) === 'json_object') {
            if (body.stream) throw new InputError("Groq does not support streaming when the 'response_format' is 'json_object'.");
            if (null !== body.stop && void 0 !== body.stop) throw new InputError("Groq does not support the 'stop' parameter when the 'response_format' is 'json_object'.");
        }
    }
    async create(body) {
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.GROQ_API_KEY;
        const client = new openai_0({
            apiKey,
            baseURL: 'https://api.groq.com/openai/v1',
            defaultHeaders: this.opts.defaultHeaders
        });
        if (void 0 === apiKey) throw new InputError('API key is required for Groq, define GROQ_API_KEY in your environment or specifty the apiKey option.');
        return client.chat.completions.create(body);
    }
}
const findLinkedToolCallName = (messages, toolCallId)=>{
    for (const message of messages)for (const toolCall of (null == message ? void 0 : message.tool_calls) ?? [])if (toolCall.id === toolCallId) return toolCall.function.name;
    throw new InputError(`Tool call with id ${toolCallId} not found in messages`);
};
const mistral_convertMessages = (messages)=>messages.map((message)=>{
        const messageContent = convertMessageContentToString(message.content);
        if ('tool' === message.role) {
            const name = findLinkedToolCallName(messages, message.tool_call_id);
            return {
                name,
                role: 'tool',
                content: messageContent,
                tool_call_id: message.tool_call_id
            };
        }
        if ('system' === message.role) return {
            role: message.role,
            content: messageContent
        };
        if ('assistant' === message.role) return {
            role: message.role,
            content: messageContent,
            tool_calls: message.tool_calls ?? null
        };
        if ('user' === message.role) {
            var _message_content;
            const content = 'string' == typeof message.content ? message.content : null == (_message_content = message.content) ? void 0 : _message_content.map((m)=>m.text);
            return {
                role: message.role,
                content
            };
        }
        throw new Error('Function messages are deprecated.');
    });
const mistral_convertTools = (tools, specificFunctionName)=>{
    if (!tools) return;
    const specifiedTool = tools.filter((tool)=>{
        if (void 0 === specificFunctionName) return true;
        return tool.function.name === specificFunctionName;
    });
    if (void 0 !== specificFunctionName && 0 === specifiedTool.length) throw new InputError(`Tool with name ${specificFunctionName} not found in tool list`);
    return specifiedTool.map((tool)=>({
            type: 'function',
            function: {
                name: tool.function.name,
                description: tool.function.description ?? '',
                parameters: tool.function.parameters ?? {}
            }
        }));
};
const mistral_convertToolConfig = (toolChoice, tools)=>{
    if ('object' == typeof toolChoice) return {
        toolChoice: 'any',
        tools: mistral_convertTools(tools, toolChoice.function.name)
    };
    switch(toolChoice){
        case 'auto':
            return {
                toolChoice: 'auto',
                tools: mistral_convertTools(tools)
            };
        case 'none':
            return {
                toolChoice: 'none',
                tools: mistral_convertTools(tools)
            };
        case 'required':
            return {
                toolChoice: 'any',
                tools: mistral_convertTools(tools)
            };
        case void 0:
            return {
                toolChoice: void 0,
                tools: mistral_convertTools(tools)
            };
        default:
            throw new InputError(`Invalid tool choice: ${toolChoice}`);
    }
};
const mistral_convertToolCalls = (toolResponse)=>{
    if (!toolResponse) return;
    return toolResponse.map((tool)=>({
            id: tool.id,
            type: 'function',
            function: {
                name: tool.function.name,
                arguments: tool.function.arguments
            }
        }));
};
const mistral_convertStreamToolCalls = (toolResponse)=>{
    var _convertToolCalls;
    if (!toolResponse) return;
    return null == (_convertToolCalls = mistral_convertToolCalls(toolResponse)) ? void 0 : _convertToolCalls.map((toolCall, index)=>({
            ...toolCall,
            index
        }));
};
async function* toStreamResponse(result) {
    for await (const chunk of result)yield {
        id: chunk.id,
        created: chunk.created,
        object: chunk.object,
        model: chunk.model,
        choices: chunk.choices.map((choice)=>({
                index: choice.index,
                delta: {
                    role: 'assistant',
                    content: choice.delta.content,
                    tool_calls: mistral_convertStreamToolCalls(choice.delta.tool_calls)
                },
                finish_reason: choice.finish_reason,
                logprobs: null
            })),
        usage: chunk.usage ?? void 0
    };
}
const toCompletionResponse = (result)=>({
        id: result.id,
        created: result.created,
        object: result.object,
        model: result.model,
        choices: result.choices.map((choice)=>({
                index: choice.index,
                message: {
                    role: 'assistant',
                    refusal: null,
                    content: choice.message.content,
                    tool_calls: mistral_convertToolCalls(choice.message.tool_calls)
                },
                finish_reason: choice.finish_reason,
                logprobs: null
            })),
        usage: result.usage
    });
class MistralHandler extends BaseHandler {
    async create(body) {
        var _body_response_format;
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.MISTRAL_API_KEY;
        if (void 0 === apiKey) throw new InputError('API key is required for Mistral, define MISTRAL_API_KEY in your environment or specifty the apiKey option.');
        const endpoint = this.opts.baseURL ?? void 0;
        const client = new mistralai(apiKey, endpoint);
        const responseFormat = (null == (_body_response_format = body.response_format) ? void 0 : _body_response_format.type) === 'json_object' ? {
            type: 'json_object'
        } : void 0;
        const temperature = 'number' == typeof body.temperature ? body.temperature / 2 : void 0;
        const { toolChoice, tools } = mistral_convertToolConfig(body.tool_choice, body.tools);
        const messages = mistral_convertMessages(body.messages);
        const options = {
            model: body.model,
            messages: messages,
            temperature,
            maxTokens: body.max_tokens ?? void 0,
            topP: body.top_p ?? void 0,
            responseFormat,
            toolChoice,
            tools
        };
        if (body.stream) {
            const chatResponseStream = client.chatStream(options);
            return toStreamResponse(chatResponseStream);
        }
        {
            const chatResponse = await client.chat(options);
            return toCompletionResponse(chatResponse);
        }
    }
}
function openai_compatible_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
async function* streamOpenAI(response) {
    for await (const chunk of response)yield chunk;
}
class OpenAICompatibleHandler extends BaseHandler {
    validateInputs(body) {
        super.validateInputs(body);
        if (!this.opts.baseURL) throw new InputError('No baseURL option provided for openai compatible provider. You must define a baseURL option to use a generic openai compatible API with Token.js');
    }
    async create(body) {
        this.validateInputs(body);
        const apiKey = this.determineAPIKey();
        const openai = new openai_0({
            ...this.opts,
            apiKey,
            defaultHeaders: this.opts.defaultHeaders
        });
        const params = body;
        delete params.provider;
        if (!body.stream) return openai.chat.completions.create(body);
        {
            const stream = await openai.chat.completions.create(body);
            return streamOpenAI(stream);
        }
    }
    constructor(...args){
        super(...args), openai_compatible_define_property(this, "determineAPIKey", ()=>{
            if (this.opts.apiKey) return this.opts.apiKey;
            if (process.env.OPENAI_COMPATIBLE_API_KEY) return process.env.OPENAI_COMPATIBLE_API_KEY;
            return '';
        });
    }
}
async function* openai_streamOpenAI(response) {
    for await (const chunk of response)yield chunk;
}
class OpenAIHandler extends BaseHandler {
    async create(body) {
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.OPENAI_API_KEY;
        const openai = new openai_0({
            ...this.opts,
            apiKey,
            defaultHeaders: this.opts.defaultHeaders
        });
        const params = body;
        delete params.provider;
        if (!body.stream) return openai.chat.completions.create(body);
        {
            const stream = await openai.chat.completions.create(body);
            return openai_streamOpenAI(stream);
        }
    }
}
async function* createSyntheticStream(response) {
    yield {
        id: response.id,
        object: 'chat.completion.chunk',
        created: response.created,
        model: response.model,
        choices: [
            {
                index: 0,
                delta: {
                    role: 'assistant'
                },
                finish_reason: null,
                logprobs: null
            }
        ],
        usage: void 0
    };
    for (const choice of response.choices){
        const content = choice.message.content;
        if (content) yield {
            id: response.id,
            object: 'chat.completion.chunk',
            created: response.created,
            model: response.model,
            choices: [
                {
                    index: choice.index,
                    delta: {
                        content
                    },
                    finish_reason: null,
                    logprobs: null
                }
            ],
            usage: void 0
        };
        if (choice.message.tool_calls) for (const toolCall of choice.message.tool_calls)yield {
            id: response.id,
            object: 'chat.completion.chunk',
            created: response.created,
            model: response.model,
            choices: [
                {
                    index: choice.index,
                    delta: {
                        tool_calls: [
                            {
                                index: 0,
                                id: toolCall.id,
                                type: toolCall.type,
                                function: toolCall.function
                            }
                        ]
                    },
                    finish_reason: null,
                    logprobs: null
                }
            ],
            usage: void 0
        };
        yield {
            id: response.id,
            object: 'chat.completion.chunk',
            created: response.created,
            model: response.model,
            choices: [
                {
                    index: choice.index,
                    delta: {},
                    finish_reason: choice.finish_reason,
                    logprobs: null
                }
            ],
            usage: response.usage
        };
    }
}
class OpenAINonStreamingHandler extends BaseHandler {
    async create(body) {
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.OPENAI_API_KEY;
        const openai = new openai_0({
            ...this.opts,
            apiKey,
            defaultHeaders: this.opts.defaultHeaders
        });
        const params = {
            ...body
        };
        delete params.provider;
        if (!body.stream) return openai.chat.completions.create(params);
        {
            const nonStreamingParams = {
                ...params,
                stream: false
            };
            const response = await openai.chat.completions.create(nonStreamingParams);
            return createSyntheticStream(response);
        }
    }
}
class OpenRouterHandler extends BaseHandler {
    validateInputs(body) {
        super.validateInputs(body);
    }
    async create(body) {
        this.validateInputs(body);
        console.log('open router');
        const apiKey = this.opts.apiKey ?? process.env.OPENROUTER_API_KEY;
        const client = new openai_0({
            apiKey,
            baseURL: 'https://openrouter.ai/api/v1',
            defaultHeaders: {
                'HTTP-Referer': 'docs.tokenjs.ai',
                'X-Title': 'Token.js',
                ...this.opts.defaultHeaders
            }
        });
        if (void 0 === apiKey) throw new InputError('API key is required for OpenRouter, define OPENROUTER_API_KEY in your environment or specifty the apiKey option.');
        return client.chat.completions.create(body);
    }
}
const PERPLEXITY_PREFIX = 'perplexity/';
async function* streamPerplexity(response) {
    for await (const chunk of response)yield chunk;
}
class PerplexityHandler extends BaseHandler {
    async create(body) {
        this.validateInputs(body);
        const apiKey = this.opts.apiKey ?? process.env.PERPLEXITY_API_KEY;
        if (void 0 === apiKey) throw new InputError('API key is required for Perplexity, define PERPLEXITY_API_KEY in your environment or specifty the apiKey option.');
        const openai = new openai_0({
            ...this.opts,
            baseURL: 'https://api.perplexity.ai',
            apiKey,
            defaultHeaders: this.opts.defaultHeaders
        });
        const model = body.model.replace(PERPLEXITY_PREFIX, '');
        const temperature = 2 === body.temperature ? 2 - Number.EPSILON : body.temperature;
        const options = {
            ...body,
            model,
            temperature
        };
        if (!options.stream) return openai.chat.completions.create(options);
        {
            const stream = await openai.chat.completions.create(options);
            return streamPerplexity(stream);
        }
    }
}
async function* azure_openai_streamOpenAI(response) {
    for await (const chunk of response)yield chunk;
}
class AzureOpenAIHandler extends BaseHandler {
    async create(body) {
        var _this_opts_azure, _this_opts_azure1, _this_opts_azure2;
        this.validateInputs(body);
        const apiVersion = (null == (_this_opts_azure = this.opts.azure) ? void 0 : _this_opts_azure.apiVersion) || '2024-10-01-preview';
        const endpoint = (null == (_this_opts_azure1 = this.opts.azure) ? void 0 : _this_opts_azure1.endpoint) ?? this.opts.baseURL;
        if (!endpoint) throw new InputError('Azure OpenAI endpoint is required.');
        const azureADTokenProvider = null == (_this_opts_azure2 = this.opts.azure) ? void 0 : _this_opts_azure2.azureADTokenProvider;
        const apiKey = this.opts.apiKey ?? process.env.AZURE_OPENAI_API_KEY;
        if (!apiKey && !azureADTokenProvider) throw new InputError('Either an API key or an Azure AD token provider must be provided for Azure OpenAI');
        const azureOpenAI = new AzureOpenAI({
            apiKey,
            azureADTokenProvider,
            apiVersion,
            endpoint,
            defaultHeaders: this.opts.defaultHeaders
        });
        const params = body;
        delete params.provider;
        if (!body.stream) return azureOpenAI.chat.completions.create(body);
        {
            const stream = await azureOpenAI.chat.completions.create(body);
            return azure_openai_streamOpenAI(stream);
        }
    }
}
const Handlers = {
    ['openai']: (opts)=>new OpenAIHandler(opts, models_models.openai.models, models_models.openai.supportsJSON, models_models.openai.supportsImages, models_models.openai.supportsToolCalls, models_models.openai.supportsN, models_models.openai.supportsStreaming),
    ['openai-non-streaming']: (opts)=>new OpenAINonStreamingHandler(opts, models_models["openai-non-streaming"].models, models_models["openai-non-streaming"].supportsJSON, models_models["openai-non-streaming"].supportsImages, models_models["openai-non-streaming"].supportsToolCalls, models_models["openai-non-streaming"].supportsN, true),
    ['anthropic']: (opts)=>new AnthropicHandler(opts, models_models.anthropic.models, models_models.anthropic.supportsJSON, models_models.anthropic.supportsImages, models_models.anthropic.supportsToolCalls, models_models.anthropic.supportsN, models_models.anthropic.supportsStreaming),
    ['gemini']: (opts)=>new GeminiHandler(opts, models_models.gemini.models, models_models.gemini.supportsJSON, models_models.gemini.supportsImages, models_models.gemini.supportsToolCalls, models_models.gemini.supportsN, models_models.gemini.supportsStreaming),
    ['mistral']: (opts)=>new MistralHandler(opts, models_models.mistral.models, models_models.mistral.supportsJSON, models_models.mistral.supportsImages, models_models.mistral.supportsToolCalls, models_models.mistral.supportsN, models_models.mistral.supportsStreaming),
    ['groq']: (opts)=>new GroqHandler(opts, models_models.groq.models, models_models.groq.supportsJSON, models_models.groq.supportsImages, models_models.groq.supportsToolCalls, models_models.groq.supportsN, models_models.groq.supportsStreaming),
    ['ai21']: (opts)=>new AI21Handler(opts, models_models.ai21.models, models_models.ai21.supportsJSON, models_models.ai21.supportsImages, models_models.ai21.supportsToolCalls, models_models.ai21.supportsN, models_models.ai21.supportsStreaming),
    ['perplexity']: (opts)=>new PerplexityHandler(opts, models_models.perplexity.models, models_models.perplexity.supportsJSON, models_models.perplexity.supportsImages, models_models.perplexity.supportsToolCalls, models_models.perplexity.supportsN, models_models.perplexity.supportsStreaming),
    ['openrouter']: (opts)=>new OpenRouterHandler(opts, models_models.openrouter.models, models_models.openrouter.supportsJSON, models_models.openrouter.supportsImages, models_models.openrouter.supportsToolCalls, models_models.openrouter.supportsN, models_models.openrouter.supportsStreaming),
    ['openai-compatible']: (opts)=>new OpenAICompatibleHandler(opts, models_models.openrouter.models, models_models.openrouter.supportsJSON, models_models.openrouter.supportsImages, models_models.openrouter.supportsToolCalls, models_models.openrouter.supportsN, models_models.openrouter.supportsStreaming),
    ['azure-openai']: (opts)=>new AzureOpenAIHandler(opts, models_models["azure-openai"].models, models_models["azure-openai"].supportsJSON, models_models["azure-openai"].supportsImages, models_models["azure-openai"].supportsToolCalls, models_models["azure-openai"].supportsN, models_models["azure-openai"].supportsStreaming)
};
const getHandler = (provider, opts)=>{
    for(const handlerKey in Handlers)if (provider === handlerKey) return Handlers[handlerKey](opts);
    throw new Error("Could not find provider for model. Are you sure the model name is correct and the provider is supported?");
};
const getTimestamp = ()=>Math.floor(new Date().getTime() / 1000);
const fetchImageAsBase64 = async (url)=>{
    const response = await fetch(url);
    const arrayBuffer = await response.arrayBuffer();
    return btoa(String.fromCharCode.apply(null, new Uint8Array(arrayBuffer)));
};
const getMimeType = (url)=>{
    const parsedUrl = new URL(url);
    const pathname = parsedUrl.pathname;
    const extension = pathname.split('.').pop();
    return lookup(extension);
};
const isUrl = (input)=>{
    try {
        const url = new URL(input);
        return 'http:' === url.protocol || 'https:' === url.protocol;
    } catch (_) {
        return false;
    }
};
const isBase64Image = (input)=>/^data:image\/[a-zA-Z]+;base64,/.test(input);
const fetchThenParseImage = async (urlOrBase64Image)=>{
    if (isUrl(urlOrBase64Image)) {
        const content = await fetchImageAsBase64(urlOrBase64Image);
        const mimeType = getMimeType(urlOrBase64Image);
        if (null === mimeType) throw new Error(`Failed to get the mime type for the URL: ${urlOrBase64Image}`);
        if (!isSupportedMIMEType(mimeType)) throw new InputError(`Unsupported MIME type: ${mimeType}`);
        return {
            content,
            mimeType
        };
    }
    if (isBase64Image(urlOrBase64Image)) return parseImage(urlOrBase64Image);
    throw new InputError('Invalid image URL.');
};
const isSupportedMIMEType = (value)=>'image/jpeg' === value || 'image/png' === value || 'image/gif' === value || 'image/webp' === value;
const parseImage = (image)=>{
    const parts = image.split(';base64,');
    if (2 === parts.length) {
        const mimeType = parts[0].replace('data:', '').toLowerCase();
        if (!isSupportedMIMEType(mimeType)) throw new InputError(`Unsupported MIME type: ${mimeType}`);
        return {
            content: parts[1],
            mimeType
        };
    }
    throw new InputError('Invalid image URL.');
};
const consoleWarn = (message)=>{
    console.warn(`Warning: ${message}\n`);
};
const isEmptyObject = (variable)=>variable && 'object' == typeof variable && variable.constructor === Object && 0 === Object.keys(variable).length;
const convertMessageContentToString = (messageContent)=>{
    if (!messageContent) return '';
    return ('string' == typeof messageContent ? messageContent : messageContent.map((m)=>m.text).join('\n')) ?? '';
};
function chat_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class LLMCompletions {
    create(body) {
        const handler = getHandler(body.provider, this.opts);
        return handler.create(body);
    }
    constructor(opts){
        chat_define_property(this, "opts", void 0);
        this.opts = opts;
    }
}
class LLMChat {
    constructor(opts){
        chat_define_property(this, "completions", void 0);
        this.completions = new LLMCompletions(opts);
    }
}
function src_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class TokenJS {
    extendedModelExist(provider, name) {
        return TokenJS.extendedModelList.some((model)=>model.provider === provider && model.name === name);
    }
    extendModelList(provider, name, featureSupport) {
        if (this.extendedModelExist(provider, name)) return this;
        const modelsRef = models_models[provider];
        modelsRef['models'] = [
            ...models_models[provider]['models'],
            name
        ];
        const isSupportedFeature = (_featureSupport, model)=>{
            if ('boolean' == typeof _featureSupport) return _featureSupport;
            return _featureSupport.includes(model);
        };
        if ('string' == typeof featureSupport) {
            if (isSupportedFeature(modelsRef.supportsJSON, featureSupport)) modelsRef['supportsJSON'] = [
                ...modelsRef.supportsJSON,
                name
            ];
            if (isSupportedFeature(modelsRef.supportsStreaming, featureSupport)) modelsRef['supportsStreaming'] = [
                ...modelsRef.supportsStreaming,
                name
            ];
            if (isSupportedFeature(modelsRef.supportsImages, featureSupport)) modelsRef['supportsImages'] = [
                ...modelsRef.supportsImages,
                name
            ];
            if (isSupportedFeature(modelsRef.supportsToolCalls, featureSupport)) modelsRef['supportsToolCalls'] = [
                ...modelsRef.supportsToolCalls,
                name
            ];
        } else {
            if (featureSupport.json) modelsRef['supportsJSON'] = [
                ...modelsRef.supportsJSON,
                name
            ];
            if (featureSupport.streaming) modelsRef['supportsStreaming'] = [
                ...modelsRef.supportsStreaming,
                name
            ];
            if (featureSupport.toolCalls) modelsRef['supportsToolCalls'] = [
                ...modelsRef.supportsToolCalls,
                name
            ];
            if (featureSupport.images) modelsRef['supportsImages'] = [
                ...modelsRef.supportsImages,
                name
            ];
        }
        TokenJS.extendedModelList.push({
            provider,
            name,
            featureSupport
        });
        return this;
    }
    constructor({ ...opts } = {}){
        src_define_property(this, "opts", void 0);
        src_define_property(this, "chat", void 0);
        this.opts = opts;
        this.chat = new LLMChat(opts);
    }
}
src_define_property(TokenJS, "extendedModelList", []);
export { TableDisplayNames, TokenJS, models_models as models };
