// @tarko/agent-interface/dist/tool.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
import { z } from "@tarko/model-provider/types";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
function isZodSchema(schema) {
    return schema instanceof z.ZodObject;
}
function isJsonSchema(schema) {
    return null !== schema && 'object' == typeof schema && !isZodSchema(schema) && ('object' === schema.type || void 0 !== schema.properties);
}
class Tool {
    hasZodSchema() {
        return isZodSchema(this.schema);
    }
    hasJsonSchema() {
        return isJsonSchema(this.schema);
    }
    constructor(options){
        _define_property(this, "name", void 0);
        _define_property(this, "description", void 0);
        _define_property(this, "schema", void 0);
        _define_property(this, "function", void 0);
        this.name = options.id;
        this.description = options.description;
        this.schema = options.parameters;
        this.function = options.function;
    }
}
export { Tool };
