// @gui-agent/shared/dist/types/actions.mjs
/**
* Copyright (c) 2025 Bytedance, Inc. and its affiliates.
* SPDX-License-Identifier: Apache-2.0
*/
const ACTION_METADATA = {
    click: {
        category: 'mouse',
        description: 'Click on an element'
    },
    right_click: {
        category: 'mouse',
        description: 'Right click on an element'
    },
    double_click: {
        category: 'mouse',
        description: 'Double click on an element'
    },
    middle_click: {
        category: 'mouse',
        description: 'Middle click on an element'
    },
    mouse_down: {
        category: 'mouse',
        description: 'Press mouse button down'
    },
    mouse_up: {
        category: 'mouse',
        description: 'Release mouse button'
    },
    mouse_move: {
        category: 'mouse',
        description: 'Move mouse to position'
    },
    drag: {
        category: 'mouse',
        description: 'Drag from one position to another'
    },
    scroll: {
        category: 'mouse',
        description: 'Scroll in a direction'
    },
    type: {
        category: 'keyboard',
        description: 'Type text'
    },
    hotkey: {
        category: 'keyboard',
        description: 'Press hotkey combination'
    },
    press: {
        category: 'keyboard',
        description: 'Press a key'
    },
    release: {
        category: 'keyboard',
        description: 'Release a key'
    },
    navigate: {
        category: 'navigation',
        description: 'Navigate to URL'
    },
    navigate_back: {
        category: 'navigation',
        description: 'Navigate back'
    },
    long_press: {
        category: 'mobile',
        description: 'Long press on element'
    },
    swipe: {
        category: 'mobile',
        description: 'Swipe gesture'
    },
    home: {
        category: 'mobile',
        description: 'Go to home'
    },
    press_home: {
        category: 'mobile',
        description: 'Press home button'
    },
    back: {
        category: 'mobile',
        description: 'Go back'
    },
    press_back: {
        category: 'mobile',
        description: 'Press back button'
    },
    open_app: {
        category: 'mobile',
        description: 'Open application'
    },
    wait: {
        category: 'wait',
        description: 'Wait for specified time'
    },
    finished: {
        category: 'system',
        description: 'Mark task as finished'
    },
    call_user: {
        category: 'system',
        description: 'Request user interaction'
    }
};
function isSupportedActionType(type) {
    return type in ACTION_METADATA;
}
function isValidActionType(type) {
    return 'screenshot' === type || isSupportedActionType(type);
}
export { ACTION_METADATA, isSupportedActionType, isValidActionType };
